From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 1 Feb 2022 00:41:27 +1000
Subject: [PATCH] Sync beacons


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index f03383f0bbb402cf6c15f950ef6fee5fe1e26757..d2880c08e47f2ffd751b31d66195b68ec9f55b6c 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -151,6 +151,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
                         BeaconBlockEntity.this.secondaryPower = BeaconBlockEntity.getValidEffectById(value);
                 }
 
+                BeaconBlockEntity.this.setChanged(); // MultiPaper - send this block entity to other server
             }
 
             @Override
