From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 23 Jan 2022 00:07:41 +1000
Subject: [PATCH] Only play beacon deactivate sound if we aren't unloading the
 chunk


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index 49ca1d45bb4b3ddafc1d5952ff9830ba69b745e2..6337273e12f3e6f757911c846829cd12cd37a41c 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -282,6 +282,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
         org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(level, worldPosition);
         new io.papermc.paper.event.block.BeaconDeactivatedEvent(block).callEvent();
         // Paper end
+        if (this.level.getChunkIfLoaded(getBlockPos()) != null && this.level.getChunkIfLoaded(getBlockPos()).loaded) // MultiPaper - Only play beacon deactivate sound if we aren't unloading the chunk
         BeaconBlockEntity.playSound(this.level, this.worldPosition, SoundEvents.BEACON_DEACTIVATE);
         super.setRemoved();
     }
