From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 26 Jun 2022 16:01:21 +1000
Subject: [PATCH] Add option to not save lighting updates


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index d497283cc468daf9a35baf2be72de3b8d7d71518..4de86df9d9c980ec1c800c2bd76fb19b4c2b8a31 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -45,6 +45,7 @@ import net.minecraft.server.MinecraftServer;
 
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
+import puregero.multipaper.config.MultiPaperConfiguration;
 
 public class ChunkHolder {
 
@@ -239,6 +240,7 @@ public class ChunkHolder {
         LevelChunk chunk = this.getSendingChunk(); // Paper - no-tick view distance
 
         if (chunk != null) {
+            if (!MultiPaperConfiguration.get().optimizations.dontSaveJustForLightingUpdates) chunk.setUnsaved(true); // MultiPaper
             int i = this.levelHeightAccessor.getSectionIndex(pos.getY());
 
             if (i < 0 || i >= this.changedBlocksPerSection.length) return; // CraftBukkit - SPIGOT-6086, SPIGOT-6296
