From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 11 Feb 2022 00:13:18 +1000
Subject: [PATCH] Add --add-plugin-dir command line option


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9c93f53fbec78dd4e99a927b9839805ea2a2714d..e822037e37871a594302fceb0b5057c7bbc614e6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -481,6 +481,20 @@ public final class CraftServer implements Server {
             }
             list.add(file);
         }
+        // MultiPaper start
+        @SuppressWarnings("unchecked")
+        List<File> dirs = (List<File>) this.console.options.valuesOf("add-plugin-dir");
+        for (File dir : dirs) {
+            File[] files = dir.listFiles();
+            if (files != null) {
+                for (File file : files) {
+                    if (file.getName().endsWith(".jar")) {
+                        list.add(file);
+                    }
+                }
+            }
+        }
+        // MultiPaper end
         return list;
     }
     // Paper end
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d39622732cbf68af57cbc2c7a9afb468a55c6d7c..d27591225c8220c5bbf18173d0a20903f1cd4e81 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -172,6 +172,12 @@ public class Main {
                         .ofType(File.class)
                         .defaultsTo(new File("multipaper.yml"))
                         .describedAs("Yml file");
+
+                acceptsAll(asList("add-plugin-dir", "add-plugin-directory", "add-extra-plugin-jars"), "Specify paths of a directory containing extra plugin jars to be loaded in addition to those in the plugins folder. This argument can be specified multiple times, once for each extra plugin directory path.")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File[] {})
+                        .describedAs("Plugin directory");
                 // MultiPaper end
 
                 // Paper start
