From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul <paul@technove.co>
Date: Tue, 6 Jul 2021 17:51:18 -0500
Subject: [PATCH] Ignore lookups if closed


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index c85617aa7a48b5dd23cde767a80d1707815915f9..f94febbd90bec145d4cc4473df3a4babd376c178 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -46,6 +46,8 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
     private final Set<String> seenIllegalAccess = Collections.newSetFromMap(new ConcurrentHashMap<>());
     private java.util.logging.Logger logger; // Paper - add field
 
+    private boolean closed = false; // Pufferfish
+
     static {
         ClassLoader.registerAsParallelCapable();
     }
@@ -185,7 +187,7 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
         }
         Class<?> result = classes.get(name);
 
-        if (result == null) {
+        if (result == null && !this.closed) { // Pufferfish
             String path = name.replace('.', '/').concat(".class");
             JarEntry entry = jar.getJarEntry(path);
 
@@ -232,6 +234,7 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
             classes.put(name, result);
         }
 
+        if (result == null) throw new ClassNotFoundException(name); // Pufferfish
         return result;
     }
 
@@ -240,6 +243,7 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
         try {
             super.close();
         } finally {
+            this.closed = true; // Pufferfish
             jar.close();
         }
     }
