From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul <paul@technove.co>
Date: Sun, 26 Sep 2021 15:10:00 -0500
Subject: [PATCH] Temporarily pull my Paper PR


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index c2d8294ffeff017a6ec9e7725b50eaef8eb75dfd..6b7006829dde81840bd305ee775d26c9f1ef24f7 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -268,6 +268,9 @@ public class PaperWorldConfig {
 
     public int softDespawnDistance;
     public int hardDespawnDistance;
+    // Airplane start - separate squared values
+    public int softDespawnDistanceSq;
+    public int hardDespawnDistanceSq;
     private void despawnDistances() {
         softDespawnDistance = getInt("despawn-ranges.soft", 32); // 32^2 = 1024, Minecraft Default
         hardDespawnDistance = getInt("despawn-ranges.hard", 128); // 128^2 = 16384, Minecraft Default
@@ -278,8 +281,9 @@ public class PaperWorldConfig {
 
         log("Living Entity Despawn Ranges:  Soft: " + softDespawnDistance + " Hard: " + hardDespawnDistance);
 
-        softDespawnDistance = softDespawnDistance*softDespawnDistance;
-        hardDespawnDistance = hardDespawnDistance*hardDespawnDistance;
+        softDespawnDistanceSq = softDespawnDistance*softDespawnDistance;
+        hardDespawnDistanceSq = hardDespawnDistance*hardDespawnDistance;
+        // Airplane end
     }
 
     public boolean keepSpawnInMemory;
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index b8e512e1c4b00b468b2d22add5653b98f4a2c81a..6051fd771dd989f2903b854b6564252a847a74e3 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -801,16 +801,16 @@ public abstract class Mob extends LivingEntity {
                 int i = this.getType().getCategory().getDespawnDistance();
                 int j = i * i;
 
-                if (d0 > (double) level.paperConfig.hardDespawnDistance) { // CraftBukkit - remove isTypeNotPersistent() check // Paper - custom despawn distances
+                if (d0 > (double) level.paperConfig.hardDespawnDistanceSq) { // CraftBukkit - remove isTypeNotPersistent() check // Paper - custom despawn distances // Airplane
                     this.discard();
                 }
 
                 int k = this.getType().getCategory().getNoDespawnDistance();
                 int l = k * k;
 
-                if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > level.paperConfig.softDespawnDistance) { // CraftBukkit - remove isTypeNotPersistent() check // Paper - custom despawn distances
+                if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > level.paperConfig.softDespawnDistanceSq) { // CraftBukkit - remove isTypeNotPersistent() check // Paper - custom despawn distances // Airplane
                     this.discard();
-                } else if (d0 < level.paperConfig.softDespawnDistance) { // Paper - custom despawn distances
+                } else if (d0 < level.paperConfig.softDespawnDistanceSq) { // Paper - custom despawn distances // Airplane
                     this.noActionTime = 0;
                 }
             }
