From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 May 2024 18:44:13 +0900
Subject: [PATCH] Add isFoliaSupported


diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index c0691a849831f99268fdcb7b0f471f80a1a2a70e..caf7bc3faf34ebdd79aab33462eec67349d2094f 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -259,6 +259,7 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
     private Set<PluginAwareness> awareness = ImmutableSet.of();
     private String apiVersion = null;
     private List<String> libraries = ImmutableList.of();
+    private String foliaSupported = null; // MultiPaper
     // Paper start - oh my goddddd
     /**
      * Don't use this.
@@ -1049,6 +1050,11 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         return libraries;
     }
 
+    // MultiPaper start
+    public boolean isFoliaSupported() {
+        return "true".equalsIgnoreCase(foliaSupported);
+    }
+
     /**
      * @return unused
      * @deprecated unused
@@ -1233,6 +1239,12 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
             libraries = ImmutableList.<String>of();
         }
 
+        // MultiPaper start
+        if (map.get("folia-supported") != null) {
+            foliaSupported = map.get("folia-supported").toString();
+        }
+        // MultiPaper end
+
         try {
             lazyPermissions = (Map<?, ?>) map.get("permissions");
         } catch (ClassCastException ex) {
@@ -1311,6 +1323,12 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
             map.put("libraries", libraries);
         }
 
+        // MultiPaper start
+        if (foliaSupported != null) {
+            map.put("folia-supported", foliaSupported);
+        }
+        // MultiPaper end
+
         if (classLoaderOf != null) {
             map.put("class-loader-of", classLoaderOf);
         }
