From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 9 Dec 2021 18:10:03 +1000
Subject: [PATCH] Add MultiPaper API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 83571abfe9a2eb8736b481de35dfd7fd4c663f5a..ece88974d21856f58e01bab20843b9f8af91b09c 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -743,6 +743,28 @@ public final class Bukkit {
         return server.getServicesManager();
     }
 
+    // MultiPaper start
+    /**
+     * Gets a multipaper notification manager.
+     *
+     * @return a multipaper notification manager
+     */
+    @NotNull
+    public static MultiPaperNotificationManager getMultiPaperNotificationManager() {
+        return server.getMultiPaperNotificationManager();
+    }
+
+    /**
+     * Get the MultiPaper name of this server.
+     *
+     * @return the MultiPaper name of this server
+     */
+    @NotNull
+    public static String getLocalServerName() {
+        return server.getLocalServerName();
+    }
+    // MultiPaper end
+
     /**
      * Gets a list of all worlds on this server.
      *
diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index eccc357cabd8317afc5666c3347936d50eb34015..20f3655b82c5e6c7ae925a71d1b47899ab747b33 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -181,6 +181,26 @@ public interface Chunk extends PersistentDataHolder {
      */
     boolean unload();
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isExternalChunk();
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isLocalChunk();
+
+    // MultiPaper end
+
     /**
      * Checks if this chunk can spawn slimes without being a swamp biome.
      *
diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
index 6bbf8468bc47e82b0aeb164e49cdb73d93bae87b..8c6328ff7148a29b6fed328b0427a3a7eb0b56f4 100644
--- a/src/main/java/org/bukkit/Location.java
+++ b/src/main/java/org/bukkit/Location.java
@@ -346,6 +346,30 @@ public class Location implements Cloneable, ConfigurationSerializable, io.paperm
         return this;
     }
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isChunkExternal() {
+        return this.getWorld().isChunkExternal(locToBlock(x) >> 4, locToBlock(z) >> 4);
+    }
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isChunkLocal() {
+        return this.getWorld().isChunkLocal(locToBlock(x) >> 4, locToBlock(z) >> 4);
+    }
+
+    // MultiPaper end
+
     /**
      * Adds the location by another.
      *
diff --git a/src/main/java/org/bukkit/MultiPaperNotificationManager.java b/src/main/java/org/bukkit/MultiPaperNotificationManager.java
new file mode 100644
index 0000000000000000000000000000000000000000..d569e2c95f07e05d9fce019dafc33f9c51c4461a
--- /dev/null
+++ b/src/main/java/org/bukkit/MultiPaperNotificationManager.java
@@ -0,0 +1,91 @@
+package org.bukkit;
+
+import org.bukkit.plugin.Plugin;
+
+import java.nio.charset.StandardCharsets;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public interface MultiPaperNotificationManager {
+
+    /**
+     * Listen to notifications sent by other servers.
+     * 
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callback A handler for any data received
+     */
+    void on(Plugin plugin, String channel, Consumer<byte[]> callback);
+
+    /**
+     * Listen to notifications sent by other servers.
+     *
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callback A handler for any data received
+     */
+    default void onString(Plugin plugin, String channel, Consumer<String> callback) {
+        on(plugin, channel, bytes -> callback.accept(new String(bytes, StandardCharsets.UTF_8)));
+    }
+
+    /**
+     * Listen to notifications sent by other servers.
+     *
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callbackWithReply A handler for any data received, and a method to reply to the server on a specified channel
+     */
+    void on(Plugin plugin, String channel, BiConsumer<byte[], BiConsumer<String, byte[]>> callbackWithReply);
+
+    /**
+     * Listen to notifications sent by other servers.
+     *
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callbackWithReply A handler for any data received, and a method to reply to the server on a specified channel
+     */
+    default void onString(Plugin plugin, String channel, BiConsumer<String, BiConsumer<String, String>> callbackWithReply) {
+        on(plugin, channel, (bytes, reply) -> callbackWithReply.accept(
+                new String(bytes, StandardCharsets.UTF_8),
+                (replyChannel, string) -> reply.accept(replyChannel, string.getBytes(StandardCharsets.UTF_8)))
+        );
+    }
+
+    /**
+     * Notify other servers.
+     *
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    void notify(String channel, byte[] data);
+
+    /**
+     * Notify other servers.
+     *
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    default void notify(String channel, String data) {
+        notify(channel, data.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Notify other servers with the specified chunk loaded
+     *
+     * @param chunk The chunk that's loaded
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    void notify(Chunk chunk, String channel, byte[] data);
+
+    /**
+     * Notify other servers with the specified chunk loaded
+     *
+     * @param chunk The chunk that's loaded
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    default void notify(Chunk chunk, String channel, String data) {
+        notify(chunk, channel, data.getBytes(StandardCharsets.UTF_8));
+    }
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e0f69edf603c2ec99bc92b16b18912272cc41bd9..6e9fed841d9ac6b3671ebe7b4e8e9d2699403ba2 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -614,6 +614,24 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public ServicesManager getServicesManager();
 
+    // MultiPaper start
+    /**
+     * Gets a multipaper notification manager.
+     *
+     * @return a multipaper notification manager
+     */
+    @NotNull
+    public MultiPaperNotificationManager getMultiPaperNotificationManager();
+
+    /**
+     * Get the MultiPaper name of this server.
+     *
+     * @return the MultiPaper name of this server
+     */
+    @NotNull
+    public String getLocalServerName();
+    // MultiPaper end
+
     /**
      * Gets a list of all worlds on this server.
      *
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index a38863ebd363f54994753937a10e041076846af0..75cbd7b307d7c774bc6b128b88565d02288581a0 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -414,6 +414,26 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @Deprecated
     public boolean isChunkInUse(int x, int z);
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isChunkExternal(int x, int z);
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isChunkLocal(int x, int z);
+
+    // MultiPaper end
+
     /**
      * Loads the {@link Chunk} at the specified coordinates.
      * <p>
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index aba40431c5a1ae43106a678445483e0653e130d1..afaf534dca773f7381357c24fff327615b807526 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -439,6 +439,28 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
      */
     int getBlockPower();
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the block is in a chunk that's running on an external
+     * server or not.
+     *
+     * @return True if the block is in an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isInExternalChunk();
+
+    /**
+     * Returns whether the block is in a chunk that's running on this server
+     * or not.
+     *
+     * @return True if the block is in a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isInLocalChunk();
+
+    // MultiPaper end
+
     /**
      * Checks if this block is empty.
      * <p>
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 8a980916176b2179168833f1d14872176837154e..65349e06d50e7d3247ae36a36ff3cdf19b2d8029 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -277,6 +277,26 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public int getEntityId();
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the entity is in a chunk that's running on an external
+     * server or not.
+     *
+     * @return True if the entity is in an external chunk.
+     */
+    public boolean isInExternalChunk();
+
+    /**
+     * Returns whether the entity is in a chunk that's running on this server
+     * or not.
+     *
+     * @return True if the entity is in a local chunk.
+     */
+    public boolean isInLocalChunk();
+
+    // MultiPaper end
+
     /**
      * Returns the entity's current fire ticks (ticks before the entity stops
      * being on fire).
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 46dd9980b4ef09153c1ad9cbf447e7089ecfa6ac..86980bd78661e28d629bc6e2efce35f00fbda67e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -381,6 +381,84 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void setBedSpawnLocation(@Nullable Location location, boolean force);
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the player is on an external server or not.
+     *
+     * @return True if the player is on an external server.
+     */
+    public boolean isExternalPlayer();
+
+    /**
+     * Returns whether the player is on this server or not.
+     *
+     * @return True if the player is on this server.
+     */
+    public boolean isLocalPlayer();
+
+    /**
+     * Returns cross-server data that is stored under the specified key. Note
+     * that all plugins share the same set of keys. This data is
+     * non-persistent, it will be lost when the player disconnects.
+     *
+     * @param key The key the data is stored under.
+     * @return The data stored under the key, or null if the key isn't set.
+     */
+    @Nullable
+    public String getData(String key);
+
+    /**
+     * Store cross-server data under the specified key. Note that all plugins
+     * share the same set of keys. This data is non-persistent, it will be
+     * lost when the player disconnects.
+     *
+     * @param key The key to store the data under.
+     * @param value The data to store under the key.
+     */
+    public void setData(String key, String value);
+
+    /**
+     * Returns cross-server data that is stored under the specified key. Note
+     * that all plugins share the same set of keys. This data is persistent,
+     * it will be saved even if the player disconnects. This persistent data is
+     * saved onto the player's .dat file.
+     *
+     * @param key The key the data is stored under.
+     * @return The data stored under the key, or null if the key isn't set.
+     */
+    @Nullable
+    public String getPersistentData(String key);
+
+    /**
+     * Store cross-server data under the specified key. Note that all plugins
+     * share the same set of keys. This data is persistent, it will be saved
+     * even if the player disconnects. This persistent data is saved onto the
+     * player's .dat file.
+     *
+     * @param key The key to store the data under.
+     * @param value The data to store under the key.
+     */
+    public void setPersistentData(String key, String value);
+
+    /**
+     * Says a message (or runs a command) on other server excluding this one.
+     *
+     * @param msg chat message to say
+     */
+    public void chatOnOtherServers(@NotNull String msg);
+
+    /**
+     * Get the bungeecord name of the server that this player is on.
+     *
+     * @return The bungeecord name of the server the player is on for external
+     *         players, or null for local players.
+     */
+    @Nullable
+    public String getExternalServerName();
+
+    // MultiPaper end
+
     /**
      * Play a note for a player at a location. This requires a note block
      * at the particular location (as far as the client is concerned). This
