From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul <paul@technove.co>
Date: Tue, 6 Jul 2021 17:51:18 -0500
Subject: [PATCH] Ignore lookups if closed


diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
index 379b49f13a8c676125e17ad9207884a5d6cb5b7b..4c84e3a5c983b7f8fe5515fff6fcfa64106150d2 100644
--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
@@ -46,6 +46,8 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
     private final Set<String> seenIllegalAccess = Collections.newSetFromMap(new ConcurrentHashMap<>());
     private java.util.logging.Logger logger; // Paper - add field
 
+    private boolean closed = false; // Airplane
+
     static {
         ClassLoader.registerAsParallelCapable();
     }
@@ -159,7 +161,7 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
         }
         Class<?> result = classes.get(name);
 
-        if (result == null) {
+        if (result == null && !this.closed) { // Airplane
             String path = name.replace('.', '/').concat(".class");
             JarEntry entry = jar.getJarEntry(path);
 
@@ -214,6 +216,7 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
         try {
             super.close();
         } finally {
+            this.closed = true; // Airplane
             jar.close();
         }
     }
