From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 9 Dec 2021 18:10:03 +1000
Subject: [PATCH] Add MultiPaper API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 944f9b87a11472ac6d7e328acc00bf09f899e648..e2a659e5097163689011843a56269cacd059671f 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -742,6 +742,18 @@ public final class Bukkit {
         return server.getServicesManager();
     }
 
+    // MultiPaper start
+    /**
+     * Gets a multipaper notification manager.
+     *
+     * @return a multipaper notification manager
+     */
+    @NotNull
+    public static MultiPaperNotificationManager getMultiPaperNotificationManager() {
+        return server.getMultiPaperNotificationManager();
+    }
+    // MultiPaper end
+
     /**
      * Gets a list of all worlds on this server.
      *
diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 049c36807d2a970842442c1b7517c06f3f150041..1d18697f78007f173be44e17eaf43356603d7441 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -182,6 +182,26 @@ public interface Chunk extends PersistentDataHolder {
      */
     boolean unload();
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isExternalChunk();
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isLocalChunk();
+
+    // MultiPaper end
+
     /**
      * Checks if this chunk can spawn slimes without being a swamp biome.
      *
diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
index 8321441b8f528a05e297f485672f928e76fe017d..15b646c6cbf74ddbb3290bc0a0b95a2483fa9284 100644
--- a/src/main/java/org/bukkit/Location.java
+++ b/src/main/java/org/bukkit/Location.java
@@ -346,6 +346,30 @@ public class Location implements Cloneable, ConfigurationSerializable {
         return this;
     }
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isChunkExternal() {
+        return this.getWorld().isChunkExternal(locToBlock(x) >> 4, locToBlock(z) >> 4);
+    }
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isChunkLocal() {
+        return this.getWorld().isChunkLocal(locToBlock(x) >> 4, locToBlock(z) >> 4);
+    }
+
+    // MultiPaper end
+
     /**
      * Adds the location by another.
      *
diff --git a/src/main/java/org/bukkit/MultiPaperNotificationManager.java b/src/main/java/org/bukkit/MultiPaperNotificationManager.java
new file mode 100644
index 0000000000000000000000000000000000000000..d569e2c95f07e05d9fce019dafc33f9c51c4461a
--- /dev/null
+++ b/src/main/java/org/bukkit/MultiPaperNotificationManager.java
@@ -0,0 +1,91 @@
+package org.bukkit;
+
+import org.bukkit.plugin.Plugin;
+
+import java.nio.charset.StandardCharsets;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public interface MultiPaperNotificationManager {
+
+    /**
+     * Listen to notifications sent by other servers.
+     * 
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callback A handler for any data received
+     */
+    void on(Plugin plugin, String channel, Consumer<byte[]> callback);
+
+    /**
+     * Listen to notifications sent by other servers.
+     *
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callback A handler for any data received
+     */
+    default void onString(Plugin plugin, String channel, Consumer<String> callback) {
+        on(plugin, channel, bytes -> callback.accept(new String(bytes, StandardCharsets.UTF_8)));
+    }
+
+    /**
+     * Listen to notifications sent by other servers.
+     *
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callbackWithReply A handler for any data received, and a method to reply to the server on a specified channel
+     */
+    void on(Plugin plugin, String channel, BiConsumer<byte[], BiConsumer<String, byte[]>> callbackWithReply);
+
+    /**
+     * Listen to notifications sent by other servers.
+     *
+     * @param plugin The plugin listening to these notifications
+     * @param channel The notification channel to listen to
+     * @param callbackWithReply A handler for any data received, and a method to reply to the server on a specified channel
+     */
+    default void onString(Plugin plugin, String channel, BiConsumer<String, BiConsumer<String, String>> callbackWithReply) {
+        on(plugin, channel, (bytes, reply) -> callbackWithReply.accept(
+                new String(bytes, StandardCharsets.UTF_8),
+                (replyChannel, string) -> reply.accept(replyChannel, string.getBytes(StandardCharsets.UTF_8)))
+        );
+    }
+
+    /**
+     * Notify other servers.
+     *
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    void notify(String channel, byte[] data);
+
+    /**
+     * Notify other servers.
+     *
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    default void notify(String channel, String data) {
+        notify(channel, data.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Notify other servers with the specified chunk loaded
+     *
+     * @param chunk The chunk that's loaded
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    void notify(Chunk chunk, String channel, byte[] data);
+
+    /**
+     * Notify other servers with the specified chunk loaded
+     *
+     * @param chunk The chunk that's loaded
+     * @param channel The notification channel to notify on
+     * @param data The data to notify other servers with
+     */
+    default void notify(Chunk chunk, String channel, String data) {
+        notify(chunk, channel, data.getBytes(StandardCharsets.UTF_8));
+    }
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index cca362e54d6ff4a5a1e60f85a7eb1b3d222d3d48..41f02d9b75ce873721b406f9a60bad053841271f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -616,6 +616,16 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public ServicesManager getServicesManager();
 
+    // MultiPaper start
+    /**
+     * Gets a multipaper notification manager.
+     *
+     * @return a multipaper notification manager
+     */
+    @NotNull
+    public MultiPaperNotificationManager getMultiPaperNotificationManager();
+    // MultiPaper end
+
     /**
      * Gets a list of all worlds on this server.
      *
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 8a688583e65cd22e0417f9fd24e51803486d095e..f383c1d476be4333ccd08796aa8f8549b4b51fe6 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -423,6 +423,26 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @Deprecated
     public boolean isChunkInUse(int x, int z);
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isChunkExternal(int x, int z);
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isChunkLocal(int x, int z);
+
+    // MultiPaper end
+
     /**
      * Loads the {@link Chunk} at the specified coordinates.
      * <p>
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 3e980c630452c8ea72227bc4cd92c605253cd41b..8365e0f692fcc2e97cfa6d81e3b591b9c26ca166 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -427,6 +427,28 @@ public interface Block extends Metadatable, net.kyori.adventure.translation.Tran
      */
     int getBlockPower();
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the block is in a chunk that's running on an external
+     * server or not.
+     *
+     * @return True if the block is in an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isInExternalChunk();
+
+    /**
+     * Returns whether the block is in a chunk that's running on this server
+     * or not.
+     *
+     * @return True if the block is in a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isInLocalChunk();
+
+    // MultiPaper end
+
     /**
      * Checks if this block is empty.
      * <p>
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 8bc6876c82935988436597161fa0ec94c032174b..f3118dc0d024260f609a4a213d05e759d3b2baa0 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -209,6 +209,26 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public int getEntityId();
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the entity is in a chunk that's running on an external
+     * server or not.
+     *
+     * @return True if the entity is in an external chunk.
+     */
+    public boolean isInExternalChunk();
+
+    /**
+     * Returns whether the entity is in a chunk that's running on this server
+     * or not.
+     *
+     * @return True if the entity is in a local chunk.
+     */
+    public boolean isInLocalChunk();
+
+    // MultiPaper end
+
     /**
      * Returns the entity's current fire ticks (ticks before the entity stops
      * being on fire).
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f954801f6a5d465b8545e75d7ff5af0352d6ec0d..360f11e025b73ee14a6921aefe974fb40d22fcd1 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -363,6 +363,84 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void setBedSpawnLocation(@Nullable Location location, boolean force);
 
+    // MultiPaper start
+
+    /**
+     * Returns whether the player is on an external server or not.
+     *
+     * @return True if the player is on an external server.
+     */
+    public boolean isExternalPlayer();
+
+    /**
+     * Returns whether the player is on this server or not.
+     *
+     * @return True if the player is on this server.
+     */
+    public boolean isLocalPlayer();
+
+    /**
+     * Returns cross-server data that is stored under the specified key. Note
+     * that all plugins share the same set of keys. This data is
+     * non-persistent, it will be lost when the player disconnects.
+     *
+     * @param key The key the data is stored under.
+     * @return The data stored under the key, or null if the key isn't set.
+     */
+    @Nullable
+    public String getData(String key);
+
+    /**
+     * Store cross-server data under the specified key. Note that all plugins
+     * share the same set of keys. This data is non-persistent, it will be
+     * lost when the player disconnects.
+     *
+     * @param key The key to store the data under.
+     * @param value The data to store under the key.
+     */
+    public void setData(String key, String value);
+
+    /**
+     * Returns cross-server data that is stored under the specified key. Note
+     * that all plugins share the same set of keys. This data is persistent,
+     * it will be saved even if the player disconnects. This persistent data is
+     * saved onto the player's .dat file.
+     *
+     * @param key The key the data is stored under.
+     * @return The data stored under the key, or null if the key isn't set.
+     */
+    @Nullable
+    public String getPersistentData(String key);
+
+    /**
+     * Store cross-server data under the specified key. Note that all plugins
+     * share the same set of keys. This data is persistent, it will be saved
+     * even if the player disconnects. This persistent data is saved onto the
+     * player's .dat file.
+     *
+     * @param key The key to store the data under.
+     * @param value The data to store under the key.
+     */
+    public void setPersistentData(String key, String value);
+
+    /**
+     * Says a message (or runs a command) on other server excluding this one.
+     *
+     * @param msg chat message to say
+     */
+    public void chatOnOtherServers(@NotNull String msg);
+
+    /**
+     * Get the bungeecord name of the server that this player is on.
+     *
+     * @return The bungeecord name of the server the player is on for external
+     *         players, or null for local players.
+     */
+    @Nullable
+    public String getExternalServerName();
+
+    // MultiPaper end
+
     /**
      * Play a note for a player at a location. This requires a note block
      * at the particular location (as far as the client is concerned). This
