From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 1 Jan 2022 19:56:02 +1000
Subject: [PATCH] Mark permissions as dirty when modified


diff --git a/src/main/java/org/bukkit/permissions/PermissibleBase.java b/src/main/java/org/bukkit/permissions/PermissibleBase.java
index cd3296fea01648592d2af89b3d80135acb6d0958..c1cbb006a2f98b6ade35a7aab77a2624462f4090 100644
--- a/src/main/java/org/bukkit/permissions/PermissibleBase.java
+++ b/src/main/java/org/bukkit/permissions/PermissibleBase.java
@@ -20,6 +20,7 @@ public class PermissibleBase implements Permissible {
     private final Permissible parent;
     private final List<PermissionAttachment> attachments = new LinkedList<PermissionAttachment>();
     private final Map<String, PermissionAttachmentInfo> permissions = new HashMap<String, PermissionAttachmentInfo>();
+    public boolean dirty = false; // MultiPaper
 
     public PermissibleBase(@Nullable ServerOperator opable) {
         this.opable = opable;
@@ -176,6 +177,8 @@ public class PermissibleBase implements Permissible {
         for (PermissionAttachment attachment : attachments) {
             calculateChildPermissions(attachment.getPermissions(), false, attachment);
         }
+
+        dirty = true; // MultiPaper
     }
 
     public synchronized void clearPermissions() {
