From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 20 Feb 2022 13:06:41 +1000
Subject: [PATCH] Fix opening a player's inventory


diff --git a/src/main/java/org/bukkit/inventory/InventoryView.java b/src/main/java/org/bukkit/inventory/InventoryView.java
index b06995aa57aa9cba0bb59f1d26d81015619a08e6..a1887329b2eff511a9ccc7398b641379eb0633a8 100644
--- a/src/main/java/org/bukkit/inventory/InventoryView.java
+++ b/src/main/java/org/bukkit/inventory/InventoryView.java
@@ -266,6 +266,12 @@ public abstract class InventoryView {
             return rawSlot;
         }
 
+        // MultiPaper start
+        if (getType() == InventoryType.PLAYER) {
+            numInTop = numInTop / 9 * 9;
+        }
+        // MultiPaper end
+
         // Move down the slot index by the top size
         int slot = rawSlot - numInTop;
 
