From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Sat, 18 Feb 2023 02:12:11 +0400
Subject: [PATCH] add not for entity#getTrackedPlayers


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 65349e06d50e7d3247ae36a36ff3cdf19b2d8029..12a69ecf648e0bdf3417b7ed04671156f2947add 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -929,6 +929,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
 
     /**
      * Returns a set of {@link Player Players} within this entity's tracking range (players that can "see" this entity).
+     * <p>
+     * Please note that this will return Local players only! {@link Player#isLocalPlayer()}
+     * <p>
+     * you can use {@link #getNearbyEntities(double, double, double)} or {@link World#getNearbyPlayers(Location, double)} to get all players including external players
      *
      * @return players in tracking range
      */
