From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Sat, 18 Feb 2023 02:12:11 +0400
Subject: [PATCH] add not for entity#getTrackedPlayers


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index a3b38ebea1f4b085166ac3d875f40e6da6aff2f9..a30006bfedd9581299bf962e4bc3a46d126d1dac 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -912,6 +912,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
 
     /**
      * Returns a set of {@link Player Players} within this entity's tracking range (players that can "see" this entity).
+     * <p>
+     * Please note that this will return Local players only! {@link Player#isLocalPlayer()}
+     * <p>
+     * you can use {@link #getNearbyEntities(double, double, double)} or {@link World#getNearbyPlayers(Location, double)} to get all players including external players
      *
      * @return players in tracking range
      */
