From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Sat, 18 Feb 2023 02:12:11 +0400
Subject: [PATCH] add not for entity#getTrackedPlayers


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index be25290e12e3f73fb090ddde466e74cbab8e227f..f13611cfcf135323f297f65ea8d39e9d0e8a0f7b 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -913,6 +913,10 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
 
     /**
      * Returns a set of {@link Player Players} within this entity's tracking range (players that can "see" this entity).
+     * <p>
+     * Please note that this will return Local players only! {@link Player#isLocalPlayer()}
+     * <p>
+     * you can use {@link #getNearbyEntities(double, double, double)} or {@link World#getNearbyPlayers(Location, double)} to get all players including external players
      *
      * @return players in tracking range
      */
