From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Noel <96398669+noelvnx@users.noreply.github.com>
Date: Mon, 30 May 2022 16:24:38 -0400
Subject: [PATCH] Add FileSyncCompleteEvent


diff --git a/src/main/java/puregero/multipaper/event/FileSyncCompleteEvent.java b/src/main/java/puregero/multipaper/event/FileSyncCompleteEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..91e66b4330a2ad59d0b2566f310ad5b0c2e9c5bd
--- /dev/null
+++ b/src/main/java/puregero/multipaper/event/FileSyncCompleteEvent.java
@@ -0,0 +1,46 @@
+package puregero.multipaper.event;
+
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.server.ServerEvent;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+
+public class FileSyncCompleteEvent extends ServerEvent {
+    public static final HandlerList HANDLER_LIST = new HandlerList();
+
+    private File file;
+    private Cause cause;
+
+    public FileSyncCompleteEvent(@NotNull File file,
+                                 @NotNull Cause cause) {
+        this.file = file;
+        this.cause = cause;
+    }
+
+    @NotNull
+    public File getFile() {
+        return this.file;
+    }
+
+    @NotNull
+    public Cause getCause() {
+        return this.cause;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    public enum Cause {
+        DOWNLOAD,
+        UPLOAD
+    }
+}
