From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 14:03:12 +1000
Subject: [PATCH] Fix decompile issues


diff --git a/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java b/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
index f7c856535cc6a1f010d82f0a0b31a34978854310..c95bde602b8dca8e91a6c2c73411a8e4694ad2af 100644
--- a/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
+++ b/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
@@ -124,8 +124,8 @@ public class ArgumentTypeInfos {
         return register(registry, "uuid", UuidArgument.class, SingletonArgumentInfo.contextFree(UuidArgument::uuid));
     }
 
-    private static <T extends ArgumentType<?>> Class<T> fixClassType(Class<? super T> clazz) {
-        return clazz;
+    private static <T extends ArgumentType<?>> Class<T> fixClassType(Class<? extends T> clazz) { // MultiPaper - fix decompile
+        return (Class<T>) clazz; // MultiPaper - fix decompile
     }
 
     public static boolean isClassRecognized(Class<?> clazz) {
@@ -137,7 +137,7 @@ public class ArgumentTypeInfos {
         if (argumentTypeInfo == null) {
             throw new IllegalArgumentException("Unrecognized argument type %s (%s)".formatted(argumentType, argumentType.getClass()));
         } else {
-            return argumentTypeInfo;
+            return (ArgumentTypeInfo<A, ?>) argumentTypeInfo; // MultiPaper - fix decompile
         }
     }
 
