From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 14:03:12 +1000
Subject: [PATCH] Fix decompile issues


diff --git a/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java b/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
index 4a6b624df834217d7305ba42cab78fdf320ec64c..40550296d2e42050212660801c70e6bf41c3f2c1 100644
--- a/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
+++ b/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
@@ -125,8 +125,8 @@ public class ArgumentTypeInfos {
         return register(registry, "uuid", UuidArgument.class, SingletonArgumentInfo.contextFree(UuidArgument::uuid));
     }
 
-    private static <T extends ArgumentType<?>> Class<T> fixClassType(Class<? super T> clazz) {
-        return clazz;
+    private static <T extends ArgumentType<?>> Class<T> fixClassType(Class<? extends T> clazz) { // MultiPaper - fix decompile
+        return (Class<T>) clazz; // MultiPaper - fix decompile
     }
 
     public static boolean isClassRecognized(Class<?> clazz) {
@@ -138,7 +138,7 @@ public class ArgumentTypeInfos {
         if (argumentTypeInfo == null) {
             throw new IllegalArgumentException(String.format(Locale.ROOT, "Unrecognized argument type %s (%s)", argumentType, argumentType.getClass()));
         } else {
-            return argumentTypeInfo;
+            return (ArgumentTypeInfo<A, ?>) argumentTypeInfo; // MultiPaper - fix decompile
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
index 2bc8917b7c4512f0607f34b1f18627a391fda07d..83c5f1460b8b002574947d5a67d3bfdb96ca0547 100644
--- a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
@@ -118,7 +118,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier {
 
     @Override
     public Brain<Allay> getBrain() {
-        return super.getBrain();
+        return (Brain<Allay>) super.getBrain(); // MultiPaper - Fix decompile error
     }
 
     public static AttributeSupplier.Builder createAttributes() {
@@ -176,8 +176,8 @@ public class Allay extends PathfinderMob implements InventoryCarrier {
     public boolean hurt(DamageSource source, float amount) {
         Entity optional = source.getEntity();
         if (optional instanceof Player player) {
-            Optional<UUID> optional = this.getBrain().getMemory(MemoryModuleType.LIKED_PLAYER);
-            if (optional.isPresent() && player.getUUID().equals(optional.get())) {
+            Optional<UUID> optional2 = this.getBrain().getMemory(MemoryModuleType.LIKED_PLAYER); // MultiPaper - Fix decompile error
+            if (optional2.isPresent() && player.getUUID().equals(optional2.get())) { // MultiPaper - Fix decompile error
                 return false;
             }
         }
