From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 14:03:12 +1000
Subject: [PATCH] Fix decompile issues


diff --git a/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java b/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
index f7c856535cc6a1f010d82f0a0b31a34978854310..c95bde602b8dca8e91a6c2c73411a8e4694ad2af 100644
--- a/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
+++ b/src/main/java/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
@@ -124,8 +124,8 @@ public class ArgumentTypeInfos {
         return register(registry, "uuid", UuidArgument.class, SingletonArgumentInfo.contextFree(UuidArgument::uuid));
     }
 
-    private static <T extends ArgumentType<?>> Class<T> fixClassType(Class<? super T> clazz) {
-        return clazz;
+    private static <T extends ArgumentType<?>> Class<T> fixClassType(Class<? extends T> clazz) { // MultiPaper - fix decompile
+        return (Class<T>) clazz; // MultiPaper - fix decompile
     }
 
     public static boolean isClassRecognized(Class<?> clazz) {
@@ -137,7 +137,7 @@ public class ArgumentTypeInfos {
         if (argumentTypeInfo == null) {
             throw new IllegalArgumentException("Unrecognized argument type %s (%s)".formatted(argumentType, argumentType.getClass()));
         } else {
-            return argumentTypeInfo;
+            return (ArgumentTypeInfo<A, ?>) argumentTypeInfo; // MultiPaper - fix decompile
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
index b6fc1b495e9173e625a7b2c856e79146734c4a9e..75811a1c66ee3249708d11ceef4c5190c6ca2a4b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
@@ -92,7 +92,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationL
 
     @Override
     public Brain<Allay> getBrain() {
-        return super.getBrain();
+        return (Brain<Allay>) super.getBrain(); // MultiPaper - Fix decompile error
     }
 
     public static AttributeSupplier.Builder createAttributes() {
@@ -143,8 +143,8 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationL
     public boolean hurt(DamageSource source, float amount) {
         Entity optional = source.getEntity();
         if (optional instanceof Player player) {
-            Optional<UUID> optional = this.getBrain().getMemory(MemoryModuleType.LIKED_PLAYER);
-            if (optional.isPresent() && player.getUUID().equals(optional.get())) {
+            Optional<UUID> optional2 = this.getBrain().getMemory(MemoryModuleType.LIKED_PLAYER); // MultiPaper - Fix decompile error
+            if (optional2.isPresent() && player.getUUID().equals(optional2.get())) { // MultiPaper - Fix decompile error
                 return false;
             }
         }
