From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 18 Jan 2021 20:05:06 +1000
Subject: [PATCH] Fix banned players, whitelists and ops getting reset when a
 new server is added to the network


diff --git a/src/main/java/net/minecraft/server/JsonList.java b/src/main/java/net/minecraft/server/JsonList.java
index 34857c7d77395b8f768a364af5dcbd8da0b8fa8b..f961eb8b0ccc4b621dd4bd652fcf67de880782d9 100644
--- a/src/main/java/net/minecraft/server/JsonList.java
+++ b/src/main/java/net/minecraft/server/JsonList.java
@@ -182,12 +182,19 @@ public abstract class JsonList<K, V extends JsonListEntry<K>> {
     }
 
     public void load() throws IOException {
-        if (this.c.exists()) {
+        if (true || this.c.exists()) { // MultiPaper
             BufferedReader bufferedreader = null; // Files.newReader(this.c, StandardCharsets.UTF_8); // MultiPaper
             Throwable throwable = null;
-
+            
+            // MultiPaper start
+            String json = MultiPaper.readJson(this.c.getName());
+            if (json == null || json.isEmpty()) {
+                return;
+            }
+            // MultiPaper end
+            
             try {
-                JsonArray jsonarray = JsonList.b.fromJson(MultiPaper.readJson(this.c.getName()), JsonArray.class); // (JsonArray) JsonList.b.fromJson(bufferedreader, JsonArray.class); // MultiPaper
+                JsonArray jsonarray = JsonList.b.fromJson(json, JsonArray.class); // (JsonArray) JsonList.b.fromJson(bufferedreader, JsonArray.class); // MultiPaper
 
                 this.d.clear();
                 Iterator iterator = jsonarray.iterator();
