From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 30 Mar 2021 00:28:39 +1000
Subject: [PATCH] Fix ConcurrentModificationException


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index ccb011b854baed3cc236e2bfe082ff126c800d7c..db7e67d501bd5da7b358b2eaab0ee0c8816b9c00 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -29,6 +29,7 @@ import org.bukkit.entity.Player;
 
 import java.io.*;
 import java.nio.file.Files;
+import java.util.ConcurrentModificationException;
 import java.util.UUID;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
@@ -72,14 +73,18 @@ public class MultiPaper {
     }
 
     private static void checkIfPlayersNeedToBeMoved() {
-        for (Player player : Bukkit.getOnlinePlayers()) {
-            if (player.getLastLogin() < System.currentTimeMillis() - 5 * 1000L) {
-                String server = Zone.getServer(player.getLocation());
+        try {
+            for (Player player : Bukkit.getOnlinePlayers()) {
+                if (player.getLastLogin() < System.currentTimeMillis() - 5 * 1000L) {
+                    String server = Zone.getServer(player.getLocation());
 
-                if (server != null) {
-                    Zone.doMerge(player, server);
+                    if (server != null) {
+                        Zone.doMerge(player, server);
+                    }
                 }
             }
+        } catch (ConcurrentModificationException e) {
+            e.printStackTrace();
         }
     }
 
