From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 21 Jan 2021 20:35:46 +1000
Subject: [PATCH] Fix server restart issues


diff --git a/src/main/java/puregero/multipaper/ExternalServer.java b/src/main/java/puregero/multipaper/ExternalServer.java
index 8943de3e4ee0d3e3d3c91864ebbc4fe58056daa9..3a35379f75384aaf7b31d74ecbd5c801179d980b 100644
--- a/src/main/java/puregero/multipaper/ExternalServer.java
+++ b/src/main/java/puregero/multipaper/ExternalServer.java
@@ -41,7 +41,7 @@ public class ExternalServer {
     }
 
     public boolean isAlive() {
-        return getLastAlive() > System.currentTimeMillis() - 2500;
+        return getLastAlive() > System.currentTimeMillis() - 2500 && getTps() > 0;
     }
 
     public ArrayList<UUID> getPlayers() {
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index a9d648030177d39f232d8785b61c0aafcb6a2f9a..d988a6fd271ff29013c13cffe55297be60ef792e 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -295,7 +295,8 @@ public class MultiPaperConnection extends Thread {
             out.writeUTF("playerList");
             out.writeDouble(20);
             out.writeInt(0);
-        } else {
+            out.send(null);
+        } else if (!MinecraftServer.getServer().isStopped()) {
             Collection<? extends Player> players = Bukkit.getOnlinePlayers();
             out.writeUTF("playerList");
             out.writeDouble(MinecraftServer.getServer().recentTps[0]);
@@ -310,8 +311,8 @@ public class MultiPaperConnection extends Thread {
                 out.writeFloat(player.getLocation().getYaw());
                 out.writeFloat(player.getLocation().getPitch());
             }
+            out.send(null);
         }
-        out.send(null);
 
         sendLoadedChunkList();
     }
diff --git a/src/main/java/puregero/multipaper/PlayerInfoManager.java b/src/main/java/puregero/multipaper/PlayerInfoManager.java
index e976c9f81faa553d37c3f17ea2407162c92d5e3d..4d9763e0f06e6535a5085b71bf2715054c0f6b81 100644
--- a/src/main/java/puregero/multipaper/PlayerInfoManager.java
+++ b/src/main/java/puregero/multipaper/PlayerInfoManager.java
@@ -23,7 +23,7 @@ public class PlayerInfoManager {
 
             if (packet.getAction() == PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER) {
                 playerInfos.put(uuid, info);
-                return;
+                continue;
             }
 
             if (packet.getAction() == PacketPlayOutPlayerInfo.EnumPlayerInfoAction.REMOVE_PLAYER) {
@@ -33,13 +33,13 @@ public class PlayerInfoManager {
                 }
 
                 playerInfos.remove(uuid);
-                return;
+                continue;
             }
 
             PacketPlayOutPlayerInfo.PlayerInfoData infoData = playerInfos.get(uuid);
 
             if (infoData == null) {
-                return;
+                continue;
             }
 
             if (info.b != 0) {
diff --git a/src/main/java/puregero/multipaper/ShutdownHandler.java b/src/main/java/puregero/multipaper/ShutdownHandler.java
index bab74c4a99e7dc4b488c62b204c8441e6369aebb..710ce810ddbfe83c902142b05a0c886383cbdcde 100644
--- a/src/main/java/puregero/multipaper/ShutdownHandler.java
+++ b/src/main/java/puregero/multipaper/ShutdownHandler.java
@@ -1,5 +1,6 @@
 package puregero.multipaper;
 
+import net.minecraft.server.MinecraftServer;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.bukkit.Bukkit;
@@ -14,9 +15,21 @@ public class ShutdownHandler {
     private static final Logger LOGGER = LogManager.getLogger();
 
     public static void onStop() {
+        markServerAsOffline();
+        savePlayers();
         movePlayersToDifferentServer();
     }
 
+    private static void markServerAsOffline() {
+        MultiPaper.sendTickTime(-1);
+    }
+
+    private static void savePlayers() {
+        if (MinecraftServer.getServer().getPlayerList() != null) {
+            MinecraftServer.getServer().getPlayerList().savePlayers();
+        }
+    }
+
     private static void movePlayersToDifferentServer() {
         HashSet<Player> moved = new HashSet<>();
         int tries = 0;
