From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index da7c899fbab162ee197a0593f455ebd9c5286d3c..f6aeb1a5a7811178847f27d93c5c17476ff67258 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -409,7 +409,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public class Spawn extends ConfigurationPart {
         @RequiresSpigotInitialization(SpawnLoadedRangeInitializer.class)
         public short keepSpawnLoadedRange = 10;
-        public boolean keepSpawnLoaded = true;
+        public boolean keepSpawnLoaded = false; // MultiPaper
         public boolean allowUsingSignsInsideSpawnProtection = false;
     }
 
