From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 5 Apr 2021 12:37:14 +1000
Subject: [PATCH] Fix broadcast messages not being broadcast


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 7cd584605bfd5a8756f33797eaacc18e2b519ba8..79ed723e807b7a457958e1641584c5322a8e8382 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -581,6 +581,7 @@ public final class CraftServer implements Server {
     @Override
     @Deprecated // Paper start
     public int broadcastMessage(String message) {
+        MultiPaper.broadcastPacket(null, message);
         this.sendMessage(io.papermc.paper.adventure.PaperAdventure.LEGACY_SECTION_UXRC.deserialize(message));
         return this.getOnlinePlayers().size() + 1;
         // Paper end
