From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 14 Feb 2021 15:03:02 +1000
Subject: [PATCH] Flush sockets after writing


diff --git a/src/main/java/puregero/multipaper/WriterWorkerThread.java b/src/main/java/puregero/multipaper/WriterWorkerThread.java
index a7c3cc5e40db63359e4da69a647b8c1ed427b4b0..310dd44d77fdb51e8feea6b3d80b32ebb5dd8d29 100644
--- a/src/main/java/puregero/multipaper/WriterWorkerThread.java
+++ b/src/main/java/puregero/multipaper/WriterWorkerThread.java
@@ -62,6 +62,7 @@ public class WriterWorkerThread extends Thread {
                 if (toWrite.setLastSocket(socket)) {
                     synchronized (socket) {
                         socket.getOutputStream().write(toWrite.getData());
+                        socket.getOutputStream().flush();
                     }
                 }
             } catch (NullPointerException | SocketException e) {
