From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 5 Apr 2021 12:37:14 +1000
Subject: [PATCH] Fix broadcast messages not being broadcast


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c1ad2e7a682593378794ae66935b8559470bee61..f56c3816485f4db85276989e6922f91b5ba42334 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -610,6 +610,7 @@ public final class CraftServer implements Server {
     @Override
     @Deprecated // Paper start
     public int broadcastMessage(String message) {
+        MultiPaper.broadcastPacket(null, message);
         return this.broadcast(message, BROADCAST_CHANNEL_USERS);
         // Paper end
     }
