From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 15 Feb 2021 14:24:42 +1000
Subject: [PATCH] Don't load raids from disk


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 488cf1a8e75636ac0caf3f993b2b32dd7e688cac..d0f128d0a3f514a9843cd5af40f8eb3478ba6b35 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -262,6 +262,11 @@ public class MultiPaper {
     }
 
     public static byte[] readData(String path) {
+        if (path.contains("raids.dat")) {
+            // Don't load raids from dist
+            return null;
+        }
+        
         try {
             return getConnection().readData(path).join();
         } catch (IOException e) {
