From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 15 Nov 2021 13:02:02 +1000
Subject: [PATCH] Add sendTickTime


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 67f8b5b13f574a7801a2f3e2f98b689e477492ff..5d93832e83b7404a80b74f8a53e6a9707d92849b 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1472,6 +1472,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         tickTimes10s.add(this.tickCount, j);
         tickTimes60s.add(this.tickCount, j);
         // Paper end
+        MultiPaper.sendTickTime(k - i, recentTps[0]); // MultiPaper
         this.frameTimer.logFrameDuration(k - i);
         //this.profiler.pop(); // Purpur
         org.spigotmc.WatchdogThread.tick(); // Spigot
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index d9fa811b2a0b5f8402ae0e811b09ddbd8509e8ad..19948c5a87dd5f0da3aef064dbb6448d306395b6 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -1,5 +1,7 @@
 package puregero.multipaper;
 
+import puregero.multipaper.mastermessagingprotocol.messages.masterbound.WriteTickTimeMessage;
+
 public class MultiPaper {
     private static MultiPaperConnection multiPaperConnection = null;
 
@@ -10,4 +12,8 @@ public class MultiPaper {
 
         return multiPaperConnection;
     }
+
+    public static void sendTickTime(long time, double tps) {
+        getConnection().send(new WriteTickTimeMessage(time, (float) tps));
+    }
 }
