From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 10 Apr 2021 19:20:53 +1000
Subject: [PATCH] Synchronize external server player list


diff --git a/src/main/java/puregero/multipaper/ExternalServer.java b/src/main/java/puregero/multipaper/ExternalServer.java
index 3a35379f75384aaf7b31d74ecbd5c801179d980b..1744af27ca1ce62aa80638b9f0b41a548a6b560d 100644
--- a/src/main/java/puregero/multipaper/ExternalServer.java
+++ b/src/main/java/puregero/multipaper/ExternalServer.java
@@ -1,6 +1,8 @@
 package puregero.multipaper;
 
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.UUID;
 
 public class ExternalServer {
@@ -9,7 +11,7 @@ public class ExternalServer {
     private int averageTickTime;
     private double tps;
     private long lastAlive;
-    private ArrayList<UUID> players = new ArrayList<>();
+    private List<UUID> players = Collections.synchronizedList(new ArrayList<>());
 
     public ExternalServer(String name, boolean me) {
         this.name = name;
@@ -44,7 +46,7 @@ public class ExternalServer {
         return getLastAlive() > System.currentTimeMillis() - 2500 && getTps() > 0;
     }
 
-    public ArrayList<UUID> getPlayers() {
+    public List<UUID> getPlayers() {
         return players;
     }
 
diff --git a/src/main/java/puregero/multipaper/Zone.java b/src/main/java/puregero/multipaper/Zone.java
index 277d3a928db6c239f4e299ccdd93be5cd52219e9..7b2cc4fafd7b21fc3c75c4f78f4d64dd80f21dc2 100644
--- a/src/main/java/puregero/multipaper/Zone.java
+++ b/src/main/java/puregero/multipaper/Zone.java
@@ -59,11 +59,13 @@ public class Zone implements Iterable<Chunk> {
     public static String getServer(Location location) {
         for (ExternalServer server : MultiPaper.getConnection().getServersMap().values()) {
             if (server.isAlive() && !server.isMe()) {
-                for (UUID uuid : server.getPlayers()) {
-                    Location loc = MultiPaper.getConnection().playerLocations.get(uuid);
-                    if (Bukkit.getPlayer(uuid) == null && loc != null && loc.getWorld() == location.getWorld()
-                            && Zone.inSameZone(loc, location)) {
-                        return server.getName();
+                synchronized (server.getPlayers()) {
+                    for (UUID uuid : server.getPlayers()) {
+                        Location loc = MultiPaper.getConnection().playerLocations.get(uuid);
+                        if (Bukkit.getPlayer(uuid) == null && loc != null && loc.getWorld() == location.getWorld()
+                                && Zone.inSameZone(loc, location)) {
+                            return server.getName();
+                        }
                     }
                 }
             }
