From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Jun 2021 00:20:17 +1000
Subject: [PATCH] Add entities region file


diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
index 916f93b097a65f95e830fe5e1567c85d304f808f..98c355ba340946fe1dee16e16c59bbfd3b0a5712 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -103,7 +103,7 @@ public class EntityStorage implements EntityPersistentStorage<Entity> {
             if (this.emptyChunks.add(chunkcoordintpair.toLong())) {
                 this.worker.store(chunkcoordintpair, (CompoundTag) null);
             }
-
+            this.emptyChunks.remove(chunkcoordintpair.toLong()); // MultiPaper - don't cache empty chunks when they're not loaded
         } else {
             ListTag nbttaglist = new ListTag();
 
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/IOWorker.java b/src/main/java/net/minecraft/world/level/chunk/storage/IOWorker.java
index 2340feb8483c2478a22b78e735bcbd6759f1b011..21023743292cd1f9d6eafdee23cbf28339cc4455 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/IOWorker.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/IOWorker.java
@@ -21,6 +21,7 @@ import net.minecraft.util.thread.StrictQueue;
 import net.minecraft.world.level.ChunkPos;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import puregero.multipaper.MultiPaper;
 
 public class IOWorker implements AutoCloseable {
     private static final Logger LOGGER = LogManager.getLogger();
@@ -66,7 +67,7 @@ public class IOWorker implements AutoCloseable {
                 return Either.left(pendingStore.data);
             } else {
                 try {
-                    CompoundTag compoundTag = this.storage.read(pos);
+                    CompoundTag compoundTag = MultiPaper.readRegionFileNBT(storage.getFolder(), pos); // this.storage.read(pos); // MultiPaper
                     return Either.left(compoundTag);
                 } catch (Exception var4) {
                     LOGGER.warn("Failed to read chunk {}", pos, var4);
@@ -129,7 +130,7 @@ public class IOWorker implements AutoCloseable {
 
     private void runStore(ChunkPos pos, IOWorker.PendingStore result) {
         try {
-            this.storage.write(pos, result.data);
+            MultiPaper.writeRegionFileNBT(storage.getFolder(), pos, result.data); // this.storage.write(pos, result.data); // MultiPaper
             result.result.complete((Void)null);
         } catch (Exception var4) {
             LOGGER.error("Failed to store chunk {}", pos, var4);
diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index 4e0006b73fbd2634aa42334ae9dde79b4eccaf38..702aedca187617e3d240dc284af919698d3a34a2 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -37,6 +37,12 @@ public class RegionFileStorage implements AutoCloseable {
         this.folder = directory;
         this.sync = dsync;
     }
+    
+    // MultiPaper start
+    public File getFolder() {
+        return folder;
+    }
+    // MultiPaper end
 
     // Paper start
     public static ChunkPos getRegionFileCoordinates(File file) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 4b3dd64f73cf3887acd9288ecd3a623ac5277751..01ae5bd905de2d2508cdacfdbcc186e2daebf62e 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -127,14 +127,37 @@ public class MultiPaper {
         getConnection().writeChunk(world, path, cx, cz, bytes);
     }
 
+    public static CompoundTag readRegionFileNBT(File path, ChunkPos chunkPos) throws IOException {
+        return readRegionFileNBT(getWorld(path), path.getName(), chunkPos.x, chunkPos.z);
+    }
+
     public static CompoundTag readRegionFileNBT(ServerLevel serverLevel, String path, ChunkPos chunkPos) throws IOException {
-        DataInput in = readRegionFile(serverLevel.convertable.getLevelId(), path, chunkPos.x, chunkPos.z);
+        return readRegionFileNBT(serverLevel.convertable.getLevelId(), path, chunkPos.x, chunkPos.z);
+    }
+
+    public static CompoundTag readRegionFileNBT(String world, String path, int cx, int cz) throws IOException {
+        DataInput in = readRegionFile(world, path, cx, cz);
 
         return in == null ? null : NbtIo.read(in);
     }
 
+    public static void writeRegionFileNBT(File path, ChunkPos chunkPos, CompoundTag compoundTag) throws IOException {
+        writeRegionFileNBT(getWorld(path), path.getName(), chunkPos.x, chunkPos.z, compoundTag);
+    }
+
     public static void writeRegionFileNBT(ServerLevel serverLevel, String path, ChunkPos chunkPos, CompoundTag compoundTag) throws IOException {
-        writeRegionFile(serverLevel.convertable.getLevelId(), path, chunkPos.x, chunkPos.z, nbtToBytes(compoundTag));
+        writeRegionFileNBT(serverLevel.convertable.getLevelId(), path, chunkPos.x, chunkPos.z, compoundTag);
+    }
+
+    public static void writeRegionFileNBT(String world, String path, int cx, int cz, CompoundTag compoundTag) throws IOException {
+        writeRegionFile(world, path, cx, cz, nbtToBytes(compoundTag));
+    }
+
+    private static String getWorld(File path) {
+        do {
+            path = path.getParentFile();
+        } while (path.getName().startsWith("DIM"));
+        return path.getName();
     }
 
     public static CompoundTag readLevel(String world) throws IOException {
@@ -339,6 +362,9 @@ public class MultiPaper {
     }
 
     private static byte[] nbtToBytes(CompoundTag compoundTag) throws IOException {
+        if (compoundTag == null) {
+            return new byte[0];
+        }
         ByteArrayOutputStream buffer = new ByteArrayOutputStream();
         NbtIo.write(compoundTag, (DataOutput) new DataOutputStream(buffer));
         return buffer.toByteArray();
