From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Jun 2021 13:19:46 +1000
Subject: [PATCH] Don't deflate 0 bytes


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index c55bd9c141a6bc7579284c192c3451977b0c0e41..8e8bf54a049285d137d26c0fd5931764951ebdbd 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -387,10 +387,13 @@ public class MultiPaperConnection extends Thread {
     }
 
     public void writeChunk(String world, String path, int cx, int cz, byte[] data) throws IOException {
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        DeflaterOutputStream deflateOut = new DeflaterOutputStream(baos);
-        deflateOut.write(data);
-        deflateOut.close();
+        if (data.length != 0) {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            DeflaterOutputStream deflateOut = new DeflaterOutputStream(baos);
+            deflateOut.write(data);
+            deflateOut.close();
+            data = baos.toByteArray();
+        }
 
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeChunk");
@@ -398,8 +401,8 @@ public class MultiPaperConnection extends Thread {
         out.writeUTF(path);
         out.writeInt(cx);
         out.writeInt(cz);
-        out.writeInt(baos.size());
-        out.write(baos.toByteArray());
+        out.writeInt(data.length);
+        out.write(data);
         out.send(in -> { /* Do nothing */ });
     }
 
