From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Jun 2021 13:19:46 +1000
Subject: [PATCH] Don't deflate 0 bytes


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 39a4666828dee016dab5f94e98143574dd3f44eb..cd1d4fddbd8987d94e071b7d734c0caf78c9cced 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -386,10 +386,13 @@ public class MultiPaperConnection extends Thread {
     }
 
     public void writeChunk(String world, String path, int cx, int cz, byte[] data) throws IOException {
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        DeflaterOutputStream deflateOut = new DeflaterOutputStream(baos);
-        deflateOut.write(data);
-        deflateOut.close();
+        if (data.length != 0) {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            DeflaterOutputStream deflateOut = new DeflaterOutputStream(baos);
+            deflateOut.write(data);
+            deflateOut.close();
+            data = baos.toByteArray();
+        }
 
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeChunk");
@@ -397,8 +400,8 @@ public class MultiPaperConnection extends Thread {
         out.writeUTF(path);
         out.writeInt(cx);
         out.writeInt(cz);
-        out.writeInt(baos.size());
-        out.write(baos.toByteArray());
+        out.writeInt(data.length);
+        out.write(data);
         out.send(in -> { /* Do nothing */ });
     }
 
