From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 27 Feb 2021 12:15:29 +1000
Subject: [PATCH] Fix join message showing when restarting a server


diff --git a/src/main/java/puregero/multipaper/Zone.java b/src/main/java/puregero/multipaper/Zone.java
index 677612f2b5642d4122a4e378ced44735671ccd7b..8566dfa3620f67170d285d91abbce3df95d4e55c 100644
--- a/src/main/java/puregero/multipaper/Zone.java
+++ b/src/main/java/puregero/multipaper/Zone.java
@@ -167,6 +167,7 @@ public class Zone {
 
         for (Player player : toSend) {
             ((CraftPlayer) player).getHandle().isBeingSentToAnotherServer = true;
+            player.saveData();
             BungeeCord.connectPlayer(player, server);
             
             // Check back in 15 seconds to see if they've moved servers yet
@@ -185,6 +186,7 @@ public class Zone {
     private static void ensureMovedServer(Player player) {
         // Remove the player from beingSentToAnotherServer will put them back on the server
         if (player.isOnline() && beingSentToAnotherServer.remove(player)) {
+            ((CraftPlayer) player).getHandle().isBeingSentToAnotherServer = false;
             zonesToUnload.remove(player);
             
             LOGGER.info(player.getName() + " failed to move servers, checking if they still need to");
