From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 22 Jun 2021 15:16:36 +1000
Subject: [PATCH] Disable cancelling chunk load functionality


diff --git a/src/main/java/com/destroystokyo/paper/io/chunk/ChunkTaskManager.java b/src/main/java/com/destroystokyo/paper/io/chunk/ChunkTaskManager.java
index 3b7ba9c755c82a6f086d5542d32b3567c0f98b99..acb758ef30302a052322b4f61fb2201c57b4bc8f 100644
--- a/src/main/java/com/destroystokyo/paper/io/chunk/ChunkTaskManager.java
+++ b/src/main/java/com/destroystokyo/paper/io/chunk/ChunkTaskManager.java
@@ -301,6 +301,8 @@ public final class ChunkTaskManager {
     }
 
     public void cancelChunkLoad(final int chunkX, final int chunkZ) {
+        if (true) return; // Disable cancel chunk load
+        
         this.chunkLoadTasks.compute(IOUtil.getCoordinateKey(chunkX, chunkZ), (final Long keyInMap, final ChunkLoadTask valueInMap) -> {
             if (valueInMap == null) {
                 return null;
