From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 9 Mar 2021 23:35:09 +1000
Subject: [PATCH] Put item on player's cursor back into their inventory when
 going between servers'


diff --git a/src/main/java/puregero/multipaper/Zone.java b/src/main/java/puregero/multipaper/Zone.java
index 8566dfa3620f67170d285d91abbce3df95d4e55c..d1eb086dcad9fb21da5eab429b02bd08ea77fcef 100644
--- a/src/main/java/puregero/multipaper/Zone.java
+++ b/src/main/java/puregero/multipaper/Zone.java
@@ -167,6 +167,10 @@ public class Zone {
 
         for (Player player : toSend) {
             ((CraftPlayer) player).getHandle().isBeingSentToAnotherServer = true;
+            if (!player.getItemOnCursor().getType().isAir()) {
+                player.getInventory().addItem(player.getItemOnCursor());
+                player.setItemOnCursor(null);
+            }
             player.saveData();
             BungeeCord.connectPlayer(player, server);
             
