From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 6 Jul 2021 18:14:56 +1000
Subject: [PATCH] Improve server stability


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 94078a53820c84c92a59b720f7eee538f59c645f..a7f43e08e48fddb59b77f1724b350a1d42f9acb6 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1050,6 +1050,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         MinecraftServer.LOGGER.info("Stopping server");
         MinecraftTimings.stopServer(); // Paper
         ShutdownHandler.onStop(); // MultiPaper
+        MultiPaper.onStop();
         // CraftBukkit start
         if (this.server != null) {
             this.server.disablePlugins();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 1a1b62cfddb7322a21822b221399b3ac0096c9a8..fbaa6c79468638a7b0fd9d3f677ba9f0fa50223e 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -168,11 +168,7 @@ public class MultiPaper {
     private static final ConcurrentHashMap<String, CompletableFuture<byte[]>> regionFileLookAhead = new ConcurrentHashMap<>();
     private static final List<String> lookAheadPaths = Arrays.asList("region", "poi", "entities");
     private static long lastRegionFileLookAheadClear = 0;
-    public static DataInput readRegionFile(String world, String path, int cx, int cz) throws IOException {      
-        if (DedicatedServer.getServer().hasStopped()) {
-            LOGGER.info("Attempted to readRegionFile " + world + "," + path + "," + cx + "," + cz + ", but server has stopped, aborting operation.");
-            return null;
-        }
+    public static DataInput readRegionFile(String world, String path, int cx, int cz) throws IOException {     
         while (true) {
             try {
                 CompletableFuture<byte[]> future;
@@ -201,10 +197,10 @@ public class MultiPaper {
             } catch (InterruptedException | ExecutionException e) {
                 throw new IOException(e);
             } catch (TimeoutException e) {
-                if (DedicatedServer.getServer().hasStopped()) {
-                    LOGGER.info("No response for readRegionFile " + world + "," + path + "," + cx + "," + cz + ", but server has stopped, aborting operation.");
-                    return null;
-                }
+//                if (DedicatedServer.getServer().hasStopped()) {
+//                    LOGGER.info("No response for readRegionFile " + world + "," + path + "," + cx + "," + cz + ", but server has stopped, aborting operation.");
+//                    return null;
+//                }
                 LOGGER.info("No response for readRegionFile " + world + "," + path + "," + cx + "," + cz + ", resending...");
             }
         }
@@ -535,4 +531,35 @@ public class MultiPaper {
     public static void setPlayerShouldTickEntities(Player player, boolean shouldTickEntities) {
         ((CraftPlayer) player).getHandle().shouldTickEntities = shouldTickEntities;
     }
+
+    private static Thread stoppingThread;
+    public static void onStop() {
+        stoppingThread = Thread.currentThread();
+        new Thread(() -> {
+            int i = 0;
+            long t = System.currentTimeMillis();
+            while (stoppingThread.isAlive()) {
+                int j = (int) ((System.currentTimeMillis() - t) / 10000);
+                if (i != j) {
+                    i = j;
+                    LOGGER.fatal("Server has taken too long to restart (" + (System.currentTimeMillis() - t) / 1000 + " seconds)");
+                    System.err.println("Stack:");
+                    for (StackTraceElement e : stoppingThread.getStackTrace()) {
+                        System.err.println("\t" + e);
+                    }
+                }
+                if (System.currentTimeMillis() > t + 60000) {
+                    LOGGER.info("Server has stopped unsuccessfully");
+                    return;
+                }
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                    return;
+                }
+            }
+            LOGGER.info("Server stopped successfully");
+        }).start();
+    }
 }
diff --git a/src/main/java/puregero/multipaper/Zone.java b/src/main/java/puregero/multipaper/Zone.java
index fbaf86b7492afe171f5c11fe37f0dd3cc2f1b8e9..82a085d2aa10f714853643813ac6eeb981bf9696 100644
--- a/src/main/java/puregero/multipaper/Zone.java
+++ b/src/main/java/puregero/multipaper/Zone.java
@@ -180,7 +180,9 @@ public class Zone implements Iterable<Chunk> {
 
     private void mapPlayers(World world) {
         for (Player player : world.getPlayers()) {
-            if (chunks.contains(player.getChunk())) {
+            int cx = player.getLocation().getBlockX() >> 4;
+            int cz = player.getLocation().getBlockZ() >> 4;
+            if (world.isChunkLoaded(cx, cz) && chunks.contains(player.getChunk())) {
                 players.add(player);
             }
         }
