From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 24 May 2024 16:44:47 +0900
Subject: [PATCH] World gen


diff --git a/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java b/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java
index 5bfdd772f2e8933acca69ed62c04755a5655fa94..d1ab812685ecef25ace31eca34b74c4c3dec8ed5 100644
--- a/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java
+++ b/src/main/java/net/minecraft/server/level/PlayerRespawnLogic.java
@@ -10,6 +10,7 @@ import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.levelgen.Heightmap;
+import puregero.multipaper.region.RegionPos;
 
 public class PlayerRespawnLogic {
     @Nullable
@@ -48,6 +49,7 @@ public class PlayerRespawnLogic {
         if (SharedConstants.debugVoidTerrain(chunkPos)) {
             return null;
         } else {
+            return world.chunkScheduler.getRegionLocker().lockRegion(RegionPos.forChunk(chunkPos), () -> { // MultiPaper - lock region
             for (int i = chunkPos.getMinBlockX(); i <= chunkPos.getMaxBlockX(); i++) {
                 for (int j = chunkPos.getMinBlockZ(); j <= chunkPos.getMaxBlockZ(); j++) {
                     BlockPos blockPos = getOverworldRespawnPos(world, i, j);
@@ -58,6 +60,7 @@ public class PlayerRespawnLogic {
             }
 
             return null;
+            }); // MultiPaper
         }
     }
 }
