From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 May 2024 02:33:24 +0900
Subject: [PATCH] Multithreaded WeakSeqLock


diff --git a/src/main/java/com/destroystokyo/paper/util/concurrent/WeakSeqLock.java b/src/main/java/com/destroystokyo/paper/util/concurrent/WeakSeqLock.java
index 4029dc68cf35d63aa70c4a76c35bf65a7fc6358f..2889d03e082427436e348e6e39d6ba8fdb19911b 100644
--- a/src/main/java/com/destroystokyo/paper/util/concurrent/WeakSeqLock.java
+++ b/src/main/java/com/destroystokyo/paper/util/concurrent/WeakSeqLock.java
@@ -16,8 +16,21 @@ public final class WeakSeqLock {
     }
 
     public void acquireWrite() {
+        // MultiPaper start - Only one thread can hold the write lock at a time
+        int failures = 0;
+        long curr;
+
+        for (curr = this.lock.get(); !this.canRead(curr) || !lock.compareAndSet(curr, curr + 1); curr = this.lock.get()) {
+
+            if (++failures > 5_000) {
+                Thread.yield();
+            }
+
+        }
+
         // must be release-type write
-        this.lock.lazySet(this.lock.get() + 1);
+        // this.lock.lazySet(this.lock.get() + 1);
+        // MultiPaper end - Only one thread can hold the write lock at a time
     }
 
     public boolean canRead(final long read) {
