From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 10 Jul 2022 21:13:49 +1000
Subject: [PATCH] Acknowledge block changes for external players


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index b4043c355bc1d8ca7c0da477338d2ed02a15d065..5a68bc74f8a5280c9a7fd27d2c6114c2a031ef7a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -249,6 +249,7 @@ import org.bukkit.inventory.InventoryView;
 import org.bukkit.inventory.SmithingInventory;
 import puregero.multipaper.ExternalPlayer;
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperAckBlockChangesHandler;
 import puregero.multipaper.externalserverprotocol.PlayerRemovePacket;
 // CraftBukkit end
 
@@ -264,7 +265,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     private final MinecraftServer server;
     public ServerPlayer player;
     private int tickCount;
-    private int ackBlockChangesUpTo = -1;
+    public int ackBlockChangesUpTo = -1; // MultiPaper - make public
     private long keepAliveTime = Util.getMillis();
     private boolean keepAlivePending;
     private long keepAliveChallenge;
@@ -2293,6 +2294,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             throw new IllegalArgumentException("Expected packet sequence nr >= 0");
         } else {
             this.ackBlockChangesUpTo = Math.max(sequence, this.ackBlockChangesUpTo);
+            MultiPaperAckBlockChangesHandler.onAckNeeded(this.player); // MultiPaper
         }
     }
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 5e1015113ba56b751e7e11e527ef04fead587b5b..7eb012e15e8092fb9c58c6db09584764b7a4c22f 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -120,6 +120,8 @@ public class MultiPaper {
                 // Ignore - plugins doing async things (HuskSync specifically)
             }
         });
+
+        MultiPaperAckBlockChangesHandler.tick();
     }
 
     public static void sendTickTime(long time, double tps) {
diff --git a/src/main/java/puregero/multipaper/MultiPaperAckBlockChangesHandler.java b/src/main/java/puregero/multipaper/MultiPaperAckBlockChangesHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..c71c4e08696dc749f47b477df43d74d3d148c26b
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperAckBlockChangesHandler.java
@@ -0,0 +1,29 @@
+package puregero.multipaper;
+
+import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
+import net.minecraft.network.protocol.game.ClientboundBlockChangedAckPacket;
+import net.minecraft.server.level.ServerPlayer;
+
+import java.util.Set;
+
+public class MultiPaperAckBlockChangesHandler {
+    private final static Set<ExternalPlayer> ackBlockChangesQueue = new ObjectLinkedOpenHashSet<>(); // Linked set for faster iteration
+
+    public static void tick() {
+        if (!ackBlockChangesQueue.isEmpty()) {
+            for (ExternalPlayer player : ackBlockChangesQueue) {
+                if (player.connection.ackBlockChangesUpTo > -1) {
+                    player.connection.send(new ClientboundBlockChangedAckPacket(player.connection.ackBlockChangesUpTo));
+                    player.connection.ackBlockChangesUpTo = -1;
+                }
+            }
+            ackBlockChangesQueue.clear();
+        }
+    }
+
+    public static void onAckNeeded(ServerPlayer player) {
+        if (player instanceof ExternalPlayer externalPlayer) {
+            ackBlockChangesQueue.add(externalPlayer);
+        }
+    }
+}
