From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 4 Dec 2021 12:23:17 +1000
Subject: [PATCH] Sync enderchests


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d38fa61c8264442a0712227d20281fbe9396a248..be331455a878093fb12a704dd7d5870706ae9785 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -433,6 +433,7 @@ public class ServerPlayer extends Player {
         this.maxHealthCache = this.getMaxHealth();
         this.cachedSingleMobDistanceMap = new com.destroystokyo.paper.util.PooledHashSets.PooledObjectLinkedOpenHashSet<>(this); // Paper
         this.spawnInvulnerableTime = world.purpurConfig.playerSpawnInvulnerableTicks; // Purpur
+        enderChestInventory.addListener(new MultiPaperEnderChestHandler(this)); // MultiPaper
     }
 
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
diff --git a/src/main/java/puregero/multipaper/MultiPaperEnderChestHandler.java b/src/main/java/puregero/multipaper/MultiPaperEnderChestHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..d8074c83ab7d10c14e59f86402b724877dedfb67
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperEnderChestHandler.java
@@ -0,0 +1,89 @@
+package puregero.multipaper;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.Container;
+import net.minecraft.world.ContainerListener;
+import net.minecraft.world.item.ItemStack;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.scheduler.CraftScheduler;
+import org.bukkit.scheduler.BukkitTask;
+import puregero.multipaper.externalserverprotocol.PlayerInventoryUpdatePacket;
+
+public class MultiPaperEnderChestHandler implements ContainerListener {
+
+    private static boolean broadcastChanges = true;
+
+    private final ServerPlayer player;
+    private ItemStack[] sentItems = new ItemStack[0];
+    private BukkitTask isScheduled = null;
+
+    public MultiPaperEnderChestHandler(ServerPlayer player) {
+        this.player = player;
+    }
+
+    @Override
+    public void containerChanged(Container container) {
+        if (container.getContainerSize() != sentItems.length) {
+            sentItems = new ItemStack[container.getContainerSize()];
+        }
+
+        if (!player.didPlayerJoinEvent) {
+            if (isScheduled == null && broadcastChanges && player.server.getPlayerList().getActivePlayer(player.getUUID()) == player) {
+                // Wait till they join to broadcast changes
+                isScheduled = ((CraftScheduler) Bukkit.getScheduler()).scheduleInternalTask(() -> {
+                    isScheduled = null;
+                    containerChanged(container);
+                }, 1, "MultiPaperEnderChestHandler-containerChanged");
+            }
+            return;
+        }
+
+        isScheduled = null;
+
+        CompoundTag itemsRoot = new CompoundTag();
+        ListTag items = new ListTag();
+        for (int i = 0; i < sentItems.length; i++) {
+            ItemStack item = container.getItem(i);
+            if (!item.equals(sentItems[i])) {
+                sentItems[i] = item.copy();
+
+                if (broadcastChanges) {
+                    CompoundTag itemToSend = new CompoundTag();
+                    itemToSend.putInt("Slot", i);
+                    item.save(itemToSend);
+                    items.add(itemToSend);
+                }
+            }
+        }
+
+        if (!items.isEmpty()) {
+            itemsRoot.put("items", items);
+            MultiPaper.broadcastPacketToExternalServers(new PlayerInventoryUpdatePacket(player, "enderchest", itemsRoot));
+        }
+    }
+
+    public static void sendFullEnderChestUpdate(ServerPlayer player, ExternalServerConnection... connections) {
+        CompoundTag itemsRoot = new CompoundTag();
+        ListTag items = new ListTag();
+        for (int i = 0; i < player.getEnderChestInventory().getContainerSize(); i++) {
+            ItemStack item = player.getEnderChestInventory().getItem(i);
+            CompoundTag itemToSend = new CompoundTag();
+            itemToSend.putInt("Slot", i);
+            item.save(itemToSend);
+            items.add(itemToSend);
+        }
+
+        itemsRoot.put("items", items);
+        for (ExternalServerConnection connection : connections) {
+            connection.send(new PlayerInventoryUpdatePacket(player, "enderchest", itemsRoot));
+        }
+    }
+
+    public static void updateInventory(ServerPlayer player, int slot, ItemStack item) {
+        broadcastChanges = false;
+        player.getEnderChestInventory().setItem(slot, item);
+        broadcastChanges = true;
+    }
+}
diff --git a/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java b/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java
index 8d083ab12060016a7ca2ea5d693957be3320f02b..fbf0bd20980073e71d67029efe596f7618d78106 100644
--- a/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java
@@ -130,6 +130,7 @@ public class MultiPaperInventoryHandler {
             case "items" -> component = player.getInventory().items;
             case "armor" -> component = player.getInventory().armor;
             case "offhand" -> component = player.getInventory().offhand;
+            case "enderchest" -> MultiPaperEnderChestHandler.updateInventory(player, slot, item);
             default -> throw new IllegalArgumentException("Unknown inventory component of " + name);
         }
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java
index bcb17ae5b57131f6627aeecde4cec91c23bef215..3d2e875ae6da723785fe3f97842e3c8f04f0525f 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java
@@ -9,10 +9,7 @@ import net.minecraft.world.level.GameType;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.bukkit.event.player.PlayerKickEvent;
-import puregero.multipaper.ExternalPlayer;
-import puregero.multipaper.ExternalServerConnection;
-import puregero.multipaper.MultiPaper;
-import puregero.multipaper.MultiPaperInventoryHandler;
+import puregero.multipaper.*;
 
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
@@ -120,5 +117,6 @@ public class PlayerCreatePacket extends ExternalServerPacket {
         send(new PlayerActionPacket(player, new ServerboundSetCarriedItemPacket(player.getInventory().selected)), connections);
 
         MultiPaperInventoryHandler.sendFullInventoryUpdate(player, connections);
+        MultiPaperEnderChestHandler.sendFullEnderChestUpdate(player, connections);
     }
 }
