From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 16:34:05 +1000
Subject: [PATCH] Only let one server tick the dragon fight


diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
index 5b333bef255d7ef61c99510837536920c6fb6e8c..cf4f12c7f4bfbd24530bb49ff27386aee5bd93a1 100644
--- a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -60,6 +60,7 @@ import net.minecraft.world.level.levelgen.feature.SpikeFeature;
 import net.minecraft.world.level.levelgen.feature.configurations.FeatureConfiguration;
 import net.minecraft.world.phys.AABB;
 import org.slf4j.Logger;
+import puregero.multipaper.MultiPaper;
 
 public class EndDragonFight {
 
@@ -227,8 +228,10 @@ public class EndDragonFight {
         List<? extends EnderDragon> list = this.level.getDragons();
 
         if (list.isEmpty()) {
+            if (MultiPaper.isChunkLocal(this.level.getChunkIfLoaded(0, 0))) { // MultiPaper - only respawn the dragon if we're ticking it
             EndDragonFight.LOGGER.debug("Haven't seen the dragon, respawning it");
             this.createNewDragon();
+            } // MultiPaper - only respawn the dragon if we're ticking it
         } else {
             EndDragonFight.LOGGER.debug("Haven't seen our dragon, but found another one to use.");
             this.dragonUUID = ((EnderDragon) list.get(0)).getUUID();
