From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Sync attack strength


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 816ef12ee70088545bc1f8cf18b0eae07af74ce7..e1ac28aedb927231acedbe86549c696d659f3603 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -127,6 +127,7 @@ import org.bukkit.event.entity.EntityExhaustionEvent;
 import org.bukkit.event.player.PlayerDropItemEvent;
 import org.bukkit.event.player.PlayerVelocityEvent;
 import puregero.multipaper.MultiPaperEntityInteractHandler;
+import puregero.multipaper.externalserverprotocol.PlayerResetAttackStrengthPacket;
 // CraftBukkit end
 
 public abstract class Player extends LivingEntity {
@@ -335,7 +336,18 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
+
         ItemStack itemstack = this.getMainHandItem();
 
         if (!ItemStack.matches(this.lastItemInMainHand, itemstack)) {
@@ -2317,6 +2329,7 @@ public abstract class Player extends LivingEntity {
     }
 
     public void resetAttackStrengthTicker() {
+        PlayerResetAttackStrengthPacket.broadcastResetAttackStrength(this);
         this.attackStrengthTicker = 0;
     }
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 0c9a1c314b4f7ab94e2f01d042e8c9e1ac5b655a..75b67bc4fbdc995467a4b7da52d13d152ab635cf 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -82,6 +82,8 @@ public class MultiPaper {
                 // Copied from Paper end
 
                 externalPlayer.checkInsideBlocks();
+
+                externalPlayer.tickAttackStrength();
             }
         }
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index a517132003d16117c0ce4031ad5bb0ddd84e901a..5e57b1977c93d3b94deaf382c32f4621408ddf3b 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -46,6 +46,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerTouchEntityPacket.class, PlayerTouchEntityPacket::new);
         addPacket(SetEndDragonFightDataPacket.class, SetEndDragonFightDataPacket::new);
         addPacket(HurtEntityPacket.class, HurtEntityPacket::new);
+        addPacket(PlayerResetAttackStrengthPacket.class, PlayerResetAttackStrengthPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerResetAttackStrengthPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerResetAttackStrengthPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..048e25da96045e81871b5024b379d8b10f80efe7
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerResetAttackStrengthPacket.java
@@ -0,0 +1,55 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.player.Player;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerResetAttackStrengthPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(PlayerResetAttackStrengthPacket.class.getSimpleName());
+    private static boolean settingAttackStrength = false;
+
+    private final UUID uuid;
+
+    public PlayerResetAttackStrengthPacket(Player player) {
+        this.uuid = player.getUUID();
+    }
+
+    public PlayerResetAttackStrengthPacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            if (player == null) {
+                LOGGER.warn("Tried to reset attack strength of a non-existent player with uuid " + uuid);
+                return;
+            }
+
+            settingAttackStrength = true;
+            player.resetAttackStrengthTicker();
+            settingAttackStrength = false;
+        });
+    }
+
+    public static void broadcastResetAttackStrength(Player player) {
+        if (!settingAttackStrength && player instanceof ServerPlayer serverPlayer) {
+            MultiPaper.broadcastPacketToExternalServers(serverPlayer, new PlayerResetAttackStrengthPacket(player));
+        }
+    }
+}
