From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index b64fbce003c00a6371a849d4835632c18ad730bb..e08c3e61d096e52d6c31d6702c81f395a3b40264 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -318,7 +318,7 @@ public class PaperWorldConfig {
 
     public boolean keepSpawnInMemory;
     private void keepSpawnInMemory() {
-        keepSpawnInMemory = getBoolean("keep-spawn-loaded", true);
+        keepSpawnInMemory = false; // getBoolean("keep-spawn-loaded", true); // MultiPaper
         log("Keep spawn chunk loaded: " + keepSpawnInMemory);
     }
 
