From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 13 Dec 2021 21:27:25 +1000
Subject: [PATCH] Sync gamerules


diff --git a/src/main/java/net/minecraft/world/level/GameRules.java b/src/main/java/net/minecraft/world/level/GameRules.java
index 6fa00d9239546aa82b9e92e25a33103868084c38..029cb0849d3c4e9178082183c3f3f384a3825408 100644
--- a/src/main/java/net/minecraft/world/level/GameRules.java
+++ b/src/main/java/net/minecraft/world/level/GameRules.java
@@ -27,6 +27,7 @@ import net.minecraft.network.protocol.game.ClientboundGameEventPacket;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerPlayer;
 import org.slf4j.Logger;
+import puregero.multipaper.externalserverprotocol.GameRuleUpdatePacket;
 
 public class GameRules {
 
@@ -353,6 +354,7 @@ public class GameRules {
         protected void updateFromArgument(CommandContext<CommandSourceStack> context, String name, GameRules.Key<BooleanValue> gameRuleKey) { // Paper start
             io.papermc.paper.event.world.WorldGameRuleChangeEvent event = new io.papermc.paper.event.world.WorldGameRuleChangeEvent(context.getSource().getBukkitWorld(), context.getSource().getBukkitSender(), (org.bukkit.GameRule<Boolean>) org.bukkit.GameRule.getByName(gameRuleKey.toString()), String.valueOf(BoolArgumentType.getBool(context, name)));
             if (!event.callEvent()) return;
+           GameRuleUpdatePacket.onGameRuleChange(event.getWorld(), event.getGameRule().getName(), event.getValue()); // MultiPaper
             this.value = Boolean.parseBoolean(event.getValue());
             // Paper end
         }
@@ -421,6 +423,7 @@ public class GameRules {
         protected void updateFromArgument(CommandContext<CommandSourceStack> context, String name, GameRules.Key<IntegerValue> gameRuleKey) { // Paper start
             io.papermc.paper.event.world.WorldGameRuleChangeEvent event = new io.papermc.paper.event.world.WorldGameRuleChangeEvent(context.getSource().getBukkitWorld(), context.getSource().getBukkitSender(), (org.bukkit.GameRule<Integer>) org.bukkit.GameRule.getByName(gameRuleKey.toString()), String.valueOf(IntegerArgumentType.getInteger(context, name)));
             if (!event.callEvent()) return;
+            GameRuleUpdatePacket.onGameRuleChange(event.getWorld(), event.getGameRule().getName(), event.getValue()); // MultiPaper
             this.value = Integer.parseInt(event.getValue());
             // Paper end
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index fc9f3dca2ae3a773bf772e2e074a0863d16acb79..90ded953128d7e5e1433f3e39b9977bc1f21fbd7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -136,6 +136,7 @@ import org.bukkit.util.StructureSearchResult;
 import org.bukkit.util.Vector;
 import org.jetbrains.annotations.NotNull;
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.GameRuleUpdatePacket;
 import puregero.multipaper.externalserverprotocol.TimeUpdatePacket;
 
 public class CraftWorld extends CraftRegionAccessor implements World {
@@ -1958,6 +1959,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         GameRule<?> gameRule = GameRule.getByName(rule);
         io.papermc.paper.event.world.WorldGameRuleChangeEvent event = new io.papermc.paper.event.world.WorldGameRuleChangeEvent(this, null, gameRule, value);
         if (!event.callEvent()) return false;
+        GameRuleUpdatePacket.onGameRuleChange(event.getWorld(), event.getGameRule().getName(), event.getValue()); // MultiPaper
         // Paper end
         GameRules.Value<?> handle = this.getHandle().getGameRules().getRule(CraftWorld.getGameRulesNMS().get(rule));
         handle.deserialize(event.getValue()); // Paper
@@ -1998,6 +2000,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         // Paper start
         io.papermc.paper.event.world.WorldGameRuleChangeEvent event = new io.papermc.paper.event.world.WorldGameRuleChangeEvent(this, null, rule, String.valueOf(newValue));
         if (!event.callEvent()) return false;
+        GameRuleUpdatePacket.onGameRuleChange(event.getWorld(), event.getGameRule().getName(), event.getValue()); // MultiPaper
         // Paper end
         GameRules.Value<?> handle = this.getHandle().getGameRules().getRule(CraftWorld.getGameRulesNMS().get(rule.getName()));
         handle.deserialize(event.getValue()); // Paper
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 7a23ecddb8b079fc29cae088092dfb9f9b01fb65..49d8998c88ac5550e0d7ff42c3abef6f88fd4fc3 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -55,6 +55,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerSayChatPacket.class, PlayerSayChatPacket::new);
         addPacket(PlayerFoodUpdatePacket.class, PlayerFoodUpdatePacket::new);
         addPacket(EntityUpdateEffectPacket.class, EntityUpdateEffectPacket::new);
+        addPacket(GameRuleUpdatePacket.class, GameRuleUpdatePacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/GameRuleUpdatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/GameRuleUpdatePacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..dbe6646c697e7cae7e22ae78740aab39aa73aac2
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/GameRuleUpdatePacket.java
@@ -0,0 +1,53 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+public class GameRuleUpdatePacket extends ExternalServerPacket {
+
+    private static boolean updatingGamerules = false;
+
+    private final String world;
+    private final String name;
+    private final String value;
+
+    public GameRuleUpdatePacket(World world, String name, String value) {
+        this.world = world.getName();
+        this.name = name;
+        this.value = value;
+    }
+
+    public GameRuleUpdatePacket(FriendlyByteBuf in) {
+        world = in.readUtf();
+        name = in.readUtf();
+        value = in.readUtf();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+        out.writeUtf(name);
+        out.writeUtf(value);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            World bukkitWorld = Bukkit.getWorld(world);
+            if (bukkitWorld != null) {
+                updatingGamerules = true;
+                bukkitWorld.setGameRuleValue(name, value);
+                updatingGamerules = false;
+            }
+        });
+    }
+
+    public static void onGameRuleChange(World world, String name, String value) {
+        if (!updatingGamerules) {
+            MultiPaper.broadcastPacketToExternalServers(new GameRuleUpdatePacket(world, name, value));
+        }
+    }
+}
