From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 6b7006829dde81840bd305ee775d26c9f1ef24f7..a31858f456ea59c186ea379e466f52d06ac55435 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -288,7 +288,7 @@ public class PaperWorldConfig {
 
     public boolean keepSpawnInMemory;
     private void keepSpawnInMemory() {
-        keepSpawnInMemory = getBoolean("keep-spawn-loaded", true);
+        keepSpawnInMemory = false; // getBoolean("keep-spawn-loaded", true); // MultiPaper
         log("Keep spawn chunk loaded: " + keepSpawnInMemory);
     }
 
