From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 13 Dec 2021 22:55:54 +1000
Subject: [PATCH] Forward sign editing


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 9a2af64e125ec1ef6c012101db4723677744cd32..e1e236f41186b66403ec96bc862a5f071b2cc1d3 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -247,6 +247,7 @@ import puregero.multipaper.MultiPaperAckBlockChangesHandler;
 import puregero.multipaper.MultiPaperEntityInteractHandler;
 import puregero.multipaper.MultiPaperInventoryHandler;
 import puregero.multipaper.MultiPaperPlayerHandler;
+import puregero.multipaper.MultiPaperSignHandler;
 import puregero.multipaper.externalserverprotocol.PlayerActionPacket;
 import puregero.multipaper.externalserverprotocol.PlayerRemovePacket;
 // CraftBukkit end
@@ -3502,6 +3503,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
     @Override
     public void handleSignUpdate(ServerboundSignUpdatePacket packet) {
+        if (MultiPaperSignHandler.handleSignUpdate(this.player, packet)) return;
         // Paper start - cap line length - modified clients can send longer data than normal
         String[] lines = packet.getLines();
         for (int i = 0; i < lines.length; ++i) {
diff --git a/src/main/java/puregero/multipaper/MultiPaperSignHandler.java b/src/main/java/puregero/multipaper/MultiPaperSignHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..08d802a86b85603465dd939ef735756404297298
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperSignHandler.java
@@ -0,0 +1,23 @@
+package puregero.multipaper;
+
+import io.papermc.paper.chunk.system.scheduling.NewChunkHolder;
+import net.minecraft.network.protocol.game.ServerboundSignUpdatePacket;
+import net.minecraft.server.level.ServerPlayer;
+import puregero.multipaper.externalserverprotocol.PlayerActionPacket;
+
+public class MultiPaperSignHandler {
+
+    /**
+     * Returns true if the sign update should be cancelled
+     */
+    public static boolean handleSignUpdate(ServerPlayer player, ServerboundSignUpdatePacket packet) {
+        NewChunkHolder newChunkHolder = MultiPaper.getChunkHolder(player);
+        if (MultiPaper.isChunkExternal(newChunkHolder)) {
+            newChunkHolder.externalOwner.getConnection().send(new PlayerActionPacket(player, packet));
+            return true;
+        }
+
+        return false;
+    }
+
+}
