From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 21:32:25 +1000
Subject: [PATCH] Subscribe to worlds


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index d1684dfd788567e0fa0afbcdfed69074d37373a8..f576eb3dafd132a9b1cb106f750269f3226eb621 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -249,6 +249,8 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.SubscribeToWorldPacket;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -1377,6 +1379,7 @@ public final class CraftServer implements Server {
             return;
         }
         this.worlds.put(world.getName().toLowerCase(java.util.Locale.ENGLISH), world);
+        MultiPaper.broadcastPacketToExternalServers(new SubscribeToWorldPacket(world.getName())); // MultiPaper
     }
 
     @Override
diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index 95165501e33fa87656b5cf71cbcb68d893cfd694..9f40f5587c376d8e3fa881824a988ec060e6b8d2 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -7,12 +7,14 @@ import java.io.*;
 import java.net.Socket;
 import java.nio.ByteBuffer;
 import java.util.concurrent.*;
+import java.util.HashSet;
 
 public class ExternalServerConnection extends Thread implements Closeable {
 
     private final Socket socket;
     public ExternalServer externalServer = null;
     private final MultiPaperExecutor sendThread = new MultiPaperExecutor(new ConcurrentLinkedQueue<>());
+    public HashSet<String> subscribedWorlds = new HashSet<>();
 
     public ExternalServerConnection(Socket socket) {
         this.socket = socket;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 8da5b18aef1c16fa3293d32b8263d7500b5a0304..de38fd3dd41815791a07ad013c37cfe7f9f2a21d 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -1,5 +1,6 @@
 package puregero.multipaper;
 
+import net.minecraft.server.level.ServerPlayer;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
 import net.minecraft.server.MinecraftServer;
@@ -134,4 +135,16 @@ public class MultiPaper {
             });
         }
     }
+
+    public static void broadcastPacketToExternalServers(String world, ExternalServerPacket packet) {
+        forEachExternalServer(externalServer -> {
+            if (externalServer.getConnection() != null && externalServer.getConnection().isOpen() && externalServer.getConnection().subscribedWorlds.contains(world)) {
+                externalServer.getConnection().send(packet);
+            }
+        });
+    }
+
+    public static void broadcastPacketToExternalServers(ServerPlayer player, ExternalServerPacket packet) {
+        broadcastPacketToExternalServers(player.getLevel().getWorld().getName(), packet);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index c1d7076b6cc2ff5ec575d1fc1f51ee350afe3199..aa9e79748ef682de68ecea928c80e021a955a791 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -11,6 +11,7 @@ public class ExternalServerPacketSerializer {
     private static final List<IOExceptionFunction<FriendlyDataInputStream, ExternalServerPacket>> PACKET_DESERIALIZERS = new ArrayList<>();
     static {
         addPacket(HelloPacket.class, HelloPacket::new);
+        addPacket(SubscribeToWorldPacket.class, SubscribeToWorldPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, IOExceptionFunction<FriendlyDataInputStream, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
index 8684b38175af859eb4b1c489866df41dddaed2cc..f447caf8e99c19f65e1af34ebb28df7489e003ab 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
@@ -2,6 +2,8 @@ package puregero.multipaper.externalserverprotocol;
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
 import puregero.multipaper.ExternalServerConnection;
 import puregero.multipaper.MultiPaper;
 
@@ -31,5 +33,9 @@ public class HelloPacket extends ExternalServerPacket {
         LOGGER.info("Connected to external server " + bungeecordName);
         connection.externalServer = MultiPaper.getConnection().getOrCreateServer(bungeecordName);
         connection.externalServer.setConnection(connection);
+
+        for (World world : Bukkit.getWorlds()) {
+            connection.send(new SubscribeToWorldPacket(world.getName()));
+        }
     }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..eb733582d7b401fc657b48d5bf5474fa00630b6c
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java
@@ -0,0 +1,43 @@
+package puregero.multipaper.externalserverprotocol;
+
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+public class SubscribeToWorldPacket extends ExternalServerPacket {
+
+    private final String world;
+
+    public SubscribeToWorldPacket(String world) {
+        this.world = world;
+    }
+
+    public SubscribeToWorldPacket(FriendlyDataInputStream in) throws IOException {
+        world = in.readUTF();
+    }
+
+    @Override
+    public void write(FriendlyDataOutputStream out) throws IOException {
+        out.writeUTF(world);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            World bukkitWorld = Bukkit.getWorld(world);
+            if (connection.subscribedWorlds.add(world) && bukkitWorld instanceof CraftWorld craftWorld) {
+                onWorldSubscribe(connection, craftWorld);
+            }
+        });
+    }
+
+    private void onWorldSubscribe(ExternalServerConnection connection, CraftWorld craftWorld) {
+
+    }
+}
