From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 15 Nov 2021 13:27:45 +1000
Subject: [PATCH] Add peer-to-peer connection


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index e2901132b78126c0a4eb04363dfe6a0dccd1313f..bb373dba9d6a655a49a195dbf4d34cf37cd5b844 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -71,6 +71,7 @@ import co.aikar.timings.MinecraftTimings; // Paper
 import org.bukkit.event.server.ServerCommandEvent;
 import org.bukkit.craftbukkit.util.Waitable;
 import org.bukkit.event.server.RemoteServerCommandEvent;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public class DedicatedServer extends MinecraftServer implements ServerInterface {
@@ -290,6 +291,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
         // CraftBukkit end
 
+        MultiPaper.onStart(); // MultiPaper
+
         if (!this.usesAuthentication()) {
             DedicatedServer.LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
             DedicatedServer.LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
diff --git a/src/main/java/puregero/multipaper/ExternalServer.java b/src/main/java/puregero/multipaper/ExternalServer.java
index 35b8352a2c798c70cfebeddd2a95dc8edf1d21d9..ceaad2b5af855905ad1b87e7308fb354304819b1 100644
--- a/src/main/java/puregero/multipaper/ExternalServer.java
+++ b/src/main/java/puregero/multipaper/ExternalServer.java
@@ -6,6 +6,7 @@ public class ExternalServer {
     private int averageTickTime;
     private double tps;
     private long lastAlive;
+    private ExternalServerConnection connection;
 
     public ExternalServer(String name, boolean me) {
         this.name = name;
@@ -47,4 +48,12 @@ public class ExternalServer {
     public void setTps(double tps) {
         this.tps = tps;
     }
+
+    public void setConnection(ExternalServerConnection connection) {
+        this.connection = connection;
+    }
+
+    public ExternalServerConnection getConnection() {
+        return connection;
+    }
 }
diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
new file mode 100644
index 0000000000000000000000000000000000000000..519333b7d1629a3e8105d11e911df0b7430e00c5
--- /dev/null
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -0,0 +1,69 @@
+package puregero.multipaper;
+
+import puregero.multipaper.externalserverprotocol.ExternalServerPacket;
+import puregero.multipaper.externalserverprotocol.ExternalServerPacketSerializer;
+import puregero.multipaper.externalserverprotocol.HelloPacket;
+
+import java.io.*;
+import java.net.Socket;
+
+public class ExternalServerConnection extends Thread implements Closeable {
+
+    private Socket socket;
+    public ExternalServer externalServer = null;
+
+    public ExternalServerConnection(Socket socket) {
+        this.socket = socket;
+        start();
+    }
+
+    public int getPort() {
+        return socket.getLocalPort();
+    }
+
+    @Override
+    public void close() throws IOException {
+        socket.close();
+    }
+
+    @Override
+    public void run() {
+        try {
+            send(new HelloPacket(MultiPaper.getBungeeCordName()));
+
+            DataInputStream in = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+            while (!socket.isClosed()) {
+                ExternalServerPacket packet = ExternalServerPacketSerializer.deserialize(in);
+                packet.handle(this);
+            }
+        } catch (EOFException e) {
+            // Ignore
+        } catch (Exception e) {
+            if (!socket.isClosed()) {
+                e.printStackTrace();
+            }
+        }
+
+        try {
+            socket.close();
+        } catch (Exception e) {
+            // Ignore
+        }
+
+        if (externalServer != null && externalServer.getConnection() == this) {
+            externalServer.setConnection(null);
+        }
+    }
+
+    public void send(ExternalServerPacket packet) {
+        try {
+            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+            ExternalServerPacketSerializer.serialize(packet, new DataOutputStream(buffer));
+            synchronized (socket) {
+                socket.getOutputStream().write(buffer.toByteArray());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+}
diff --git a/src/main/java/puregero/multipaper/ExternalServerSocket.java b/src/main/java/puregero/multipaper/ExternalServerSocket.java
new file mode 100644
index 0000000000000000000000000000000000000000..87a0ea1ce48d9543007558df2e5326aba123e745
--- /dev/null
+++ b/src/main/java/puregero/multipaper/ExternalServerSocket.java
@@ -0,0 +1,41 @@
+package puregero.multipaper;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ServerSocket;
+
+public class ExternalServerSocket extends Thread implements Closeable {
+
+    private final ServerSocket socket;
+
+    public ExternalServerSocket() {
+        try {
+            socket = new ServerSocket(0);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        start();
+    }
+
+    public int getPort() {
+        return socket.getLocalPort();
+    }
+
+    @Override
+    public void close() throws IOException {
+        socket.close();
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (true) {
+                new ExternalServerConnection(socket.accept());
+            }
+        } catch (Exception e) {
+            if (!socket.isClosed()) {
+                e.printStackTrace();
+            }
+        }
+    }
+}
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 5ddb4c40c42f94b6a4b76dc0479c51dbcfe65f6e..2382f0980bef9704f227a331b5a441d748aff268 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -6,6 +6,7 @@ import java.nio.file.Files;
 
 public class MultiPaper {
     private static MultiPaperConnection multiPaperConnection = null;
+    private static ExternalServerSocket externalServerSocket = null;
 
     public static MultiPaperConnection getConnection() {
         if (multiPaperConnection == null) {
@@ -38,4 +39,13 @@ public class MultiPaper {
             e.printStackTrace();
         }
     }
+
+    public static void onStart() {
+        try {
+            externalServerSocket = new ExternalServerSocket();
+            getConnection().sendStart(externalServerSocket.getPort());
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 4aaa90a7023245791003dfe274af04bcf42ca1d9..6366cd1defa3de59963e8fd5ec4e87a567b81b54 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -9,6 +9,7 @@ import java.net.Socket;
 import java.net.SocketException;
 import java.nio.file.Files;
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Consumer;
 
@@ -128,6 +129,20 @@ public class MultiPaperConnection extends Thread {
         server.setLastAlive(System.currentTimeMillis());
     }
 
+    public void start(DataInputStream in, DataOutputSender out) throws IOException {
+        String address = in.readUTF();
+        int port = in.readInt();
+        CompletableFuture.runAsync(() -> {
+            try {
+                LOGGER.info("Connecting to external server " + address + ":" + port + "...");
+                Socket socket = new Socket(address, port);
+                new ExternalServerConnection(socket);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        });
+    }
+
     public void writeTickTime(long time, double tps) throws IOException {
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeTickTime");
@@ -135,4 +150,11 @@ public class MultiPaperConnection extends Thread {
         out.writeFloat((float) tps);
         out.send(null);
     }
+
+    public void sendStart(int port) throws IOException {
+        DataOutputSender out = new DataOutputSender(this);
+        out.writeUTF("start");
+        out.writeInt(port);
+        out.send(null);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..81d61a2d671fedd031eb2725ac3ab45622e196b1
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacket.java
@@ -0,0 +1,14 @@
+package puregero.multipaper.externalserverprotocol;
+
+import puregero.multipaper.ExternalServerConnection;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+public abstract class ExternalServerPacket {
+
+    public abstract void handle(ExternalServerConnection connection);
+
+    public abstract void write(DataOutputStream out) throws IOException;
+
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
new file mode 100644
index 0000000000000000000000000000000000000000..c21f3dc975820b3d979a4fc5c0fa981296375f0b
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -0,0 +1,39 @@
+package puregero.multipaper.externalserverprotocol;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ExternalServerPacketSerializer {
+
+    private static final List<Class<? extends ExternalServerPacket>> PACKETS = new ArrayList<>();
+    private static final List<IOExceptionFunction<DataInputStream, ExternalServerPacket>> PACKET_DESERIALIZERS = new ArrayList<>();
+    static {
+        addPacket(HelloPacket.class, HelloPacket::new);
+    }
+
+    private static void addPacket(Class<? extends ExternalServerPacket> clazz, IOExceptionFunction<DataInputStream, ExternalServerPacket> deserializer) {
+        PACKETS.add(clazz);
+        PACKET_DESERIALIZERS.add(deserializer);
+    }
+
+    private static int getPacketId(ExternalServerPacket packet) {
+        return PACKETS.indexOf(packet.getClass());
+    }
+
+    public static void serialize(ExternalServerPacket packet, DataOutputStream out) throws IOException {
+        out.write(getPacketId(packet));
+        packet.write(out);
+    }
+
+    public static ExternalServerPacket deserialize(DataInputStream in) throws IOException {
+        int packetId = in.read();
+        if (packetId == -1) {
+            throw new EOFException("End of stream reached while reading packet id");
+        }
+        return PACKET_DESERIALIZERS.get(packetId).apply(in);
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..98df09ee6eb37797ce764bbca69f482f4b90ab4d
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
@@ -0,0 +1,38 @@
+package puregero.multipaper.externalserverprotocol;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServer;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+public class HelloPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(HelloPacket.class.getSimpleName());
+
+    private final String bungeecordName;
+
+    public HelloPacket(String bungeecordName) {
+        this.bungeecordName = bungeecordName;
+    }
+
+    public HelloPacket(DataInputStream in) throws IOException {
+        bungeecordName = in.readUTF();
+    }
+
+    @Override
+    public void write(DataOutputStream out) throws IOException {
+        out.writeUTF(bungeecordName);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        LOGGER.info("Connected to external server " + bungeecordName);
+        connection.externalServer = MultiPaper.getConnection().getServersMap().computeIfAbsent(bungeecordName, key -> new ExternalServer(key, key.equals(MultiPaper.getBungeeCordName())));
+        connection.externalServer.setConnection(connection);
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/IOExceptionFunction.java b/src/main/java/puregero/multipaper/externalserverprotocol/IOExceptionFunction.java
new file mode 100644
index 0000000000000000000000000000000000000000..e75e8d222336483ce2d5cda77e570e161ea86cbb
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/IOExceptionFunction.java
@@ -0,0 +1,10 @@
+package puregero.multipaper.externalserverprotocol;
+
+import java.io.IOException;
+
+@FunctionalInterface
+public interface IOExceptionFunction<T, R> {
+
+    R apply(final T elem) throws IOException;
+
+}
