From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 14 Dec 2021 13:41:18 +1000
Subject: [PATCH] Use target UUID when saving conduits


diff --git a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
index 4dcac3899a500d8586580bcfd5b4516e1dcdcd4a..9e301e8d6a11c3360809055d52b5c40212d706cb 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -69,6 +69,7 @@ public class ConduitBlockEntity extends BlockEntity {
     @Override
     protected void saveAdditional(CompoundTag nbt) {
         super.saveAdditional(nbt);
+        if (this.destroyTargetUUID != null) nbt.putUUID("Target", this.destroyTargetUUID); // MultiPaper - Save UUID
         if (this.destroyTarget != null) {
             nbt.putUUID("Target", this.destroyTarget.getUUID());
         }
