From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 1fa17557091e430559ed7645090b26a0bcb2c8fc..5e49961c6503b6fc5a99929ea453cc0fdf043d6f 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -363,7 +363,7 @@ public class PaperWorldConfig {
 
     public boolean keepSpawnInMemory;
     private void keepSpawnInMemory() {
-        keepSpawnInMemory = getBoolean("keep-spawn-loaded", true);
+        keepSpawnInMemory = getBoolean("keep-spawn-loaded", false); // MultiPaper
         log("Keep spawn chunk loaded: " + keepSpawnInMemory);
     }
 
