From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 32f0cd29d1198fe320d10ccfe0b02f8632ac12aa..03cee3ac7dfaf1313345114811dafa53bf811cc0 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -347,7 +347,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public class Spawn extends ConfigurationPart {
         @RequiresSpigotInitialization(SpawnLoadedRangeInitializer.class)
         public short keepSpawnLoadedRange = 10;
-        public boolean keepSpawnLoaded = true;
+        public boolean keepSpawnLoaded = false; // MultiPaper
         public boolean allowUsingSignsInsideSpawnProtection = false;
     }
 
