From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 4598c54007a90e9ebf9294f3e99c9f340c4b4fe8..3dde3664a84693fac6ea7eb36612e048c5f1db27 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -320,7 +320,7 @@ public class PaperWorldConfig {
 
     public boolean keepSpawnInMemory;
     private void keepSpawnInMemory() {
-        keepSpawnInMemory = getBoolean("keep-spawn-loaded", true);
+        keepSpawnInMemory = getBoolean("keep-spawn-loaded", false); // MultiPaper
         log("Keep spawn chunk loaded: " + keepSpawnInMemory);
     }
 
