From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index fba36a0fd7036ba516e14c8d3521b470b0df17d1..c3289055139c95129cfb9d993c39cf06d82798d3 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -353,7 +353,7 @@ public class PaperWorldConfig {
 
     public boolean keepSpawnInMemory;
     private void keepSpawnInMemory() {
-        keepSpawnInMemory = getBoolean("keep-spawn-loaded", true);
+        keepSpawnInMemory = getBoolean("keep-spawn-loaded", false); // MultiPaper
         log("Keep spawn chunk loaded: " + keepSpawnInMemory);
     }
 
