From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 4532f3a0d74feae0a1249b53e1bfbc18a8808b32..a004ef05b107b554e586ed1a93b79c994048f642 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -355,7 +355,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public class Spawn extends ConfigurationPart {
         @RequiresSpigotInitialization(SpawnLoadedRangeInitializer.class)
         public short keepSpawnLoadedRange = 10;
-        public boolean keepSpawnLoaded = true;
+        public boolean keepSpawnLoaded = false; // MultiPaper
         public boolean allowUsingSignsInsideSpawnProtection = false;
     }
 
