From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 16 Nov 2021 01:58:28 +1000
Subject: [PATCH] Forward packets to the real player


diff --git a/src/main/java/puregero/multipaper/ExternalPlayer.java b/src/main/java/puregero/multipaper/ExternalPlayer.java
index ec509a18e6cc9937f1216b38f0f4c3f3df66c3d8..75609c1074a45f20137ef1bb3d6bae9c96a61618 100644
--- a/src/main/java/puregero/multipaper/ExternalPlayer.java
+++ b/src/main/java/puregero/multipaper/ExternalPlayer.java
@@ -6,6 +6,7 @@ import io.netty.util.concurrent.GenericFutureListener;
 import net.minecraft.network.Connection;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketFlow;
+import net.minecraft.network.protocol.game.*;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.network.ServerGamePacketListenerImpl;
@@ -66,11 +67,21 @@ public class ExternalPlayer extends ServerPlayer {
 
         @Override
         public void send(Packet<?> packet, @Nullable GenericFutureListener<? extends Future<? super Void>> callback) {
-            if (sendPackets) {
-
+            if (sendPackets && !(packet instanceof ClientboundPlayerAbilitiesPacket)
+                    && !(packet instanceof ClientboundSetPlayerTeamPacket)
+                    && !(packet instanceof ClientboundSetScorePacket)
+                    && !(packet instanceof ClientboundSetObjectivePacket)
+                    && !(packet instanceof ClientboundSetDisplayObjectivePacket)) {
+//                LOGGER.info("Forwarding packet " + packet);
+                externalServerConnection.sendPacket(ExternalPlayer.this, packet);
             } else {
 //                LOGGER.info("Not sending packet " + packet.getClass().getSimpleName());
             }
         }
     }
+
+    @Override
+    public boolean broadcastToPlayer(ServerPlayer spectator) {
+        return false;
+    }
 }
diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index c3468e42aca422c71d76c50e5d2769780fe05580..ee45bf94ec98d1f26aeb6c7fb0a5056d13ec78e3 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -1,16 +1,18 @@
 package puregero.multipaper;
 
+import net.minecraft.network.protocol.Packet;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.server.level.ServerPlayer;
 import puregero.multipaper.externalserverprotocol.ExternalServerPacket;
 import puregero.multipaper.externalserverprotocol.ExternalServerPacketSerializer;
 import puregero.multipaper.externalserverprotocol.HelloPacket;
+import puregero.multipaper.externalserverprotocol.SendPacketPacket;
 
 import java.io.*;
 import java.net.Socket;
 import java.nio.ByteBuffer;
-import java.util.HashSet;
+import java.util.*;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
@@ -22,6 +24,9 @@ public class ExternalServerConnection extends Thread implements Closeable {
     private final ExecutorService sendThread = Executors.newSingleThreadExecutor();
     public HashSet<String> subscribedWorlds = new HashSet<>();
 
+    private static final Queue<List<ExternalPlayer>> externalPlayerListPool = new LinkedList<>();
+    private final HashMap<Packet<?>, List<ExternalPlayer>> packetsToSend = new LinkedHashMap<>();
+
     public ExternalServerConnection(Socket socket) {
         this.socket = socket;
         start();
@@ -104,4 +109,32 @@ public class ExternalServerConnection extends Thread implements Closeable {
             }, sendThread);
         }
     }
+
+    public void tick() {
+        // Send the packets after each vanilla tick
+        synchronized (packetsToSend) {
+            packetsToSend.forEach((packet, players) -> {
+                send(new SendPacketPacket(players, packet));
+                players.clear();
+                externalPlayerListPool.add(players);
+            });
+            packetsToSend.clear();
+        }
+    }
+
+    public void sendPacket(ExternalPlayer player, Packet<?> packet) {
+        // Combine all the players that the packet's being sent to together
+        // so that the packet only needs to be sent to the external server
+        // just once, not duplicated for each player
+        synchronized (packetsToSend) {
+            List<ExternalPlayer> players = packetsToSend.computeIfAbsent(packet, key -> {
+                List<ExternalPlayer> list = externalPlayerListPool.poll();
+                if (list == null) {
+                    list = new ArrayList<>();
+                }
+                return list;
+            });
+            players.add(player);
+        }
+    }
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 2ca7ed1bc3b2cb28144b92be423b18d6bc1ace49..5fb10c2a35342685306d5c64fbd86342072af38b 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -98,6 +98,12 @@ public class MultiPaper {
     }
 
     public static void tick() {
+        for (ExternalServer server : getConnection().getServersMap().values()) {
+            if (server.getConnection() != null) {
+                // This tick function must be run after the vanilla tick
+                server.getConnection().tick();
+            }
+        }
     }
 
     public static void sendTickTime(long time, double tps) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index e6088d96e11356c811bd3731461d9a0b1a79254a..c1bac1bf5148126a0fcad517bf499833d31cf050 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -19,6 +19,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerChangeGamemodePacket.class, PlayerChangeGamemodePacket::new);
         addPacket(PlayerRespawnPacket.class, PlayerRespawnPacket::new);
         addPacket(PlayerChangeDimensionPacket.class, PlayerChangeDimensionPacket::new);
+        addPacket(SendPacketPacket.class, SendPacketPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, IOExceptionFunction<DataInputStream, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendPacketPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendPacketPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..086f6064400d03de8eb9ffe9fa1a5f349cdf8fe5
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendPacketPacket.java
@@ -0,0 +1,91 @@
+package puregero.multipaper.externalserverprotocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import net.minecraft.network.ConnectionProtocol;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.PacketFlow;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+public class SendPacketPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(SendPacketPacket.class.getSimpleName());
+
+    private final UUID[] uuids;
+    private final Packet<?> packet;
+
+    public SendPacketPacket(List<? extends ServerPlayer> players, Packet<?> packet) {
+        this.uuids = new UUID[players.size()];
+        for (int i = 0; i < players.size(); i++) {
+            this.uuids[i] = players.get(i).getUUID();
+        }
+        this.packet = packet;
+    }
+
+    public SendPacketPacket(ServerPlayer player, Packet<?> packet) {
+        this.uuids = new UUID[1];
+        this.uuids[0] = player.getUUID();
+        this.packet = packet;
+    }
+
+    public SendPacketPacket(DataInputStream in) throws IOException {
+        uuids = new UUID[in.readInt()];
+        for (int i = 0; i < uuids.length; i++) {
+            this.uuids[i] = readUUID(in);
+        }
+
+        int length = in.readInt();
+        byte[] bytes = new byte[length];
+        in.readFully(bytes);
+        ByteBuf buf = Unpooled.wrappedBuffer(bytes);
+        FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(buf);
+        int packetId = friendlyByteBuf.readVarInt();
+        packet = ConnectionProtocol.PLAY.createPacket(PacketFlow.CLIENTBOUND, packetId, friendlyByteBuf);
+    }
+
+    @Override
+    public void write(DataOutputStream out) throws IOException {
+        out.writeInt(uuids.length);
+        for (UUID uuid : uuids) {
+            writeUUID(out, uuid);
+        }
+
+        ConnectionProtocol protocol = ConnectionProtocol.getProtocolForPacket(packet);
+        Integer id = protocol.getPacketId(PacketFlow.CLIENTBOUND, packet);
+        ByteBuf buf = Unpooled.buffer();
+        FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(buf);
+        friendlyByteBuf.writeVarInt(id);
+        packet.write(friendlyByteBuf);
+        byte[] bytes = buf.array();
+        out.writeInt(bytes.length);
+        out.write(bytes);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            for (UUID uuid : uuids) {
+                ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+                if (player == null) {
+                    LOGGER.warn("Tried to send a packet to a non-existent player uuid " + player);
+                    return;
+                }
+
+                player.connection.send(packet);
+            }
+        });
+    }
+}
