From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 25 Nov 2021 18:10:39 +1000
Subject: [PATCH] Paper config changes


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 7e1add0f18da3ac2d93b084979d419cf45d2bca3..56e552eb327a5fdc49b6cb10f82b454618f81f7d 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -351,7 +351,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public class Spawn extends ConfigurationPart {
         @RequiresSpigotInitialization(SpawnLoadedRangeInitializer.class)
         public short keepSpawnLoadedRange = 10;
-        public boolean keepSpawnLoaded = true;
+        public boolean keepSpawnLoaded = false; // MultiPaper
         public boolean allowUsingSignsInsideSpawnProtection = false;
     }
 
