From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 21 Dec 2021 00:37:18 +1000
Subject: [PATCH] Sync experience


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index edb49e08f91021345f3928b8f6b717bcf988ac54..aaf77e08d0d9f391796150ecfc74147478cb5ff8 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -173,10 +173,7 @@ import org.bukkit.event.player.PlayerTeleportEvent;
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.inventory.MainHand;
 import puregero.multipaper.*;
-import puregero.multipaper.externalserverprotocol.PlayerChangeDimensionPacket;
-import puregero.multipaper.externalserverprotocol.PlayerChangeGamemodePacket;
-import puregero.multipaper.externalserverprotocol.PlayerCreatePacket;
-import puregero.multipaper.externalserverprotocol.PlayerRemovePacket;
+import puregero.multipaper.externalserverprotocol.*;
 // CraftBukkit end
 
 public class ServerPlayer extends Player {
@@ -627,6 +624,20 @@ public class ServerPlayer extends Player {
         this.lastSentExp = -1;
     }
 
+    // MultiPaper start - sync xp
+    public float lastExperienceProgress = 0;
+    public int lastTotalExperience = 0;
+    public int lastExperienceLevel = 0;
+    public void syncExperience() {
+        if (experienceProgress != lastExperienceProgress || totalExperience != lastTotalExperience || experienceLevel != lastExperienceLevel) {
+            lastExperienceProgress = experienceProgress;
+            lastTotalExperience = totalExperience;
+            lastExperienceLevel = experienceLevel;
+            MultiPaper.broadcastPacketToExternalServers(this, new PlayerExperienceUpdatePacket(this, experienceProgress, totalExperience, experienceLevel));
+        }
+    }
+    // MultiPaper end
+
     public void initMenu(AbstractContainerMenu screenHandler) {
         screenHandler.addSlotListener(this.containerListener);
         screenHandler.setSynchronizer(this.containerSynchronizer);
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 2906fd26bc135a7b8c1b0519902db1211ddd13b1..0b6aee5d5bd8646bfcc7ecda9487ea0de7f3a65b 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -86,6 +86,8 @@ public class MultiPaper {
 
                 externalPlayer.tickAttackStrength();
             }
+
+            player.syncExperience();
         }
 
         for (ExternalServer server : getConnection().getServersMap().values()) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index c559c3dc0f927dfa0c7d24d39323ba484e43a899..33375df7a8431f71b0573726c1cfc1de78023810 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -59,6 +59,7 @@ public class ExternalServerPacketSerializer {
         addPacket(AdvancementGrantProgressPacket.class, AdvancementGrantProgressPacket::new);
         addPacket(WeatherUpdatePacket.class, WeatherUpdatePacket::new);
         addPacket(PlayerStatsIncreasePacket.class, PlayerStatsIncreasePacket::new);
+        addPacket(PlayerExperienceUpdatePacket.class, PlayerExperienceUpdatePacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerExperienceUpdatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerExperienceUpdatePacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..cb0f815c026ce2547a64cacef28e646d2ce44c81
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerExperienceUpdatePacket.java
@@ -0,0 +1,58 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerExperienceUpdatePacket extends ExternalServerPacket {
+    private static final Logger LOGGER = LogManager.getLogger(PlayerExperienceUpdatePacket.class.getSimpleName());
+
+    private final UUID uuid;
+    private final float progress;
+    private final int total;
+    private final int level;
+
+    public PlayerExperienceUpdatePacket(ServerPlayer player, float progress, int total, int level) {
+        this.uuid = player.getUUID();
+        this.progress = progress;
+        this.total = total;
+        this.level = level;
+    }
+
+    public PlayerExperienceUpdatePacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+        progress = in.readFloat();
+        total = in.readInt();
+        level = in.readInt();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeFloat(progress);
+        out.writeInt(total);
+        out.writeInt(level);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            if (player == null) {
+                LOGGER.warn("Could not find player " + uuid);
+                return;
+            }
+
+            player.experienceProgress = player.lastExperienceProgress = progress;
+            player.totalExperience = player.lastTotalExperience = total;
+            player.experienceLevel = player.lastExperienceLevel = level;
+        });
+    }
+}
