From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 8 Mar 2022 21:43:27 +1000
Subject: [PATCH] Store the known entity


diff --git a/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java b/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
index 2d3aacdae95963385ea228e73a2073a6fd96e640..335123dbb1c8d48374549621e5ee6062508e7b0e 100644
--- a/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -14,11 +14,7 @@ import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.io.Writer;
-import java.util.List;
-import java.util.Objects;
-import java.util.Queue;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
@@ -39,7 +35,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory;
 public class PersistentEntitySectionManager<T extends EntityAccess> implements AutoCloseable {
 
     static final Logger LOGGER = LogUtils.getLogger();
-    final Set<UUID> knownUuids = Sets.newHashSet();
+    public final HashMap<UUID, T> knownUuids = new HashMap<>(); // MultiPaper - store the known entity
     final LevelCallback<T> callbacks;
     public final EntityPersistentStorage<T> permanentStorage;
     private final EntityLookup<T> visibleEntityStorage = new EntityLookup<>();
@@ -139,7 +135,7 @@ public class PersistentEntitySectionManager<T extends EntityAccess> implements A
 
     private boolean addEntityUuid(T entity) {
         org.spigotmc.AsyncCatcher.catchOp("Entity add by UUID"); // Paper
-        if (!this.knownUuids.add(entity.getUUID())) {
+        if (this.knownUuids.put(entity.getUUID(), entity) != null) {
             // Paper start
             T conflict = this.visibleEntityStorage.getEntity(entity.getUUID());
             if (conflict != null && ((Entity) conflict).isRemoved()) {
@@ -467,7 +463,7 @@ public class PersistentEntitySectionManager<T extends EntityAccess> implements A
     }
 
     public boolean isLoaded(UUID uuid) {
-        return this.knownUuids.contains(uuid);
+        return this.knownUuids.containsKey(uuid);
     }
 
     public LevelEntityGetter<T> getEntityGetter() {
