From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 21:32:25 +1000
Subject: [PATCH] Subscribe to worlds


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1fc201ee73fbc5e56f7fa753096ba21fdee6609d..56ca6d80e859a1be72d6f988855ab62be310ead7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -246,6 +246,8 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.SubscribeToWorldPacket;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -1358,6 +1360,7 @@ public final class CraftServer implements Server {
             return;
         }
         this.worlds.put(world.getName().toLowerCase(java.util.Locale.ENGLISH), world);
+        MultiPaper.broadcastPacketToExternalServers(new SubscribeToWorldPacket(world.getName())); // MultiPaper
     }
 
     @Override
diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index dba7879ae903f15f47790db45e9c2e7fdd5e07dd..2b05d88d64987d40d4080ce8c4557ceebff6454e 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -10,12 +10,14 @@ import puregero.multipaper.externalserverprotocol.*;
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.util.HashSet;
 
 public class ExternalServerConnection extends ChannelInitializer<SocketChannel> implements Closeable {
 
     private Channel channel;
     public long nanoTime = 0;
     public ExternalServer externalServer = null;
+    public HashSet<String> subscribedWorlds = new HashSet<>();
 
     public ExternalServerConnection() {
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index e82475730f2be2a2c8a326924610c75f260515fb..cd4c4917b0266cb4dcba6bbaf96c96b09342bda6 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -1,5 +1,6 @@
 package puregero.multipaper;
 
+import net.minecraft.server.level.ServerPlayer;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
 import net.minecraft.server.MinecraftServer;
@@ -127,4 +128,16 @@ public class MultiPaper {
             });
         }
     }
+
+    public static void broadcastPacketToExternalServers(String world, ExternalServerPacket packet) {
+        forEachExternalServer(externalServer -> {
+            if (externalServer.getConnection() != null && externalServer.getConnection().isOpen() && externalServer.getConnection().subscribedWorlds.contains(world)) {
+                externalServer.getConnection().send(packet);
+            }
+        });
+    }
+
+    public static void broadcastPacketToExternalServers(ServerPlayer player, ExternalServerPacket packet) {
+        broadcastPacketToExternalServers(player.getLevel().getWorld().getName(), packet);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index a18780b676d0e4ee1a5bf89d0bde396e9eea3804..4f85ce75025c090c08b1def93d881b1fbbed2202 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -12,6 +12,7 @@ public class ExternalServerPacketSerializer {
     private static final List<Function<FriendlyByteBuf, ExternalServerPacket>> PACKET_DESERIALIZERS = new ArrayList<>();
     static {
         addPacket(HelloPacket.class, HelloPacket::new);
+        addPacket(SubscribeToWorldPacket.class, SubscribeToWorldPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
index 8fd3e37dea839ccc39aac19b50790705016beda6..772381766bebf388664e7488fd38ebb0bdbf913f 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
@@ -3,6 +3,8 @@ package puregero.multipaper.externalserverprotocol;
 import net.minecraft.network.FriendlyByteBuf;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
 import puregero.multipaper.ExternalServerConnection;
 import puregero.multipaper.MultiPaper;
 
@@ -61,5 +63,11 @@ public class HelloPacket extends ExternalServerPacket {
             connection.nanoTime = nanoTime;
             connection.send(new HelloPacket(MultiPaper.getBungeeCordName(), nanoTime));
         }
+
+        MultiPaper.runSync(() -> {
+            for (World world : Bukkit.getWorlds()) {
+                connection.send(new SubscribeToWorldPacket(world.getName()));
+            }
+        });
     }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..5cbf42b51d99d13545b17be258e52f0638d73ef9
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java
@@ -0,0 +1,40 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+public class SubscribeToWorldPacket extends ExternalServerPacket {
+
+    private final String world;
+
+    public SubscribeToWorldPacket(String world) {
+        this.world = world;
+    }
+
+    public SubscribeToWorldPacket(FriendlyByteBuf in) {
+        world = in.readUtf();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            World bukkitWorld = Bukkit.getWorld(world);
+            if (connection.subscribedWorlds.add(world) && bukkitWorld instanceof CraftWorld craftWorld) {
+                onWorldSubscribe(connection, craftWorld);
+            }
+        });
+    }
+
+    private void onWorldSubscribe(ExternalServerConnection connection, CraftWorld craftWorld) {
+
+    }
+}
