From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 21:32:25 +1000
Subject: [PATCH] Subscribe to worlds


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index fa125f9946dc6b5eaa18b3f9f94d2833c40bf15f..98fee13364c5c57d2e530d21629571f8c7aa30f7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -243,6 +243,8 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.SubscribeToWorldPacket;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -1353,6 +1355,7 @@ public final class CraftServer implements Server {
             return;
         }
         this.worlds.put(world.getName().toLowerCase(java.util.Locale.ENGLISH), world);
+        MultiPaper.broadcastPacketToExternalServers(new SubscribeToWorldPacket(world.getName())); // MultiPaper
     }
 
     @Override
diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index f7c086c83edd7205d05b6413935fad4b7495c571..b091d54c1b599647b7db33315983594712ffb0bc 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -18,6 +18,7 @@ import puregero.multipaper.mastermessagingprotocol.MessageLengthEncoder;
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.util.HashSet;
 
 public class ExternalServerConnection extends ChannelInitializer<SocketChannel> implements Closeable {
 
@@ -26,6 +27,7 @@ public class ExternalServerConnection extends ChannelInitializer<SocketChannel>
     private Channel channel;
     public long nanoTime = 0;
     public ExternalServer externalServer = null;
+    public HashSet<String> subscribedWorlds = new HashSet<>();
 
     public ExternalServerConnection() {
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 63f573a93c8f46f2bc41cd6d81ae9d1e3fdc4e64..7cd6e0c934944b92852f650736bfaf718f78c499 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -1,6 +1,7 @@
 package puregero.multipaper;
 
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
 import puregero.multipaper.externalserverprotocol.ExternalServerPacket;
 import puregero.multipaper.mastermessagingprotocol.messages.masterbound.StartMessage;
 import puregero.multipaper.mastermessagingprotocol.messages.masterbound.WriteTickTimeMessage;
@@ -67,4 +68,16 @@ public class MultiPaper {
             });
         }
     }
+
+    public static void broadcastPacketToExternalServers(String world, ExternalServerPacket packet) {
+        forEachExternalServer(externalServer -> {
+            if (externalServer.getConnection() != null && externalServer.getConnection().isOpen() && externalServer.getConnection().subscribedWorlds.contains(world)) {
+                externalServer.getConnection().send(packet);
+            }
+        });
+    }
+
+    public static void broadcastPacketToExternalServers(ServerPlayer player, ExternalServerPacket packet) {
+        broadcastPacketToExternalServers(player.getLevel().getWorld().getName(), packet);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 8ae5d51ce27c013595462029c8237a1cffe22571..e3cf79d81b58fabeddf5f42a9351f36a5f2197be 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -13,6 +13,7 @@ public class ExternalServerPacketSerializer {
     static {
         addPacket(HelloPacket.class, HelloPacket::new);
         addPacket(SetCompressionPacket.class, SetCompressionPacket::new);
+        addPacket(SubscribeToWorldPacket.class, SubscribeToWorldPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
index 75f5853169bfa114beb9e9d30abc3919a6c06728..7a57ad7cebafd7d582b968b566587b718c404dc5 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/HelloPacket.java
@@ -5,6 +5,8 @@ import io.netty.handler.codec.compression.Zstd;
 import net.minecraft.network.FriendlyByteBuf;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
 import puregero.multipaper.ExternalServerConnection;
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperConfig;
@@ -101,5 +103,11 @@ public class HelloPacket extends ExternalServerPacket {
             connection.nanoTime = nanoTime;
             connection.send(new HelloPacket(MultiPaperConfig.bungeeCordName, nanoTime, connection.getSupportedCompressionFlags()));
         }
+
+        MultiPaper.runSync(() -> {
+            for (World world : Bukkit.getWorlds()) {
+                connection.send(new SubscribeToWorldPacket(world.getName()));
+            }
+        });
     }
 }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..5cbf42b51d99d13545b17be258e52f0638d73ef9
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SubscribeToWorldPacket.java
@@ -0,0 +1,40 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+public class SubscribeToWorldPacket extends ExternalServerPacket {
+
+    private final String world;
+
+    public SubscribeToWorldPacket(String world) {
+        this.world = world;
+    }
+
+    public SubscribeToWorldPacket(FriendlyByteBuf in) {
+        world = in.readUtf();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            World bukkitWorld = Bukkit.getWorld(world);
+            if (connection.subscribedWorlds.add(world) && bukkitWorld instanceof CraftWorld craftWorld) {
+                onWorldSubscribe(connection, craftWorld);
+            }
+        });
+    }
+
+    private void onWorldSubscribe(ExternalServerConnection connection, CraftWorld craftWorld) {
+
+    }
+}
