From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 23:10:18 +1000
Subject: [PATCH] Don't suck from external minecarts and boats


diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 679af17f5bb673e285f1ee599cb9a6a864a2df40..c3a8fc6bec2fce703dc872bc4ae977f77363a390 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -45,6 +45,7 @@ import org.bukkit.event.inventory.InventoryPickupItemEvent;
 import org.bukkit.inventory.Inventory;
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperContainerHandler;
+import puregero.multipaper.MultiPaperEntityInteractHandler;
 import puregero.multipaper.externalserverprotocol.PullItemFromContainerPacket;
 // CraftBukkit end
 
@@ -510,7 +511,12 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public static boolean suckInItems(Level world, Hopper hopper) {
         Container iinventory = HopperBlockEntity.getSourceContainer(world, hopper);
 
-        if (iinventory != null) {
+        // MultiPaper start - don't suck from external minecarts and boats
+        if (iinventory != null
+                && !(iinventory instanceof Entity entity && MultiPaperEntityInteractHandler.getOwner(entity) != null)
+                && !(iinventory instanceof CompoundContainer compoundContainer && !MultiPaper.isChunkLocal(world.getChunkIfLoaded(MultiPaperContainerHandler.getPrimaryChest(compoundContainer).getBlockPos())))
+                && !(iinventory instanceof BlockEntity blockEntity && !MultiPaper.isChunkLocal(world.getChunkIfLoaded(blockEntity.getBlockPos())))) {
+        // Multipaper end
             Direction enumdirection = Direction.DOWN;
 
             // Paper start - optimize hoppers and remove streams
