From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 4 Dec 2021 12:23:17 +1000
Subject: [PATCH] Sync enderchests


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 11ff76168ed83b54a191237a31f0ceeff2475a89..105e9da2b3316f6752eb03ae08b5213f4d9d25e0 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -166,6 +166,7 @@ import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.inventory.MainHand;
 import puregero.multipaper.ExternalServer;
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperEnderChestHandler;
 import puregero.multipaper.externalserverprotocol.PlayerChangeDimensionPacket;
 import puregero.multipaper.externalserverprotocol.PlayerChangeGamemodePacket;
 // CraftBukkit end
@@ -335,6 +336,7 @@ public class ServerPlayer extends Player {
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
         this.cachedSingleMobDistanceMap = new com.destroystokyo.paper.util.PooledHashSets.PooledObjectLinkedOpenHashSet<>(this); // Paper
+        enderChestInventory.addListener(new MultiPaperEnderChestHandler(this)); // MultiPaper
     }
     // Paper start - Chunk priority
     public BlockPos getPointInFront(double inFront) {
diff --git a/src/main/java/puregero/multipaper/MultiPaperEnderChestHandler.java b/src/main/java/puregero/multipaper/MultiPaperEnderChestHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..b26300e5bacaf6e6e9ba9c0f262a2e065d02ae54
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperEnderChestHandler.java
@@ -0,0 +1,63 @@
+package puregero.multipaper;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.Container;
+import net.minecraft.world.ContainerListener;
+import net.minecraft.world.item.ItemStack;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.scheduler.CraftScheduler;
+import puregero.multipaper.externalserverprotocol.PlayerInventoryUpdatePacket;
+
+public class MultiPaperEnderChestHandler implements ContainerListener {
+
+    private static boolean broadcastChanges = true;
+
+    private final ServerPlayer player;
+    private ItemStack[] sentItems = new ItemStack[0];
+
+    public MultiPaperEnderChestHandler(ServerPlayer player) {
+        this.player = player;
+    }
+
+    @Override
+    public void containerChanged(Container container) {
+        if (container.getContainerSize() != sentItems.length) {
+            sentItems = new ItemStack[container.getContainerSize()];
+        }
+
+        if (!player.didPlayerJoinEvent && broadcastChanges) {
+            // Wait till they join to broadcast changes
+            ((CraftScheduler) Bukkit.getScheduler()).scheduleInternalTask(() -> containerChanged(container), 1, "MultiPaperEnderChestHandler-containerChanged");
+            return;
+        }
+
+        CompoundTag itemsRoot = new CompoundTag();
+        ListTag items = new ListTag();
+        for (int i = 0; i < sentItems.length; i++) {
+            ItemStack item = container.getItem(i);
+            if (!item.equals(sentItems[i])) {
+                sentItems[i] = item.copy();
+
+                if (broadcastChanges) {
+                    CompoundTag itemToSend = new CompoundTag();
+                    itemToSend.putInt("Slot", i);
+                    item.save(itemToSend);
+                    items.add(itemToSend);
+                }
+            }
+        }
+
+        if (!items.isEmpty()) {
+            itemsRoot.put("items", items);
+            MultiPaper.broadcastPacketToExternalServers(new PlayerInventoryUpdatePacket(player, "enderchest", itemsRoot));
+        }
+    }
+
+    public static void updateInventory(ServerPlayer player, int slot, ItemStack item) {
+        broadcastChanges = false;
+        player.getEnderChestInventory().setItem(slot, item);
+        broadcastChanges = true;
+    }
+}
diff --git a/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java b/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java
index 0b5ec3d93a0a1ff18baff8c3994d81980187c818..a91b0e620c19c51fd30f530162ff613c2d124ab3 100644
--- a/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperInventoryHandler.java
@@ -86,6 +86,7 @@ public class MultiPaperInventoryHandler {
             case "items" -> player.getInventory().items.set(slot, item);
             case "armor" -> player.getInventory().armor.set(slot, item);
             case "offhand" -> player.getInventory().offhand.set(slot, item);
+            case "enderchest" -> MultiPaperEnderChestHandler.updateInventory(player, slot, item);
             default -> throw new IllegalArgumentException("Unknown inventory component of " + name);
         }
         item.dirty = false;
