From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 21 Nov 2021 01:24:00 +1000
Subject: [PATCH] Sync time


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 0a2a5709ad06263cf849a61f6a43279dccb539cf..732cca1df8364ae7f40ac497bfc78f0e5382dc16 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1574,6 +1574,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             long worldTime = world.getGameTime();
             final ClientboundSetTimePacket worldPacket = new ClientboundSetTimePacket(worldTime, dayTime, doDaylight);
             for (Player entityhuman : world.players()) {
+                if (MultiPaper.isExternalPlayer(entityhuman)) continue; // MultiPaper - Don't send the time to external players
                 if (!(entityhuman instanceof ServerPlayer) || (tickCount + entityhuman.getId()) % 20 != 0) {
                     continue;
                 }
diff --git a/src/main/java/net/minecraft/server/commands/TimeCommand.java b/src/main/java/net/minecraft/server/commands/TimeCommand.java
index 29ccc4286786cdf0b11d7c28f6bfc453fb27fd45..7294bc593c7a24c1645e0e89dd7e7365dab118d1 100644
--- a/src/main/java/net/minecraft/server/commands/TimeCommand.java
+++ b/src/main/java/net/minecraft/server/commands/TimeCommand.java
@@ -11,6 +11,8 @@ import net.minecraft.server.level.ServerLevel;
 // CraftBukkit start
 import org.bukkit.Bukkit;
 import org.bukkit.event.world.TimeSkipEvent;
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.TimeUpdatePacket;
 // CraftBukkit end
 
 public class TimeCommand {
@@ -61,6 +63,7 @@ public class TimeCommand {
             Bukkit.getPluginManager().callEvent(event);
             if (!event.isCancelled()) {
                 worldserver.setDayTime((long) worldserver.getDayTime() + event.getSkipAmount());
+                MultiPaper.broadcastPacketToExternalServers(new TimeUpdatePacket(worldserver, true)); // MultiPaper
             }
             // CraftBukkit end
         }
@@ -80,6 +83,7 @@ public class TimeCommand {
             Bukkit.getPluginManager().callEvent(event);
             if (!event.isCancelled()) {
                 worldserver.setDayTime(worldserver.getDayTime() + event.getSkipAmount());
+                MultiPaper.broadcastPacketToExternalServers(new TimeUpdatePacket(worldserver, true)); // MultiPaper
             }
             // CraftBukkit end
         }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index f4ab73db8c2c98462722f68350e7036097b9f784..0ba73edbd8ebd458bbf4acb3903d0950774bb444 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -163,6 +163,7 @@ import org.bukkit.event.world.TimeSkipEvent;
 import it.unimi.dsi.fastutil.ints.IntArrayList; // Paper
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
+import puregero.multipaper.externalserverprotocol.TimeUpdatePacket;
 
 public class ServerLevel extends Level implements WorldGenLevel {
 
@@ -592,6 +593,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 getCraftServer().getPluginManager().callEvent(event);
                 if (!event.isCancelled()) {
                     this.setDayTime(this.getDayTime() + event.getSkipAmount());
+                    MultiPaper.broadcastPacketToExternalServers(new TimeUpdatePacket(this, true)); // MultiPaper
                 }
             }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index c28f7074a07051ea9aee5dc67dfef9193426a42d..a8246278328d7712435173fc73a3359e846a7206 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -116,6 +116,8 @@ import org.bukkit.util.BoundingBox;
 import org.bukkit.util.Consumer;
 import org.bukkit.util.RayTraceResult;
 import org.bukkit.util.Vector;
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.TimeUpdatePacket;
 
 public class CraftWorld extends CraftRegionAccessor implements World {
     public static final int CUSTOM_DIMENSION_OFFSET = 10;
@@ -819,6 +821,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         }
 
         this.world.setDayTime(this.world.getDayTime() + event.getSkipAmount());
+        MultiPaper.broadcastPacketToExternalServers(new TimeUpdatePacket(world, true)); // MultiPaper
 
         // Forces the client to update to the new time immediately
         for (Player p : this.getPlayers()) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index f6a01b7dcbc307f2f75cd1382eae5754d250bbfc..9a650f1370788710f0f8051529a6a7e1e8d6f40a 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -187,6 +187,12 @@ public class MultiPaper {
                 externalPlayer.checkInsideBlocks();
             }
         }
+
+        if (MinecraftServer.getServer().getTickCount() % 20 == 0) {
+            for (ServerLevel level : MinecraftServer.getServer().getAllLevels()) {
+                broadcastPacketToExternalServers(new TimeUpdatePacket(level));
+            }
+        }
     }
 
     public static void sendTickTime(long time, double tps) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 1b58d7b815e9294c5bce5de02e4c8a1b3952fbfe..a086c845e915de7d65457bb8111f7083b23541f2 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -24,6 +24,7 @@ public class ExternalServerPacketSerializer {
         addPacket(SendTickListPacket.class, SendTickListPacket::new);
         addPacket(PlayerActionPacket.class, PlayerActionPacket::new);
         addPacket(PlayerInventoryUpdatePacket.class, PlayerInventoryUpdatePacket::new);
+        addPacket(TimeUpdatePacket.class, TimeUpdatePacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, IOExceptionFunction<FriendlyDataInputStream, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/TimeUpdatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/TimeUpdatePacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..ca7f6c514bfc1406c7d1f134f6b59fed26528b06
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/TimeUpdatePacket.java
@@ -0,0 +1,60 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.storage.PrimaryLevelData;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+public class TimeUpdatePacket extends ExternalServerPacket {
+
+    private final String world;
+    private final long gameTime;
+    private final long dayTime;
+    private final boolean force;
+
+    public TimeUpdatePacket(Level level) {
+        this(level, false);
+    }
+
+    public TimeUpdatePacket(Level level, boolean force) {
+        this.world = level.getWorld().getName();
+        this.gameTime = level.getGameTime();
+        this.dayTime = level.getDayTime();
+        this.force = force;
+    }
+
+    public TimeUpdatePacket(FriendlyDataInputStream in) throws IOException {
+        world = in.readUTF();
+        gameTime = in.readLong();
+        dayTime = in.readLong();
+        force = in.readBoolean();
+    }
+
+    @Override
+    public void write(FriendlyDataOutputStream out) throws IOException {
+        out.writeUTF(world);
+        out.writeLong(gameTime);
+        out.writeLong(dayTime);
+        out.writeBoolean(force);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        World bukkitWorld = Bukkit.getWorld(world);
+        if (bukkitWorld != null) {
+            ServerLevel level = ((CraftWorld) bukkitWorld).getHandle();
+            if (force || level.getGameTime() < gameTime - 20) {
+                // We're more than a second behind, update us
+                ((PrimaryLevelData) level.levelData).setGameTime(gameTime);
+                ((PrimaryLevelData) level.levelData).setDayTime(dayTime);
+            }
+        }
+    }
+}
