From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 Dec 2021 16:39:52 +1000
Subject: [PATCH] Reobf error with ServerPlaceRecipe


diff --git a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
index c8e9e85728d6b60bb829e69c015ab3c5172afd54..1b0df32c493f32ce0899e3a3cdf85997732b2b32 100644
--- a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -188,7 +188,7 @@ public class ServerPlaceRecipe<C extends Container> implements PlaceRecipe<Integ
     private int getAmountOfFreeSlotsInInventory() {
         int i = 0;
 
-        for(ItemStack itemStack : this.inventory.items) {
+        for(ItemStack itemStack : this.inventory.items) { // MultiPaper - patch on this line to fix reobf error
             if (itemStack.isEmpty()) {
                 ++i;
             }
