From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 17:06:56 +1000
Subject: [PATCH] Ensure spawn chunk is loaded for end to overworld portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index cefb360dc9bdebae7d31a63fb32c460819c2608d..8c7ea35d6729102e069df771f0c713433a91348e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3347,6 +3347,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
                 // Paper end
                 blockposition1 = ServerLevel.END_SPAWN_POINT;
             } else {
+                // MultiPaper start - Ensure spawn chunk is always loaded before calculating Y coordinate - is Paper's patch above alright? it literally does not accomplish its goal in the slightest
+                destination.getChunkAt(destination.getSharedSpawnPos());
+                // MultiPaper end
                 blockposition1 = destination.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, destination.getSharedSpawnPos());
             }
             // CraftBukkit start
