From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 17:06:56 +1000
Subject: [PATCH] Ensure spawn chunk is loaded for end to overworld portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e68125618c028c4bdf81c141a861a3daa8a17339..f33d694d05bf720a43767afd4334c4448b34b2f6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3340,6 +3340,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
                 // Paper end
                 blockposition1 = ServerLevel.END_SPAWN_POINT;
             } else {
+                // MultiPaper start - Ensure spawn chunk is always loaded before calculating Y coordinate - is Paper's patch above alright? it literally does not accomplish its goal in the slightest
+                destination.getChunkAt(destination.getSharedSpawnPos());
+                // MultiPaper end
                 blockposition1 = destination.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, destination.getSharedSpawnPos());
             }
             // CraftBukkit start
