From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 29 Dec 2021 15:01:38 +1000
Subject: [PATCH] Add temporary debug logs for checking if a chunk can be
 loaded twice


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index b12b0432516770fa53f4b0076effd03d0b8fc41e..fce9a2b78d92139b72f4f178e5600729f6054a28 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -1043,7 +1043,16 @@ public class LevelChunk extends ChunkAccess {
 
     }
 
+    // MultiPaper start - temporary debug logs for checking if a chunk can be loaded twice
+    private boolean hasBeenLoadedBefore = false;
     public void setLoaded(boolean loadedToWorld) {
+        if (loaded) {
+            if (hasBeenLoadedBefore) {
+                new Exception("Loading a chunk that's already been loaded " + level.getWorld().getName() + "," + chunkPos.x + "," + chunkPos.z).printStackTrace();
+            }
+            hasBeenLoadedBefore = true;
+        }
+    // MultiPaper end - temporary debug logs for checking if a chunk can be loaded twice
         this.loaded = loadedToWorld;
     }
 
