From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Tick attack strength for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index c3e40d95f1a5d9587cf5018b6960d32c60619480..45f496eb92158658b48d0a35e0123eeedfbb3e59 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -298,6 +298,12 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
         ItemStack itemstack = this.getMainHandItem();
 
@@ -308,6 +314,10 @@ public abstract class Player extends LivingEntity {
 
             this.lastItemInMainHand = itemstack.copy();
         }
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
 
         this.turtleHelmetTick();
         this.cooldowns.tick();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 390da3100df08cca13faf3a98386bdf7640ddf2c..eb5e38063ac84a90982c3a60e367b5b116bccbf8 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -113,6 +113,7 @@ public class MultiPaper {
         for (ServerPlayer player : MinecraftServer.getServer().getPlayerList().players) {
             if (player instanceof ExternalPlayer) {
                 player.detectEquipmentUpdates();
+                player.tickAttackStrength();
             }
         }
     }
