From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 30 Dec 2021 15:09:05 +1000
Subject: [PATCH] Broadcast block change when right-clicking a block to ensure
 external servers are in sync


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 9d69cd83996cfc9b06793ebcfda9cd8219ad7e22..05c857a1df2519e64acae4a59a7ac57ab771158f 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2130,6 +2130,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
                     this.player.connection.send(new ClientboundBlockUpdatePacket(worldserver, blockposition));
                     this.player.connection.send(new ClientboundBlockUpdatePacket(worldserver, blockposition.relative(enumdirection)));
+
+                // MultiPaper - broadcast block change to other servers as they may be out of sync
+                worldserver.chunkSource.blockChanged(blockposition);
+                worldserver.chunkSource.blockChanged(blockposition.relative(enumdirection));
+                // MultiPaper - broadcast block change to other servers as they may be out of sync
                 } else {
                     ServerGamePacketListenerImpl.LOGGER.warn("Rejecting UseItemOnPacket from {}: Location {} too far away from hit block {}.", new Object[]{this.player.getGameProfile().getName(), vec3d, blockposition});
                 }
