From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 17:06:56 +1000
Subject: [PATCH] Ensure spawn chunk is loaded for end to overworld portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 34113186a8c08699db7c8758a1703f25eab48b2c..40cfbcd454a72fe88222697ff679d4062379595f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3337,6 +3337,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
                 // Paper end
                 blockposition1 = ServerLevel.END_SPAWN_POINT;
             } else {
+                // MultiPaper start - Ensure spawn chunk is always loaded before calculating Y coordinate - is Paper's patch above alright? it literally does not accomplish its goal in the slightest
+                destination.getChunkAt(destination.getSharedSpawnPos());
+                // MultiPaper end
                 blockposition1 = destination.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, destination.getSharedSpawnPos());
             }
             // CraftBukkit start
