From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 17:06:56 +1000
Subject: [PATCH] Ensure spawn chunk is loaded for end to overworld portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0963b09c91b14d5fbffc5bd1804fdf79cd462845..1858ee485a1c8b51bfac551aa8cd5df524d3a968 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3349,6 +3349,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
                 // Paper end
                 blockposition1 = ServerLevel.END_SPAWN_POINT;
             } else {
+                // MultiPaper start - Ensure spawn chunk is always loaded before calculating Y coordinate - is Paper's patch above alright? it literally does not accomplish its goal in the slightest
+                destination.getChunkAt(destination.getSharedSpawnPos());
+                // MultiPaper end
                 blockposition1 = destination.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, destination.getSharedSpawnPos());
             }
             // CraftBukkit start
