From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 29 Dec 2021 15:12:40 +1000
Subject: [PATCH] Ensure projectileSource is loaded from the ownerUUID


diff --git a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
index 4d8b6de520fb03e68cddb133f8757a27e1099c11..07036d0f568d4baa1ffa36c39a7747cdd16a459d 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
@@ -100,6 +100,8 @@ public abstract class Projectile extends Entity {
                 }
             }
             // Paper end
+            this.projectileSource = null; // MultiPaper - set the projectileSource
+            if (this.cachedOwner != null && !this.cachedOwner.isRemoved()) this.projectileSource = (ProjectileSource) cachedOwner.getBukkitEntity(); // MultiPaper - set the projectileSource
             return this.cachedOwner;
         } else {
             return null;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
index 40ae8e43f40f9bf457d2917ac4f131b21e4f8dd2..f589a4adcfde78579853e8f69348abdc46b3f74d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
@@ -63,6 +63,7 @@ public class CraftArrow extends AbstractProjectile implements AbstractArrow {
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index 2358e0c8a4a135d6deb47100c490a9c145f44827..8730c15698fd7586124f546c3d9ba25c5ce7e12d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -35,6 +35,7 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
index 9a68f4ef68870d0baab5b6464d6c0a82a8fd105d..73dfd74b9793681ecc362ed4d1a589275d191e90 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
@@ -12,6 +12,7 @@ public abstract class CraftProjectile extends AbstractProjectile implements Proj
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
index ca8a9b2773d70a8800b2179b164ce33d7e2bdc5e..58504cac4908c93cd12a945fcd70bfbf2e151370 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
@@ -15,6 +15,7 @@ public class CraftShulkerBullet extends AbstractProjectile implements ShulkerBul
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
