From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Tick attack strength for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 978eaf50764c3d1ca79100c3bdfdf5d9b6600785..2c81dbee8034bf9d708cad29e054262546751023 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -300,6 +300,12 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
         ItemStack itemstack = this.getMainHandItem();
 
@@ -310,6 +316,10 @@ public abstract class Player extends LivingEntity {
 
             this.lastItemInMainHand = itemstack.copy();
         }
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
 
         this.turtleHelmetTick();
         this.cooldowns.tick();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index cdeef29cd561b6962baa1156c8a7b426bd66ce14..cc88b01a9e5ed00ce5e7df7795ec957f16ddbf47 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -114,6 +114,7 @@ public class MultiPaper {
         for (ServerPlayer player : MinecraftServer.getServer().getPlayerList().players) {
             if (player instanceof ExternalPlayer) {
                 player.detectEquipmentUpdates();
+                player.tickAttackStrength();
             }
         }
     }
