From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jan 2022 16:43:52 +1000
Subject: [PATCH] Send server port to the master server


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index ab27b4dc9c790426b7977b20174b7a86eeec4bf0..40a080143f0fe3426a4bfccd4e4cd58f78b3e137 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -267,6 +267,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             this.setPort(dedicatedserverproperties.serverPort);
         }
         bindAddress = new java.net.InetSocketAddress(inetaddress, this.getPort());
+        MultiPaper.setPort(this.getPort()); // MultiPaper
         }
         // Paper end
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 272b8e9a0d56116be4d04c80e2536b430a0a6588..e22e420df7af71edbf95f620381cb1e0a9346ccc 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -591,4 +591,8 @@ public class MultiPaper {
     public static void chunkChangedStatus(ServerLevel level, ChunkPos pos, ChunkStatus status) {
         getConnection().send(new ChunkChangedStatusMessage(level.getWorld().getName(), pos.x, pos.z, BuiltInRegistries.CHUNK_STATUS.getKey(status).toString()));
     }
+
+    public static void setPort(int port) {
+        getConnection().setPort(port);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 6044faafa283d4f50a1700330d1a444f9ba24143..6b53aa706bc00ca63a5ec612c5d24b34446d0a31 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -62,6 +62,7 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
     private final Map<String, ExternalServer> serversMap = new ConcurrentHashMap<>();
     public String secret;
     public Map<String, byte[]> dataCache = Maps.newHashMap();
+    public int port = -1;
 
     public MultiPaperConnection() {
         System.setProperty("multipaper.netty.threads", System.getProperty("multipaper.netty.threads", Integer.toString(Math.min(Runtime.getRuntime().availableProcessors(), 3))));
@@ -99,6 +100,10 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
         LOGGER.info("Connected to " + channel.remoteAddress());
         channel.write(new HelloMessage(myName, thisServersUuid));
 
+        if (port >= 0) {
+            channel.write(new SetPortMessage(port));
+        }
+
         for (MasterBoundMessage unhandledRequest : unhandledRequests) {
             channel.write(unhandledRequest);
         }
@@ -168,6 +173,14 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
         return serversMap;
     }
 
+    public void setPort(int port) {
+        if (!MultiPaperConfiguration.get().masterConnection.advertiseToBuiltInProxy) {
+            return;
+        }
+        this.port = port;
+        send(new SetPortMessage(port));
+    }
+
     @Override
     public void handle(ServerInfoUpdateMessage message) {
         ExternalServer server = serversMap.computeIfAbsent(message.name, key -> new ExternalServer(key, key.equals(myName)));
