From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jan 2022 16:43:52 +1000
Subject: [PATCH] Send server port to the master server


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 4a3c0eb3f194440aa02b380f4aeb60dcdaa47609..8f5c83a1d35986907c910aafb7747b7c1601ecbb 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -256,6 +256,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             this.setPort(dedicatedserverproperties.serverPort);
         }
         bindAddress = new java.net.InetSocketAddress(inetaddress, this.getPort());
+        MultiPaper.setPort(this.getPort()); // MultiPaper
         }
         // Paper end
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index f863560a2671994c166048c2077092b4f34412ec..c966df61713574d68aa4fa02d6e2dd65a249965b 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -585,4 +585,8 @@ public class MultiPaper {
     public static void chunkChangedStatus(ServerLevel level, ChunkPos pos, ChunkStatus status) {
         getConnection().send(new ChunkChangedStatusMessage(level.getWorld().getName(), pos.x, pos.z, BuiltInRegistries.CHUNK_STATUS.getKey(status).toString()));
     }
+
+    public static void setPort(int port) {
+        getConnection().setPort(port);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 527fcdebeeca72b18af59dcc48612ac4b86a0394..c696f2ed90439c607542143fb97ea7d786dade8e 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -59,6 +59,7 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
     private final Map<String, ExternalServer> serversMap = new ConcurrentHashMap<>();
     public String secret;
     public Map<String, byte[]> dataCache = Maps.newHashMap();
+    public int port = -1;
 
     public MultiPaperConnection() {
         System.setProperty("multipaper.netty.threads", System.getProperty("multipaper.netty.threads", Integer.toString(Math.min(Runtime.getRuntime().availableProcessors(), 3))));
@@ -96,6 +97,10 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
         LOGGER.info("Connected to " + channel.remoteAddress());
         channel.write(new HelloMessage(myName, thisServersUuid));
 
+        if (port >= 0) {
+            channel.write(new SetPortMessage(port));
+        }
+
         for (MasterBoundMessage unhandledRequest : unhandledRequests) {
             channel.write(unhandledRequest);
         }
@@ -165,6 +170,14 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
         return serversMap;
     }
 
+    public void setPort(int port) {
+        if (!MultiPaperConfiguration.get().masterConnection.advertiseToBuiltInProxy) {
+            return;
+        }
+        this.port = port;
+        send(new SetPortMessage(port));
+    }
+
     @Override
     public void handle(ServerInfoUpdateMessage message) {
         ExternalServer server = serversMap.computeIfAbsent(message.name, key -> new ExternalServer(key, key.equals(myName)));
