From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Tick attack strength for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 3a35c423ce90173fad1b93bbc04d927d277d8eab..5607b1fc54529c8f0f79dd27e94c3669ad8cd327 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -298,6 +298,12 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
         ItemStack itemstack = this.getMainHandItem();
 
@@ -308,6 +314,10 @@ public abstract class Player extends LivingEntity {
 
             this.lastItemInMainHand = itemstack.copy();
         }
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
 
         this.turtleHelmetTick();
         this.cooldowns.tick();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 9106cc5e1446ca07740c6104f31b0f230c654297..10d18185eb090659326622f918fed1a09075856d 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -91,6 +91,8 @@ public class MultiPaper {
                 // Copied from Paper end
 
                 externalPlayer.checkInsideBlocks();
+
+                externalPlayer.tickAttackStrength();
             }
         }
 
