From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 9 Jan 2022 15:23:24 +1000
Subject: [PATCH] Sync player cameras


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 3901ebd2d8025aa20c757f56ead4caa1133a8814..f5962fba7b9598c42b3d588b3806c3e079066d8d 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2328,6 +2328,8 @@ public class ServerPlayer extends Player {
 
             this.connection.send(new ClientboundSetCameraPacket(this.camera));
             this.connection.resetPosition();
+
+            if (!PlayerSetCameraPacket.handlingSetCamera) MultiPaper.broadcastPacketToExternalServers(this, new PlayerSetCameraPacket(this, entity)); // MultiPaper
         }
     }
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index c98c3235a7f82fafaf084a59f8b4fab62c9ecc91..dda10372b067684167c455b08dbc4ff582b76af0 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -61,6 +61,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerStatsIncreasePacket.class, PlayerStatsIncreasePacket::new);
         addPacket(PlayerExperienceUpdatePacket.class, PlayerExperienceUpdatePacket::new);
         addPacket(PlayerListNameUpdatePacket.class, PlayerListNameUpdatePacket::new);
+        addPacket(PlayerSetCameraPacket.class, PlayerSetCameraPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerSetCameraPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerSetCameraPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..3dd0390908805cb96a59a3ec09eed243df1f08bd
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerSetCameraPacket.java
@@ -0,0 +1,51 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerSetCameraPacket extends ExternalServerPacket {
+
+    public static boolean handlingSetCamera = false;
+
+    private final UUID uuid;
+    private final UUID uuidCamera;
+
+    public PlayerSetCameraPacket(ServerPlayer player, Entity camera) {
+        this.uuid = player.getUUID();
+        this.uuidCamera = camera == null ? null : camera.getUUID();
+    }
+
+    public PlayerSetCameraPacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+        uuidCamera = in.readBoolean() ? in.readUUID() : null;
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeBoolean(uuidCamera != null);
+        if (uuidCamera != null) {
+            out.writeUUID(uuidCamera);
+        }
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            handlingSetCamera = true;
+
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+            Entity entity = ((ServerLevel) player.level()).getEntity(uuidCamera);
+            player.setCamera(entity);
+
+            handlingSetCamera = false;
+        });
+    }
+}
