From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Tick attack strength for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 03bd0c1812ad8ac4eb9c6ec212259eecae1130c9..bb50cb1bb138eac1ade9bef7f2447d3da3033da8 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -299,6 +299,12 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
         ItemStack itemstack = this.getMainHandItem();
 
@@ -309,6 +315,10 @@ public abstract class Player extends LivingEntity {
 
             this.lastItemInMainHand = itemstack.copy();
         }
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
 
         this.turtleHelmetTick();
         this.cooldowns.tick();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 0210ba4be9abf5b4404669baf4fd62ee3ab7b144..96def16d01ce325e5a06d87cbb7c9f729b24ddf3 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -91,6 +91,8 @@ public class MultiPaper {
                 // Copied from Paper end
 
                 externalPlayer.checkInsideBlocks();
+
+                externalPlayer.tickAttackStrength();
             }
         }
 
