From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 9 Jan 2022 15:50:30 +1000
Subject: [PATCH] Add temporary debug log for dispensers


diff --git a/src/main/java/net/minecraft/world/level/block/DispenserBlock.java b/src/main/java/net/minecraft/world/level/block/DispenserBlock.java
index 4720b884e1e311f44a012f1219fe648a411247b4..6c38906e681481ba64ed0b1b6a969cbcc7fb68b4 100644
--- a/src/main/java/net/minecraft/world/level/block/DispenserBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DispenserBlock.java
@@ -78,6 +78,7 @@ public class DispenserBlock extends BaseEntityBlock {
     }
 
     public void dispenseFrom(ServerLevel world, BlockPos pos) {
+        if (pos.getZ() == -43 && pos.getY() == 107 && pos.getX() == -132) new Exception("Dispensing " + pos).printStackTrace();
         BlockSourceImpl sourceblock = new BlockSourceImpl(world, pos);
         DispenserBlockEntity tileentitydispenser = (DispenserBlockEntity) sourceblock.getEntity();
         int i = tileentitydispenser.getRandomSlot(world.random);
diff --git a/src/main/java/net/minecraft/world/level/block/DropperBlock.java b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
index 801c4c120e98584bcf218a4ef9bd66d7d18c1097..aa9140695ef140e872dcf73c5536da52f5fb2aaa 100644
--- a/src/main/java/net/minecraft/world/level/block/DropperBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
@@ -39,6 +39,7 @@ public class DropperBlock extends DispenserBlock {
 
     @Override
     public void dispenseFrom(ServerLevel world, BlockPos pos) {
+        if (pos.getZ() == -43 && pos.getY() == 107 && pos.getX() == -132) new Exception("Dispensing " + pos).printStackTrace();
         BlockSourceImpl sourceblock = new BlockSourceImpl(world, pos);
         DispenserBlockEntity tileentitydispenser = (DispenserBlockEntity) sourceblock.getEntity();
         int i = tileentitydispenser.getRandomSlot(world.random);
