From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Sync attack strength


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index b3d488983a7f3e0a57a717e67a7a8486ca5c69e4..2ff6c8a5c8782e1bee8571181268b3e777f0d922 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -124,6 +124,7 @@ import org.bukkit.event.entity.EntityExhaustionEvent;
 import org.bukkit.event.player.PlayerDropItemEvent;
 import org.bukkit.event.player.PlayerVelocityEvent;
 import puregero.multipaper.MultiPaperEntityInteractHandler;
+import puregero.multipaper.externalserverprotocol.PlayerResetAttackStrengthPacket;
 // CraftBukkit end
 
 public abstract class Player extends LivingEntity {
@@ -302,7 +303,18 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
+
         ItemStack itemstack = this.getMainHandItem();
 
         if (!ItemStack.matches(this.lastItemInMainHand, itemstack)) {
@@ -2229,6 +2241,7 @@ public abstract class Player extends LivingEntity {
     }
 
     public void resetAttackStrengthTicker() {
+        PlayerResetAttackStrengthPacket.broadcastResetAttackStrength(this);
         this.attackStrengthTicker = 0;
     }
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index d86ba0a3b7051ab18c05ec18ff116d59c89fba44..7d2116d1f967cffe73a62725fc8fc8efd4012ec1 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -82,6 +82,8 @@ public class MultiPaper {
                 // Copied from Paper end
 
                 externalPlayer.checkInsideBlocks();
+
+                externalPlayer.tickAttackStrength();
             }
         }
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index aabc8ed6261846cbed62ab68c5d8ab8a135a0fed..019b98edfa8d7f0120099ffc4aad0134cab9094a 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -45,6 +45,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerActionOnEntityPacket.class, PlayerActionOnEntityPacket::new);
         addPacket(PlayerTouchEntityPacket.class, PlayerTouchEntityPacket::new);
         addPacket(HurtEntityPacket.class, HurtEntityPacket::new);
+        addPacket(PlayerResetAttackStrengthPacket.class, PlayerResetAttackStrengthPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerResetAttackStrengthPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerResetAttackStrengthPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..048e25da96045e81871b5024b379d8b10f80efe7
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerResetAttackStrengthPacket.java
@@ -0,0 +1,55 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.player.Player;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerResetAttackStrengthPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(PlayerResetAttackStrengthPacket.class.getSimpleName());
+    private static boolean settingAttackStrength = false;
+
+    private final UUID uuid;
+
+    public PlayerResetAttackStrengthPacket(Player player) {
+        this.uuid = player.getUUID();
+    }
+
+    public PlayerResetAttackStrengthPacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            if (player == null) {
+                LOGGER.warn("Tried to reset attack strength of a non-existent player with uuid " + uuid);
+                return;
+            }
+
+            settingAttackStrength = true;
+            player.resetAttackStrengthTicker();
+            settingAttackStrength = false;
+        });
+    }
+
+    public static void broadcastResetAttackStrength(Player player) {
+        if (!settingAttackStrength && player instanceof ServerPlayer serverPlayer) {
+            MultiPaper.broadcastPacketToExternalServers(serverPlayer, new PlayerResetAttackStrengthPacket(player));
+        }
+    }
+}
