From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 13 Dec 2021 22:55:54 +1000
Subject: [PATCH] Forward sign editing


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index c1a1fd15e206cd9b28d58f7673402f99d31a1d96..ebb7f0bd51c1a79a31b95103c16752162aaf2ed9 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3045,6 +3045,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
     @Override
     public void handleSignUpdate(ServerboundSignUpdatePacket packet) {
+        if (MultipaperSignHandler.handleSignUpdate(this.player, packet)) return;
+
         List<String> list = (List) Stream.of(packet.getLines()).map(ChatFormatting::stripFormatting).collect(Collectors.toList());
 
         this.filterTextPacket(list, (list1) -> {
diff --git a/src/main/java/puregero/multipaper/MultipaperSignHandler.java b/src/main/java/puregero/multipaper/MultipaperSignHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..4520f03cc32ee2eedf47c5e575d64f161fe9daf6
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultipaperSignHandler.java
@@ -0,0 +1,23 @@
+package puregero.multipaper;
+
+import net.minecraft.network.protocol.game.ServerboundSignUpdatePacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.level.chunk.LevelChunk;
+import puregero.multipaper.externalserverprotocol.PlayerActionPacket;
+
+public class MultipaperSignHandler {
+
+    /**
+     * Returns true if the sign update should be cancelled
+     */
+    public static boolean handleSignUpdate(ServerPlayer player, ServerboundSignUpdatePacket packet) {
+        LevelChunk chunk = player.level.getChunkIfLoaded(packet.getPos());
+        if (MultiPaper.isChunkExternal(chunk)) {
+            chunk.externalOwner.getConnection().send(new PlayerActionPacket(player, packet));
+            return true;
+        }
+
+        return false;
+    }
+
+}
