From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 16 Jan 2022 11:51:16 +1000
Subject: [PATCH] Lock the chunk when writting it


diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
index 834bf755736dfd59a93e9c080ac0f1b2a522ef08..6dca0c6ddf452d7b699fc630a1f07ba984d9104e 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
@@ -1906,6 +1906,7 @@ public final class NewChunkHolder {
             if (unloading) {
                 try {
                     chunk.setUnsaved(false); // MultiPaper - set as saved before serializing
+                    MultiPaper.willSaveChunk(this.world, this.chunkX, this.chunkZ); // MultiPaper
                     final ChunkSerializer.AsyncSaveData asyncSaveData = ChunkSerializer.getAsyncSaveData(this.world, chunk);
 
                     final PrioritisedExecutor.PrioritisedTask task = this.scheduler.loadExecutor.createTask(new AsyncChunkSerializeTask(this.world, chunk, asyncSaveData, this));
@@ -1930,6 +1931,7 @@ public final class NewChunkHolder {
 
             if (unloading) {
                 completing = true;
+                MultiPaper.willSaveChunk(this.world, this.chunkX, this.chunkZ); // MultiPaper
                 this.completeAsyncChunkDataSave(save);
                 LOGGER.info("Successfully serialized chunk data (" + this.chunkX + "," + this.chunkZ + ") in world '" + this.world.getWorld().getName() + "' synchronously");
             } else {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index e22e420df7af71edbf95f620381cb1e0a9346ccc..c0e2975fde7de0c1cf0890f63748a57138f5d35e 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -511,6 +511,10 @@ public class MultiPaper {
         newChunkHolder.hasExternalLockRequest = false;
     }
 
+    public static void willSaveChunk(ServerLevel level, int x, int z) {
+        getConnection().send(new WillSaveChunkLaterMessage(level.getWorld().getName(), x, z));
+    }
+
     public static byte[] nbtToBytes(CompoundTag compoundTag) throws IOException {
         if (compoundTag == null) {
             return new byte[0];
