From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 5 Dec 2021 12:48:03 +1000
Subject: [PATCH] Don't send the horse armour equip sound when loading a horse


diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java b/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java
index 34299274e7dd57076a1043d735a41b9c34f0a349..3730f9f61136abda0fbb8414514ed1d977d9bdfa 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Horse.java
@@ -36,6 +36,8 @@ public class Horse extends AbstractHorse {
     private static final UUID ARMOR_MODIFIER_UUID = UUID.fromString("556E1665-8B10-40C8-8F9D-CF9B1667F295");
     private static final EntityDataAccessor<Integer> DATA_ID_TYPE_VARIANT = SynchedEntityData.defineId(Horse.class, EntityDataSerializers.INT);
 
+    private boolean loadingArmor = false; // MultiPaper
+
     public Horse(EntityType<? extends Horse> type, Level world) {
         super(type, world);
     }
@@ -74,6 +76,7 @@ public class Horse extends AbstractHorse {
 
     @Override
     public void readAdditionalSaveData(CompoundTag nbt) {
+        loadingArmor = true; // MultiPaper
         super.readAdditionalSaveData(nbt);
         this.setTypeVariant(nbt.getInt("Variant"));
         if (nbt.contains("ArmorItem", 10)) {
@@ -84,6 +87,7 @@ public class Horse extends AbstractHorse {
         }
 
         this.updateContainerEquipment();
+        loadingArmor = false; // MultiPaper
     }
 
     private void setTypeVariant(int variant) {
@@ -134,7 +138,7 @@ public class Horse extends AbstractHorse {
         ItemStack itemStack = this.getArmor();
         super.containerChanged(sender);
         ItemStack itemStack2 = this.getArmor();
-        if (this.tickCount > 20 && this.isArmor(itemStack2) && itemStack != itemStack2) {
+        if (this.tickCount > 20 && this.isArmor(itemStack2) && itemStack != itemStack2 && !loadingArmor) {
             this.playSound(SoundEvents.HORSE_ARMOR, 0.5F, 1.0F);
         }
 
