From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 23 Jan 2022 00:07:41 +1000
Subject: [PATCH] Only play beacon deactivate sound if we aren't unloading the
 chunk


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index 2979c1eceaac79a82da19eba5d8fb1db6537ff45..8f57bf34c05e600fec7f4714cdccd96332af0ccd 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -298,6 +298,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
         // Paper end
         // Paper start - fix MC-153086
         if (this.levels > 0 && !this.beamSections.isEmpty()) {
+        if (this.level.getChunkIfLoaded(getBlockPos()) != null && this.level.getChunkIfLoaded(getBlockPos()).loaded) // MultiPaper - Only play beacon deactivate sound if we aren't unloading the chunk
         BeaconBlockEntity.playSound(this.level, this.worldPosition, SoundEvents.BEACON_DEACTIVATE);
         }
         // Paper end
