From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 4 Dec 2021 22:39:00 +1000
Subject: [PATCH] Sync mob navigation when handing over to another server


diff --git a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 3f672d7c2377fca16a6d8d31cf7aaae4f009fdce..9b35202584ae090cb9f64c2bd3d0a07ffd8621c2 100644
--- a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -33,7 +33,7 @@ public abstract class PathNavigation {
     protected final Level level;
     @Nullable
     protected Path path;
-    protected double speedModifier;
+    public double speedModifier; // MultiPaper - make public
     protected int tick;
     protected int lastStuckCheck;
     protected Vec3 lastStuckCheckPos = Vec3.ZERO;
diff --git a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
index c2c9e05f2d7a675582d809d9cae0c3617b2d7510..5d630422bc63bafef8c01d4bec14196b7defeda2 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
@@ -130,6 +130,14 @@ public class MultiPaperEntitiesHandler {
                 // Leaving our jurisdiction, do a full entity update to ensure the new external server has all the required info
                 if (!(entity instanceof ServerPlayer)) { // Ignore players as they aren't ticked by the new external server
                     MultiPaper.runSync(() -> MultiPaper.broadcastPacketToExternalServers(chunkTo.externalEntitiesSubscribers, () -> new EntityUpdateNBTPacket(entity)));
+                    if (entity instanceof Mob mob) {
+                        MultiPaper.runSync(() -> {
+                            BlockPos goal = mob.getNavigation().getTargetPos();
+                            if (goal != null) {
+                                MultiPaper.broadcastPacketToExternalServers(chunkTo.externalEntitiesSubscribers, () -> new MobSetNavigationGoalPacket(mob, goal));
+                            }
+                        });
+                    }
                 }
             }
             for (ExternalServer fromServer : chunkFrom.externalEntitiesSubscribers) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index a8d65f014702b229e9500f881231f71ff761ee3f..7dd289fd1155bfe6f519fd49d142ac7e1bf1c0a7 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -49,6 +49,7 @@ public class ExternalServerPacketSerializer {
         addPacket(AddItemToEntityContainerPacket.class, AddItemToEntityContainerPacket::new);
         addPacket(AddItemToContainerPacket.class, AddItemToContainerPacket::new);
         addPacket(PullItemFromContainerPacket.class, PullItemFromContainerPacket::new);
+        addPacket(MobSetNavigationGoalPacket.class, MobSetNavigationGoalPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/MobSetNavigationGoalPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/MobSetNavigationGoalPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..1369ebb95e38919acaaefa64bf7f9faaa7e9fe13
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/MobSetNavigationGoalPacket.java
@@ -0,0 +1,63 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.ai.goal.WrappedGoal;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class MobSetNavigationGoalPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(MobSetNavigationGoalPacket.class.getSimpleName());
+
+    private final String world;
+    private final UUID uuid;
+    private final BlockPos goal;
+    private final double speed;
+
+    public MobSetNavigationGoalPacket(Mob mob, BlockPos goal) {
+        this.world = ((ServerLevel) mob.level).convertable.getLevelId();
+        this.uuid = mob.getUUID();
+        this.goal = goal;
+        this.speed = mob.getNavigation().speedModifier;
+    }
+
+    public MobSetNavigationGoalPacket(FriendlyByteBuf in) {
+        this.world = in.readUtf();
+        this.uuid = in.readUUID();
+        this.goal = BlockPos.of(in.readLong());
+        this.speed = in.readDouble();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+        out.writeUUID(uuid);
+        out.writeLong(goal.asLong());
+        out.writeDouble(speed);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerLevel level = ((CraftWorld) Bukkit.getWorld(world)).getHandle();
+            Entity entity = level.getEntity(uuid);
+            if (entity instanceof Mob mob) {
+                mob.goalSelector.getRunningGoals().forEach(WrappedGoal::stop);
+                mob.targetSelector.getRunningGoals().forEach(WrappedGoal::stop);
+                mob.getNavigation().moveTo(mob.getNavigation().createPath(goal, 0), speed);
+            } else {
+                LOGGER.warn("Couldn't find mob " + uuid + " for navigation goal");
+            }
+        });
+    }
+}
