From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 11 Dec 2021 22:31:23 +1000
Subject: [PATCH] Sync potion effects


diff --git a/src/main/java/net/minecraft/world/effect/MobEffectInstance.java b/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
index 0dadea0c9559d99c7de04dbda68b3e743c9eeecb..8d918a9f867084d24944cee2f14ba63a8f6c619a 100644
--- a/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
+++ b/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
@@ -78,7 +78,7 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
             this.amplifier = that.amplifier;
             this.duration = that.duration;
             bl = true;
-        } else if (that.duration > this.duration) {
+        } else if (that.duration >= this.duration) { // MultiPaper - Override the effect even if they're the same duration
             if (that.amplifier == this.amplifier) {
                 this.duration = that.duration;
                 bl = true;
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 01287b9aa12f8e08065214e501b53f443cc0d8de..57fa2df605c780973f5bc0aa26f2d11f1742f63a 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -139,6 +139,7 @@ import org.bukkit.event.entity.EntityTeleportEvent;
 import org.bukkit.event.player.PlayerItemConsumeEvent;
 import puregero.multipaper.ExternalServerConnection;
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperEffectsHandler;
 import puregero.multipaper.MultiPaperEntityInteractHandler;
 import puregero.multipaper.externalserverprotocol.HurtEntityPacket;
 // CraftBukkit end
@@ -1198,29 +1199,32 @@ public abstract class LivingEntity extends Entity {
 
     protected void onEffectAdded(MobEffectInstance effect, @Nullable Entity source) {
         this.effectsDirty = true;
-        if (!this.level.isClientSide) {
+        if (!this.level.isClientSide && !MultiPaper.isExternalPlayer(this)) { // MultiPaper
             effect.getEffect().addAttributeModifiers(this, this.getAttributes(), effect.getAmplifier());
         }
 
+        MultiPaperEffectsHandler.onEffectAdd(this, effect); // MultiPaper
     }
 
     protected void onEffectUpdated(MobEffectInstance effect, boolean reapplyEffect, @Nullable Entity source) {
         this.effectsDirty = true;
-        if (reapplyEffect && !this.level.isClientSide) {
+        if (reapplyEffect && !MultiPaper.isExternalPlayer(this)) { // MultiPaper
             MobEffect mobeffectlist = effect.getEffect();
 
             mobeffectlist.removeAttributeModifiers(this, this.getAttributes(), effect.getAmplifier());
             mobeffectlist.addAttributeModifiers(this, this.getAttributes(), effect.getAmplifier());
         }
 
+        MultiPaperEffectsHandler.onEffectAdd(this, effect); // MultiPaper
     }
 
     protected void onEffectRemoved(MobEffectInstance effect) {
         this.effectsDirty = true;
-        if (!this.level.isClientSide) {
+        if (!this.level.isClientSide && !MultiPaper.isExternalPlayer(this)) { // MultiPaper
             effect.getEffect().removeAttributeModifiers(this, this.getAttributes(), effect.getAmplifier());
         }
 
+        MultiPaperEffectsHandler.onEffectRemove(this, effect); // MultiPaper
     }
 
     // CraftBukkit start - Delegate so we can handle providing a reason for health being regained
diff --git a/src/main/java/puregero/multipaper/MultiPaperEffectsHandler.java b/src/main/java/puregero/multipaper/MultiPaperEffectsHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..d0aacf35c7d10113dca9f37dd2c9c71a593f0eac
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperEffectsHandler.java
@@ -0,0 +1,54 @@
+package puregero.multipaper;
+
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.effect.MobEffectInstance;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import puregero.multipaper.externalserverprotocol.EntityUpdateEffectPacket;
+import puregero.multipaper.externalserverprotocol.ExternalServerPacket;
+
+import java.util.function.Supplier;
+
+public class MultiPaperEffectsHandler {
+    private static boolean updatingEffects = false;
+
+    public static void onEffectAdd(LivingEntity entity, MobEffectInstance effect) {
+        if (updatingEffects) {
+            return;
+        }
+
+        broadcast(entity, () -> new EntityUpdateEffectPacket(entity, effect, false));
+    }
+
+    public static void onEffectRemove(LivingEntity entity, MobEffectInstance effect) {
+        if (updatingEffects) {
+            return;
+        }
+
+        broadcast(entity, () -> new EntityUpdateEffectPacket(entity, effect, true));
+
+
+    }
+
+    private static void broadcast(LivingEntity entity, Supplier<ExternalServerPacket> packetSupplier) {
+        if (entity instanceof ServerPlayer) {
+            MultiPaper.broadcastPacketToExternalServers(packetSupplier.get());
+        } else {
+            LevelChunk chunk = entity.level.getChunkIfLoaded(entity.blockPosition());
+            if (chunk != null) {
+                MultiPaper.broadcastPacketToExternalServers(chunk.externalEntitiesSubscribers, packetSupplier);
+            }
+        }
+    }
+
+    public static void handle(Entity entity, MobEffectInstance effect, boolean remove) {
+        updatingEffects = true;
+        if (remove) {
+            ((LivingEntity) entity).removeEffect(effect.getEffect());
+        } else {
+            ((LivingEntity) entity).addEffect(effect);
+        }
+        updatingEffects = false;
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/EntityUpdateEffectPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/EntityUpdateEffectPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..38c8469e96a680b74a36c096bdbbfd1ffce75ac8
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/EntityUpdateEffectPacket.java
@@ -0,0 +1,69 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.effect.MobEffectInstance;
+import net.minecraft.world.entity.Entity;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperEffectsHandler;
+
+import java.io.IOException;
+import java.util.UUID;
+
+public class EntityUpdateEffectPacket extends ExternalServerPacket {
+
+    private final String world;
+    private final UUID uuid;
+    private final boolean remove;
+    private final byte[] data;
+
+    public EntityUpdateEffectPacket(Entity entity, MobEffectInstance effect, boolean remove) {
+        this.world = ((ServerLevel) entity.level).convertable.getLevelId();
+        this.uuid = entity.getUUID();
+        this.remove = remove;
+
+        CompoundTag tag = effect.save(new CompoundTag());
+
+        try {
+            this.data = MultiPaper.nbtToBytes(tag);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public EntityUpdateEffectPacket(FriendlyByteBuf in) {
+        this.world = in.readUtf();
+        this.uuid = in.readUUID();
+        this.remove = in.readBoolean();
+        this.data = in.readByteArray();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+        out.writeUUID(uuid);
+        out.writeBoolean(remove);
+        out.writeByteArray(data);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            try {
+                ServerLevel level = ((CraftWorld) Bukkit.getWorld(world)).getHandle();
+                Entity entity = level.getEntity(uuid);
+                CompoundTag tag = MultiPaper.nbtFromBytes(data);
+                MobEffectInstance effect = MobEffectInstance.load(tag);
+                if (entity != null) {
+                    MultiPaperEffectsHandler.handle(entity, effect, remove);
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        });
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index b7bf24fdbd8f5a6529a05017f6140005d5673708..561793810a28e46d5516f9fa6e9c60450a5437b0 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -50,6 +50,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerDataUpdatePacket.class, PlayerDataUpdatePacket::new);
         addPacket(PluginNotificationPacket.class, PluginNotificationPacket::new);
         addPacket(PlayerFoodUpdatePacket.class, PlayerFoodUpdatePacket::new);
+        addPacket(EntityUpdateEffectPacket.class, EntityUpdateEffectPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
