From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 31 Jan 2022 13:35:42 +1000
Subject: [PATCH] Send crit animations as the entity being critted


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index cd82d82b995b1ffa7ac028f4c16967b73e42adcf..36444968551b569bed20ef25606ce27d12731bde 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2031,12 +2031,12 @@ public class ServerPlayer extends Player {
 
     @Override
     public void crit(Entity target) {
-        this.serverLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(target, 4));
+        this.serverLevel().getChunkSource().broadcastAndSend(target, new ClientboundAnimatePacket(target, 4)); // MultiPaper
     }
 
     @Override
     public void magicCrit(Entity target) {
-        this.serverLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(target, 5));
+        this.serverLevel().getChunkSource().broadcastAndSend(target, new ClientboundAnimatePacket(target, 5)); // MultiPaper
     }
 
     @Override
