From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 1 Feb 2022 00:41:27 +1000
Subject: [PATCH] Sync beacons


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index b6e8ae36348214be5a11156ee83328abf97a0caf..45e2cf64c727af2a0e040dd5c7ff5ecdf79ea296 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -151,6 +151,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
                         BeaconBlockEntity.this.secondaryPower = BeaconBlockEntity.getValidEffectById(value);
                 }
 
+                BeaconBlockEntity.this.setChanged(); // MultiPaper - send this block entity to other server
             }
 
             @Override
