From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 16:12:51 +1000
Subject: [PATCH] Tick takexpdelay on external players


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 5694c9f9961f86fdb5274aa4912404e064adf89b..bc2d66a663a99590d3c780323f177271103e4173 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -201,6 +201,10 @@ public class MultiPaper {
             if (player instanceof ExternalPlayer) {
                 player.detectEquipmentUpdates();
                 player.tickAttackStrength();
+
+                if (player.takeXpDelay > 0) {
+                    --player.takeXpDelay;
+                }
             }
 
             player.syncExperience();
