From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 16:12:51 +1000
Subject: [PATCH] Tick takexpdelay on external players


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index c8293611cf859add77cbaae02c88822c9f22548c..05e9befc73ffec7a224be532f546bde8cc88529b 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -199,6 +199,10 @@ public class MultiPaper {
             if (player instanceof ExternalPlayer) {
                 player.detectEquipmentUpdates();
                 player.tickAttackStrength();
+
+                if (player.takeXpDelay > 0) {
+                    --player.takeXpDelay;
+                }
             }
 
             player.syncExperience();
