From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 14 Dec 2021 13:41:18 +1000
Subject: [PATCH] Use target UUID when saving conduits


diff --git a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
index b7ec481bb9159bb203ff1688dd7f6b9f9199bccc..f51e51c2cf5246c3a73b12261d4e82486e1ea5f6 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -70,6 +70,7 @@ public class ConduitBlockEntity extends BlockEntity {
     @Override
     protected void saveAdditional(CompoundTag nbt) {
         super.saveAdditional(nbt);
+        if (this.destroyTargetUUID != null) nbt.putUUID("Target", this.destroyTargetUUID); // MultiPaper - Save UUID
         if (this.destroyTarget != null) {
             nbt.putUUID("Target", this.destroyTarget.getUUID());
         }
