From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 17 Dec 2021 20:07:06 +1000
Subject: [PATCH] Ensure Villager getOffers is run on the main thread


diff --git a/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java b/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
index 5eab7d50734551d96128dfebee126a1da4c51375..a98164a8963e67b163e28264ca69a7bf2476a44f 100644
--- a/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -10,6 +10,7 @@ import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.MCUtil;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
@@ -40,6 +41,7 @@ import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.inventory.CraftMerchant;
 import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe;
 import org.bukkit.event.entity.VillagerAcquireTradeEvent;
+import org.spigotmc.AsyncCatcher;
 // CraftBukkit end
 
 public abstract class AbstractVillager extends AgeableMob implements InventoryCarrier, Npc, Merchant {
@@ -125,6 +127,7 @@ public abstract class AbstractVillager extends AgeableMob implements InventoryCa
     @Override
     public MerchantOffers getOffers() {
         if (this.offers == null) {
+            if (!MCUtil.isMainThread()) return new MerchantOffers(); // MultiPaper - don't calculate trade offers async
             this.offers = new MerchantOffers();
             this.updateTrades();
         }
