From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 21 Dec 2021 00:37:18 +1000
Subject: [PATCH] Sync experience


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 6f703df89d4873a05e4ccf588775bf4149b3e7bd..5b09178d36ed74bc00a2c20fdf17c03447a70f78 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -165,10 +165,7 @@ import org.bukkit.event.player.PlayerTeleportEvent;
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.inventory.MainHand;
 import puregero.multipaper.*;
-import puregero.multipaper.externalserverprotocol.PlayerChangeDimensionPacket;
-import puregero.multipaper.externalserverprotocol.PlayerChangeGamemodePacket;
-import puregero.multipaper.externalserverprotocol.PlayerCreatePacket;
-import puregero.multipaper.externalserverprotocol.PlayerRemovePacket;
+import puregero.multipaper.externalserverprotocol.*;
 // CraftBukkit end
 
 public class ServerPlayer extends Player {
@@ -592,6 +589,20 @@ public class ServerPlayer extends Player {
         this.lastSentExp = -1;
     }
 
+    // MultiPaper start - sync xp
+    public float lastExperienceProgress = 0;
+    public int lastTotalExperience = 0;
+    public int lastExperienceLevel = 0;
+    public void syncExperience() {
+        if (experienceProgress != lastExperienceProgress || totalExperience != lastTotalExperience || experienceLevel != lastExperienceLevel) {
+            lastExperienceProgress = experienceProgress;
+            lastTotalExperience = totalExperience;
+            lastExperienceLevel = experienceLevel;
+            MultiPaper.broadcastPacketToExternalServers(new PlayerExperienceUpdatePacket(this, experienceProgress, totalExperience, experienceLevel));
+        }
+    }
+    // MultiPaper end
+
     public void initMenu(AbstractContainerMenu screenHandler) {
         screenHandler.addSlotListener(this.containerListener);
         screenHandler.setSynchronizer(this.containerSynchronizer);
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 926d4248e0bdf6ed71defd829a0a3158adfff4b3..4ce32c8abf43e3c0bc93f4b354bfed90a4720f87 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -205,6 +205,8 @@ public class MultiPaper {
                 player.detectEquipmentUpdates();
                 player.tickAttackStrength();
             }
+
+            player.syncExperience();
         }
 
         MultiPaperExternalBlocksHandler.tick();
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 42bc717b1248a0841ef79b62b1e2b34bbb296c67..0b33cab2e9c8487f0aeddf2da28c527118b845ea 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -47,6 +47,7 @@ public class ExternalServerPacketSerializer {
         addPacket(AdvancementGrantProgressPacket.class, AdvancementGrantProgressPacket::new);
         addPacket(WeatherUpdatePacket.class, WeatherUpdatePacket::new);
         addPacket(PlayerStatsIncreasePacket.class, PlayerStatsIncreasePacket::new);
+        addPacket(PlayerExperienceUpdatePacket.class, PlayerExperienceUpdatePacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerExperienceUpdatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerExperienceUpdatePacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..e558a400216b4386d9f3dbeaeb9d9d25bf7645ec
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerExperienceUpdatePacket.java
@@ -0,0 +1,49 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerExperienceUpdatePacket extends ExternalServerPacket {
+    private final UUID uuid;
+    private final float progress;
+    private final int total;
+    private final int level;
+
+    public PlayerExperienceUpdatePacket(ServerPlayer player, float progress, int total, int level) {
+        this.uuid = player.getUUID();
+        this.progress = progress;
+        this.total = total;
+        this.level = level;
+    }
+
+    public PlayerExperienceUpdatePacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+        progress = in.readFloat();
+        total = in.readInt();
+        level = in.readInt();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeFloat(progress);
+        out.writeInt(total);
+        out.writeInt(level);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            player.experienceProgress = player.lastExperienceProgress = progress;
+            player.totalExperience = player.lastTotalExperience = total;
+            player.experienceLevel = player.lastExperienceLevel = level;
+        });
+    }
+}
