From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 24 Dec 2021 13:40:24 +1000
Subject: [PATCH] Don't let a player login if they're on an external server


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 530bc07a42a2b57f0cc141ef5302d15fe7930cf6..edc79fe07f340cbe57b96f6789fc49cb81eb779c 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -708,6 +708,12 @@ public abstract class PlayerList {
         for (int i = 0; i < this.players.size(); ++i) {
             entityplayer = (ServerPlayer) this.players.get(i);
             if (entityplayer.getUUID().equals(uuid)) {
+                // MultiPaper start - don't let a player login if they're on an external server
+                if (MultiPaper.isExternalPlayer(entityplayer)) {
+                    loginlistener.disconnect(PaperAdventure.asVanilla(net.kyori.adventure.text.Component.text("You are already logged in.")));
+                    return null;
+                }
+                // MultiPaper end - don't let a player login if they're on an external server
                 list.add(entityplayer);
             }
         }
