From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 24 Dec 2021 13:40:24 +1000
Subject: [PATCH] Don't let a player login if they're on an external server


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index e995bd07e9e19cbe7105411a860be844c4d208e0..f00c6b0e3776cd3f9a40e97339b2c23ce7774513 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -708,6 +708,12 @@ public abstract class PlayerList {
         for (int i = 0; i < this.players.size(); ++i) {
             entityplayer = (ServerPlayer) this.players.get(i);
             if (entityplayer.getUUID().equals(uuid) || (com.destroystokyo.paper.PaperConfig.isProxyOnlineMode() && entityplayer.getGameProfile().getName().equalsIgnoreCase(gameprofile.getName()))) { // Paper - validate usernames
+                // MultiPaper start - don't let a player login if they're on an external server
+                if (MultiPaper.isExternalPlayer(entityplayer)) {
+                    loginlistener.disconnect(PaperAdventure.asVanilla(net.kyori.adventure.text.Component.text("You are already logged in.")));
+                    return null;
+                }
+                // MultiPaper end - don't let a player login if they're on an external server
                 list.add(entityplayer);
             }
         }
