From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 16:12:51 +1000
Subject: [PATCH] Tick takexpdelay on external players


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 38e06bdf1bbe477cd2846b4c4cb21818d6b4ab5c..2a7f37e9b53ad79e498baec9196c65be10d257d2 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -154,6 +154,10 @@ public class MultiPaper {
                 externalPlayer.checkInsideBlocks();
 
                 externalPlayer.tickAttackStrength();
+
+                if (player.takeXpDelay > 0) {
+                    --player.takeXpDelay;
+                }
             }
 
             player.syncExperience();
