From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 16:12:51 +1000
Subject: [PATCH] Tick takexpdelay on external players


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index b8cfe12f642ff352f9f8618efa826cfdb75c08ab..fc2dbda71ee71aed9c131a4eef05924573195c96 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -205,6 +205,10 @@ public class MultiPaper {
             if (player instanceof ExternalPlayer) {
                 player.detectEquipmentUpdates();
                 player.tickAttackStrength();
+
+                if (player.takeXpDelay > 0) {
+                    --player.takeXpDelay;
+                }
             }
 
             player.syncExperience();
