From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 27 Feb 2022 11:24:14 +1000
Subject: [PATCH] Update external player spam counters


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 953cb1cc7d541e5b5c38a6fe3ff092622172474a..bbf83f1bd3f156a8816c5c90c98a8006871f3105 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -458,20 +458,23 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         // Paper end
 
         //this.server.getProfiler().pop(); // Purpur
-        // CraftBukkit start
-        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
-        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
-        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
-        /* Use thread-safe field access instead
-        if (this.chatSpamTickCount > 0) {
-            --this.chatSpamTickCount;
-        }
-        */
-        // CraftBukkit end
-
-        if (this.dropSpamTickCount > 0) {
-            --this.dropSpamTickCount;
-        }
+// MultiPaper start - move to reduceSpamCounters()
+//        // CraftBukkit start
+//        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
+//        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
+//        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
+//        /* Use thread-safe field access instead
+//        if (this.chatSpamTickCount > 0) {
+//            --this.chatSpamTickCount;
+//        }
+//        */
+//        // CraftBukkit end
+//
+//        if (this.dropSpamTickCount > 0) {
+//            --this.dropSpamTickCount;
+//        }
+        reduceSpamCounters();
+// MultiPaper end
 
         if (this.player.getLastActionTime() > 0L && this.server.getPlayerIdleTimeout() > 0 && Util.getMillis() - this.player.getLastActionTime() > (long) this.server.getPlayerIdleTimeout() * 1000L * 60L && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits.
             // Purpur start
@@ -502,6 +505,25 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         this.lastGoodZ = this.player.getZ();
     }
 
+    // MultiPaper start
+    public void reduceSpamCounters() {
+        // CraftBukkit start
+        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
+        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
+        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
+        /* Use thread-safe field access instead
+        if (this.chatSpamTickCount > 0) {
+            --this.chatSpamTickCount;
+        }
+        */
+        // CraftBukkit end
+
+        if (this.dropSpamTickCount > 0) {
+            --this.dropSpamTickCount;
+        }
+    }
+    // MultiPaper end
+
     @Override
     public boolean isAcceptingMessages() {
         return this.connection.isConnected();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index b22af9eca0febb6e13c64ea2ae8fd49b814ef9c1..12e42d35111bef768c42e6a9d2a520a142c8bea0 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -98,6 +98,8 @@ public class MultiPaper {
             }
 
             player.syncExperience();
+
+            player.connection.reduceSpamCounters();
         }
 
         for (ExternalServer server : getConnection().getServersMap().values()) {
