From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 23:10:18 +1000
Subject: [PATCH] Don't suck from external minecarts


diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 47d1dd59c81a95e83c1e7801e4aae72f397a3c7c..52d16628e10f2f436a38aebae2a97df108273ed9 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -38,6 +38,7 @@ import org.bukkit.entity.HumanEntity;
 import org.bukkit.event.inventory.InventoryMoveItemEvent;
 import org.bukkit.event.inventory.InventoryPickupItemEvent;
 import org.bukkit.inventory.Inventory;
+import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperContainerHandler;
 // CraftBukkit end
 
@@ -483,7 +484,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public static boolean suckInItems(Level world, Hopper hopper) {
         Container iinventory = HopperBlockEntity.getSourceContainer(world, hopper);
 
-        if (iinventory != null) {
+        if (iinventory != null && !(iinventory instanceof Entity entity && MultiPaper.isChunkExternal(entity.level.getChunkIfLoaded(entity.blockPosition())))) { // MultiPaper - don't suck from external minecarts
             Direction enumdirection = Direction.DOWN;
 
             // Paper start - optimize hoppers and remove streams
