From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 16:12:51 +1000
Subject: [PATCH] Tick takexpdelay on external players


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 92ebdedae7601cc6d324d606cbc961da2303e71d..7917dd398db79926570885d92bd8503dd2a163c9 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -94,6 +94,10 @@ public class MultiPaper {
                 externalPlayer.checkInsideBlocks();
 
                 externalPlayer.tickAttackStrength();
+
+                if (player.takeXpDelay > 0) {
+                    --player.takeXpDelay;
+                }
             }
 
             player.syncExperience();
