From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 17:48:55 +1000
Subject: [PATCH] Load chunks when entities go through an end portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0667c5f9f943da1c29cdbd760ed35577981d1e53..57b9ea15a3d80eabef3f033c758aaec916922969 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3407,6 +3407,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 Entity entity = this.getType().create(worldserver);
 
                 if (entity != null) {
+                    // MultiPaper start - Load chunk fully for portals
+                    BlockPos pos = new BlockPos(position);
+                    worldserver.getChunkSource().addRegionTicket(TicketType.PORTAL, new ChunkPos(pos), 1, pos);
+                    worldserver.getChunkAt(pos);
+                    // MultiPaper end - Load chunk fully for portals
                     entity.restoreFrom(this);
                     entity.moveTo(position.x, position.y, position.z, yaw, pitch); // Paper - use EntityPortalExitEvent values
                     entity.setDeltaMovement(velocity); // Paper - use EntityPortalExitEvent values
