From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 23:10:18 +1000
Subject: [PATCH] Don't suck from external minecarts


diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 9de538478654a3e80133d4e01efc5c683ffc9562..04a0730373bce15c7fad32606de7e0884609cab7 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -37,6 +37,7 @@ import org.bukkit.entity.HumanEntity;
 import org.bukkit.event.inventory.InventoryMoveItemEvent;
 import org.bukkit.event.inventory.InventoryPickupItemEvent;
 import org.bukkit.inventory.Inventory;
+import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperContainerHandler;
 // CraftBukkit end
 
@@ -482,7 +483,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public static boolean suckInItems(Level world, Hopper hopper) {
         Container iinventory = HopperBlockEntity.getSourceContainer(world, hopper);
 
-        if (iinventory != null) {
+        if (iinventory != null && !(iinventory instanceof Entity entity && MultiPaper.isChunkExternal(entity.level.getChunkIfLoaded(entity.blockPosition())))) { // MultiPaper - don't suck from external minecarts
             Direction enumdirection = Direction.DOWN;
 
             // Paper start - optimize hoppers and remove streams
