From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 29 Dec 2021 15:12:40 +1000
Subject: [PATCH] Ensure projectileSource is loaded from the ownerUUID


diff --git a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
index 7bea03806f9c339392b0a97628addf64684a8788..43ff5a69fd1ec40ecd6b0dba833ede6a3012cf3e 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
@@ -99,6 +99,7 @@ public abstract class Projectile extends Entity {
                 }
             }
             // Paper end
+            if (!this.cachedOwner.isRemoved()) setOwner(this.cachedOwner); // MultiPaper - ensure projectileSource is updated
             return this.cachedOwner;
         } else {
             return null;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
index 454c8fab2f0b60aa3afd73805ea3586881605450..69b173c771b916613b7425abde3f063e9ea98603 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
@@ -63,6 +63,7 @@ public class CraftArrow extends AbstractProjectile implements AbstractArrow {
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index 2358e0c8a4a135d6deb47100c490a9c145f44827..8730c15698fd7586124f546c3d9ba25c5ce7e12d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -35,6 +35,7 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
index 9a68f4ef68870d0baab5b6464d6c0a82a8fd105d..73dfd74b9793681ecc362ed4d1a589275d191e90 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
@@ -12,6 +12,7 @@ public abstract class CraftProjectile extends AbstractProjectile implements Proj
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
index 20b54f8896be1f8744a29e1d0205e58d27049f1f..9556f22400dc96c4cc61f6217d9bda32238be224 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
@@ -15,6 +15,7 @@ public class CraftShulkerBullet extends AbstractProjectile implements ShulkerBul
 
     @Override
     public ProjectileSource getShooter() {
+        this.getHandle().getOwner(); // MultiPaper - ensure owner is loaded
         return this.getHandle().projectileSource;
     }
 
