From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 17:48:55 +1000
Subject: [PATCH] Load chunks when entities go through an end portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index b54c4c2fbb831a9f04e2c6b53a5380e401a3e2d1..64c1317961cd866b64b38e2332d2ef66f000dc26 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3525,6 +3525,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 Entity entity = this.getType().create(worldserver);
 
                 if (entity != null) {
+                    // MultiPaper start - Load chunk fully for portals
+                    BlockPos pos = new BlockPos(position);
+                    worldserver.getChunkSource().addRegionTicket(TicketType.PORTAL, new ChunkPos(pos), 1, pos);
+                    worldserver.getChunkAt(pos);
+                    // MultiPaper end - Load chunk fully for portals
                     entity.restoreFrom(this);
                     entity.moveTo(position.x, position.y, position.z, yaw, pitch); // Paper - use EntityPortalExitEvent values
                     entity.setDeltaMovement(velocity); // Paper - use EntityPortalExitEvent values
