From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 17:48:55 +1000
Subject: [PATCH] Load chunks when entities go through an end portal


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index ce8c95fc06c89c58bedf627b47d09a301de69e62..abc6b51c5f85f635427017eac7001af116afbeb7 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3536,6 +3536,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 Entity entity = this.getType().create(worldserver);
 
                 if (entity != null) {
+                    // MultiPaper start - Load chunk fully for portals
+                    BlockPos pos = new BlockPos(position);
+                    worldserver.getChunkSource().addRegionTicket(TicketType.PORTAL, new ChunkPos(pos), 1, pos);
+                    worldserver.getChunkAt(pos);
+                    // MultiPaper end - Load chunk fully for portals
                     entity.restoreFrom(this);
                     entity.moveTo(position.x, position.y, position.z, yaw, pitch); // Paper - use EntityPortalExitEvent values
                     entity.setDeltaMovement(velocity); // Paper - use EntityPortalExitEvent values
