From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 27 Feb 2022 19:21:10 +1000
Subject: [PATCH] Don't double save the json lists


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
index 2402b0e9931ee7026c22d872c675c05e3e53f961..1f98c57072bd2f86c7f085bcc2740d7f98a09f37 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -47,13 +47,13 @@ public class DedicatedPlayerList extends PlayerList {
     @Override
     public void op(GameProfile profile) {
         super.op(profile);
-        this.saveOps();
+        // this.saveOps(); // MultiPaper - don't double save the op list
     }
 
     @Override
     public void deop(GameProfile profile) {
         super.deop(profile);
-        this.saveOps();
+        // this.saveOps(); // MultiPaper - don't double save the op list
     }
 
     @Override
