From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 23 Dec 2021 00:05:35 +1000
Subject: [PATCH] Close minecart containers if they change to another server


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index f1b987faa92ae3eaee8ce740f74ade5a8499df2d..0d288d45cca7021815925dc4d045d1e6223b6b52 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -489,6 +489,7 @@ public class MultiPaper {
 
     public static void unlockChunk(LevelChunk chunk) {
         if (MultiPaper.isChunkLocal(chunk)) {
+            chunk.level.entityManager.getEntities(chunk.getPos()).forEach(MultiPaperEntitiesHandler::onEntityUnlock);
             broadcastPacketToExternalServers(chunk.externalEntitiesSubscribers, () -> new SendEntitiesPacket(chunk));
             broadcastPacketToExternalServers(chunk.externalSubscribers, () -> new SendTickListPacket(chunk));
             for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
diff --git a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
index bf715c64d79f665ddf0b8ec918f8caf3dcf21c56..8ac77b05705320c6f79e2b53eadba657bfe3c5e6 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
@@ -8,6 +8,7 @@ import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.server.level.ChunkMap;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.Container;
 import net.minecraft.world.entity.*;
 import net.minecraft.world.entity.ai.attributes.AttributeInstance;
 import net.minecraft.world.entity.ai.attributes.AttributeModifier;
@@ -25,6 +26,7 @@ import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang.ArrayUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.entity.HumanEntity;
 import puregero.multipaper.config.MultiPaperConfiguration;
 import puregero.multipaper.externalserverprotocol.*;
 import puregero.multipaper.mastermessagingprotocol.messages.masterbound.UnsubscribeEntitiesMessage;
@@ -122,6 +124,7 @@ public class MultiPaperEntitiesHandler {
             if (!MultiPaper.isChunkLocal(chunkTo)) {
                 // Leaving our jurisdiction, do a full entity update to ensure the new external server has all the required info
                 if (!(entity instanceof ServerPlayer)) { // Ignore players as they aren't ticked by the new external server
+                    onEntityUnlock(entity);
                     MultiPaper.runSync(() -> MultiPaper.broadcastPacketToExternalServers(chunkTo.externalEntitiesSubscribers, () -> new EntityUpdateNBTPacket(entity)));
                     if (entity instanceof Mob mob) {
                         MultiPaper.runSync(() -> {
@@ -158,6 +161,12 @@ public class MultiPaperEntitiesHandler {
         }
     }
 
+    public static void onEntityUnlock(Entity entity) {
+        if (entity instanceof Container container) {
+            new ArrayList<>(container.getViewers()).forEach(HumanEntity::closeInventory);
+        }
+    }
+
     private static void setRemovedRecursive(Entity entity) {
         for (Entity passenger : entity.getPassengers()) {
             if (!(passenger instanceof ServerPlayer)) {
