From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 22 Dec 2021 23:10:18 +1000
Subject: [PATCH] Don't suck from external minecarts and boats


diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 7d55b3d06afac9b7669ea803bcb138e0a9213978..4f3b7b506679f292105fa9eced4a8f73fcfcc69f 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -40,6 +40,7 @@ import org.bukkit.event.inventory.InventoryPickupItemEvent;
 import org.bukkit.inventory.Inventory;
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperContainerHandler;
+import puregero.multipaper.MultiPaperEntityInteractHandler;
 import puregero.multipaper.externalserverprotocol.PullItemFromContainerPacket;
 // CraftBukkit end
 
@@ -499,7 +500,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public static boolean suckInItems(Level world, Hopper hopper) {
         Container iinventory = HopperBlockEntity.getSourceContainer(world, hopper);
 
-        if (iinventory != null) {
+        if (iinventory != null && !(iinventory instanceof Entity entity && MultiPaperEntityInteractHandler.getOwner(entity) != null)) { // MultiPaper - don't suck from external minecarts and boats
             Direction enumdirection = Direction.DOWN;
 
             // Paper start - optimize hoppers and remove streams
