From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 31 Dec 2021 19:47:45 +1000
Subject: [PATCH] Coordinate the tick ourselves


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 14bfbde68184c4bfec098be107b7a33a1f7e25ef..938d4143b45f1cf3e766e23be754274dfb64cd40 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -271,6 +271,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     private boolean mayHaveDelayedTasks;
     private final PackRepository packRepository;
     private final ServerScoreboard scoreboard;
+    public boolean shouldTick = true; // MultiPaper - coordinate the tick ourselves
     @Nullable
     private CommandStorage commandStorage;
     private final CustomBossEvents customBossEvents;
@@ -1224,6 +1225,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 long start = System.nanoTime(), curTime, tickSection = start; // Paper - Further improve server tick loop
                 lastTick = start - TICK_TIME; // Paper
                 while (this.running) {
+                    shouldTick = false; // MultiPaper - coordinate the tick ourselves
                     long i = ((curTime = System.nanoTime()) / (1000L * 1000L)) - this.nextTickTime; // Paper
 
                     if (i > 5000L && this.nextTickTime - this.lastOverloadWarning >= 30000L) { // CraftBukkit
@@ -1348,7 +1350,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     private boolean canSleepForTickNoOversleep() {
-        return this.forceTicks || this.runningTask() || Util.getMillis() < this.nextTickTime;
+        return this.forceTicks || this.runningTask() || !shouldTick; //Util.getMillis() < this.nextTickTime; // MultiPaper - coordinate the tick ourselves
     }
     // Paper end
 
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index be282a42391efc55692e8707e28cca16f125f877..320968e6c53703621d765759639b6f9a476abcb9 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -387,6 +387,12 @@ public class MultiPaperConnection extends Thread {
         }
     }
 
+    public void tick(DataInputStream in, DataOutputSender out) throws IOException {
+        if (MinecraftServer.getServer() != null) {
+            MinecraftServer.getServer().shouldTick = true;
+        }
+    }
+
     public void writeTickTime(long time, double tps) throws IOException {
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeTickTime");
