From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 24 Dec 2021 16:53:05 +1000
Subject: [PATCH] Allow external players to push minecarts


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index eec7d7a5b558830111831792c42665724613af23..69c4af4f46f76769eec471d3010b3e603eb6bd34 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -53,6 +53,7 @@ import org.bukkit.event.vehicle.VehicleDamageEvent;
 import org.bukkit.event.vehicle.VehicleDestroyEvent;
 import org.bukkit.event.vehicle.VehicleEntityCollisionEvent;
 import org.bukkit.util.Vector;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public abstract class AbstractMinecart extends Entity {
@@ -488,7 +489,28 @@ public abstract class AbstractMinecart extends Entity {
 
             this.firstTick = false;
         }
+
+        getPushedByExternalPlayers(); // MultiPaper
+    }
+
+    // MultiPaper start
+    public void getPushedByExternalPlayers() {
+        List<Entity> list = this.level.getEntities(this, this.getBoundingBox(), EntitySelector.pushable(this, level.paperConfig().collisions.fixClimbingBypassingCrammingRule));
+
+        if (!list.isEmpty()) {
+            this.numCollisions = Math.max(0, this.numCollisions - this.level.paperConfig().collisions.maxEntityCollisions);
+            for (int j = 0; j < list.size() && this.numCollisions < this.level.paperConfig().collisions.maxEntityCollisions; ++j) {
+                Entity entity = list.get(j);
+
+                if (MultiPaper.isExternalPlayer(entity)) {
+                    this.numCollisions++;
+
+                    this.push(entity);
+                }
+            }
+        }
     }
+    // MultiPaper end
 
     protected double getMaxSpeed() {
         return (this.isInWater() ? this.maxSpeed / 2.0D: this.maxSpeed); // CraftBukkit
