From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 24 Dec 2021 16:53:05 +1000
Subject: [PATCH] Allow external players to push minecarts


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index 96a7bfd921e59f298f0ee502d356cc3552c30ce8..f1bf0f313d058913fbd4c5c3e39b48d17f06911f 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -54,6 +54,7 @@ import org.bukkit.event.vehicle.VehicleDamageEvent;
 import org.bukkit.event.vehicle.VehicleDestroyEvent;
 import org.bukkit.event.vehicle.VehicleEntityCollisionEvent;
 import org.bukkit.util.Vector;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public abstract class AbstractMinecart extends Entity {
@@ -479,7 +480,28 @@ public abstract class AbstractMinecart extends Entity {
 
             this.firstTick = false;
         }
+
+        getPushedByExternalPlayers(); // MultiPaper
+    }
+
+    // MultiPaper start
+    public void getPushedByExternalPlayers() {
+        List<Entity> list = this.level.getEntities(this, this.getBoundingBox(), EntitySelector.pushable(this, level.paperConfig.fixClimbingBypassingCrammingRule));
+
+        if (!list.isEmpty()) {
+            this.numCollisions = Math.max(0, this.numCollisions - this.level.paperConfig.maxCollisionsPerEntity);
+            for (int j = 0; j < list.size() && this.numCollisions < this.level.paperConfig.maxCollisionsPerEntity; ++j) {
+                Entity entity = list.get(j);
+
+                if (MultiPaper.isExternalPlayer(entity)) {
+                    this.numCollisions++;
+
+                    this.push(entity);
+                }
+            }
+        }
     }
+    // MultiPaper end
 
     protected double getMaxSpeed() {
         return (this.isInWater() ? this.maxSpeed / 2.0D: this.maxSpeed); // CraftBukkit
