From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 Dec 2021 16:39:52 +1000
Subject: [PATCH] Reobf error with ServerPlaceRecipe


diff --git a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
index d3d6627901ba559292152e051a2f11f225f8dd20..d44c6efaef352987ad1615cee9c8bdc80d4558e7 100644
--- a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -189,7 +189,7 @@ public class ServerPlaceRecipe<C extends Container> implements PlaceRecipe<Integ
     private int getAmountOfFreeSlotsInInventory() {
         int i = 0;
 
-        for(ItemStack itemStack : this.inventory.items) {
+        for(ItemStack itemStack : this.inventory.items) { // MultiPaper - patch on this line to fix reobf error
             if (itemStack.isEmpty()) {
                 ++i;
             }
