From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 29 Dec 2021 15:01:38 +1000
Subject: [PATCH] Add temporary debug logs for checking if a chunk can be
 loaded twice


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 395fe597a2b53e28df65f9e92fb18c7eda01d97e..94f94d468300cbf0f29f21d7e7f5f7c1f93f1bb4 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -973,7 +973,16 @@ public class LevelChunk extends ChunkAccess {
         });
     }
 
+    // MultiPaper start - temporary debug logs for checking if a chunk can be loaded twice
+    private boolean hasBeenLoadedBefore = false;
     public void setLoaded(boolean loadedToWorld) {
+        if (loaded) {
+            if (hasBeenLoadedBefore) {
+                new Exception("Loading a chunk that's already been loaded " + level.getWorld().getName() + "," + chunkPos.x + "," + chunkPos.z).printStackTrace();
+            }
+            hasBeenLoadedBefore = true;
+        }
+    // MultiPaper end - temporary debug logs for checking if a chunk can be loaded twice
         this.loaded = loadedToWorld;
     }
 
