From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Mar 2022 16:48:05 +1000
Subject: [PATCH] Only touch tridents that we are ticking


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
index 0ea182962d6647629fc98c9e7406f7b7ce012c5e..92bf58e0fca5daed1d4e5e2ca93c0e9d9a70d2a0 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -22,6 +22,7 @@ import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
+import puregero.multipaper.MultiPaper;
 
 public class ThrownTrident extends AbstractArrow {
 
@@ -197,6 +198,7 @@ public class ThrownTrident extends AbstractArrow {
 
     @Override
     public void playerTouch(Player player) {
+        if (MultiPaper.isChunkLocal(this.level().getChunkIfLoaded(blockPosition))) // MultiPaper - ensure we're still ticking this entity
         if (this.ownedBy(player) || this.getOwner() == null) {
             super.playerTouch(player);
         }
