From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 6 Jan 2022 16:17:48 +1000
Subject: [PATCH] Run all the queued tasks before the next tick


diff --git a/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java b/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java
index dcce05d2f4ab16424db4ab103a12188e207a457b..86910dbaad3b1bbdc4bde45094ca6db0f9869146 100644
--- a/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java
+++ b/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java
@@ -128,11 +128,12 @@ public abstract class BlockableEventLoop<R extends Runnable> implements Profiler
         ++this.blockingCount;
 
         try {
-            while(!stopCondition.getAsBoolean()) {
-                if (!this.pollTask()) {
-                    this.waitForTasks();
-                }
-            }
+            // MultiPaper start - run all the queued tasks before the next tick
+            do {
+                while (this.pollTask());
+                this.waitForTasks();
+            } while (!stopCondition.getAsBoolean());
+            // MultiPaper end - run all the queued tasks before the next tick
         } finally {
             --this.blockingCount;
         }
