From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 13 Feb 2022 14:02:25 +1000
Subject: [PATCH] Run tasks while reading chunk data


diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 4bfe5607682de68e7261a848e1fee92051fcc1ae..5a91b1aba7d9df8c968cd1ce8963c669d8475ae7 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1,6 +1,7 @@
 package net.minecraft.server.level;
 
 import co.aikar.timings.Timing; // Paper
+import com.destroystokyo.paper.io.PaperFileIOThread;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableList.Builder;
 import com.google.common.collect.Iterables;
@@ -1716,9 +1717,17 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
     @Override
     public CompoundTag readSync(ChunkPos chunkcoordintpair) throws IOException {
         if (Thread.currentThread() != com.destroystokyo.paper.io.PaperFileIOThread.Holder.INSTANCE) {
-            CompoundTag ret = com.destroystokyo.paper.io.PaperFileIOThread.Holder.INSTANCE
-                .loadChunkDataAsyncFuture(this.level, chunkcoordintpair.x, chunkcoordintpair.z, com.destroystokyo.paper.io.IOUtil.getPriorityForCurrentThread(),
-                    false, true, true).join().chunkData;
+            // MultiPaper start - don't block the main thread, run tasks in the meantime
+            CompletableFuture<PaperFileIOThread.ChunkData> completablefuture = com.destroystokyo.paper.io.PaperFileIOThread.Holder.INSTANCE
+                    .loadChunkDataAsyncFuture(this.level, chunkcoordintpair.x, chunkcoordintpair.z, com.destroystokyo.paper.io.IOUtil.getPriorityForCurrentThread(),
+                            false, true, true);
+
+            if (!completablefuture.isDone()) {
+                level.chunkSource.mainThreadProcessor.managedBlock(completablefuture::isDone);
+            }
+
+            CompoundTag ret = completablefuture.join().chunkData;
+            // MultiPaper end
 
             if (ret == com.destroystokyo.paper.io.PaperFileIOThread.FAILURE_VALUE) {
                 throw new IOException("See logs for further detail");
