From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 30 Dec 2021 15:09:05 +1000
Subject: [PATCH] Broadcast block change when right-clicking a block to ensure
 external servers are in sync


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 99904f233d95ce0e25cc9a5efade3ad973905c38..eaf598ede9bff4d248a856aa94ff432e1eb0250b 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1897,6 +1897,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
                 this.player.connection.send(new ClientboundBlockUpdatePacket(worldserver, blockposition));
                 this.player.connection.send(new ClientboundBlockUpdatePacket(worldserver, blockposition.relative(enumdirection)));
+
+                // MultiPaper - broadcast block change to other servers as they may be out of sync
+                worldserver.chunkSource.blockChanged(blockposition);
+                worldserver.chunkSource.blockChanged(blockposition.relative(enumdirection));
+                // MultiPaper - broadcast block change to other servers as they may be out of sync
             } else {
                 ServerGamePacketListenerImpl.LOGGER.warn("Rejecting UseItemOnPacket from {}: Location {} too far away from hit block {}.", new Object[]{this.player.getGameProfile().getName(), vec3d, blockposition});
             }
