From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 29 Dec 2021 15:01:38 +1000
Subject: [PATCH] Add temporary debug logs for checking if a chunk can be
 loaded twice


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 840b89baec31fe79b51d7b0fa272ff214f56f731..455ee65c8eaa77b315b6d81417c25553398f1faf 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -993,7 +993,16 @@ public class LevelChunk extends ChunkAccess {
         });
     }
 
+    // MultiPaper start - temporary debug logs for checking if a chunk can be loaded twice
+    private boolean hasBeenLoadedBefore = false;
     public void setLoaded(boolean loadedToWorld) {
+        if (loaded) {
+            if (hasBeenLoadedBefore) {
+                new Exception("Loading a chunk that's already been loaded " + level.getWorld().getName() + "," + chunkPos.x + "," + chunkPos.z).printStackTrace();
+            }
+            hasBeenLoadedBefore = true;
+        }
+    // MultiPaper end - temporary debug logs for checking if a chunk can be loaded twice
         this.loaded = loadedToWorld;
     }
 
