From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Mar 2022 20:12:39 +1000
Subject: [PATCH] Add option to reduce tracking of inactive entities


diff --git a/src/main/java/co/aikar/timings/WorldTimingsHandler.java b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
index 22687667ec69a954261e55e59261286ac1b8b8cd..80d8292a8c197acb1bbba066e5f0b0fce368bcf3 100644
--- a/src/main/java/co/aikar/timings/WorldTimingsHandler.java
+++ b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
@@ -36,6 +36,7 @@ public class WorldTimingsHandler {
     public final Timing tileEntityPending;
     public final Timing tracker1;
     public final Timing tracker2;
+    public final Timing trackerInactive; // MultiPaper
     public final Timing doTick;
     public final Timing tickEntities;
     public final Timing chunks;
@@ -110,6 +111,7 @@ public class WorldTimingsHandler {
 
         tracker1 = Timings.ofSafe(name + "tracker stage 1");
         tracker2 = Timings.ofSafe(name + "tracker stage 2");
+        trackerInactive = Timings.ofSafe(name + "tracker multipaper inactive stage"); // MultiPaper
         doTick = Timings.ofSafe(name + "doTick");
         tickEntities = Timings.ofSafe(name + "tickEntities");
 
diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 42c790768b83036f8ef32d77d8cf9a747ae76865..7c72eb2183c011d4cbeb7c8188a0bb43c5851ab4 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1259,6 +1259,8 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
 
     // Paper start - optimised tracker
     private final void processTrackQueue() {
+        if (this.level.inactiveTracker.tick()) return; // MultiPaper - inactive tracker
+
         //this.level.timings.tracker1.startTiming(); // Purpur
         try {
             for (TrackedEntity tracker : this.entityMap.values()) {
@@ -1531,7 +1533,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
         // Paper start - use distance map to optimise tracker
         com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> lastTrackerCandidates;
 
-        final void updatePlayers(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) {
+        public final void updatePlayers(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> newTrackerCandidates) { // MultiPaper
             com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> oldTrackerCandidates = this.lastTrackerCandidates;
             this.lastTrackerCandidates = newTrackerCandidates;
 
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index add11f81f95dda1899ab534becb3569c87b8fead..0222abd23dbea1f6fc0c29a5cfab6afa3484e33d 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -58,7 +58,7 @@ public class ServerEntity {
     private static final int TOLERANCE_LEVEL_ROTATION = 1;
     private final ServerLevel level;
     private final Entity entity;
-    private final int updateInterval;
+    public final int updateInterval; // MultiPaper
     private final boolean trackDelta;
     private final Consumer<Packet<?>> broadcast;
     private final VecDeltaCodec positionCodec = new VecDeltaCodec();
@@ -73,6 +73,7 @@ public class ServerEntity {
     private boolean wasOnGround;
     @Nullable
     private List<SynchedEntityData.DataValue<?>> trackedDataValues;
+    public int lastChangesSent; // MultiPaper
     // CraftBukkit start
     public final Set<ServerPlayerConnection> trackedPlayers; // Paper - private -> package // Purpur - package -> public
 
@@ -154,7 +155,7 @@ public class ServerEntity {
             }
         }
 
-        if (this.tickCount % this.updateInterval == 0 || this.entity.hasImpulse || this.entity.getEntityData().isDirty()) {
+        if (this.tickCount % this.updateInterval == 0 || this.entity.hasImpulse || this.entity.getEntityData().isDirty() || this.entity.sendChangesThisTick) { // MultiPaper
             int i;
             int j;
 
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 8f45de99e56d3911d7ccdf020b0311df042e3fd5..6109bb4b08b30ca3af3cc4c29ce126706bc81c7d 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -182,6 +182,7 @@ import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet; // Paper
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
 import puregero.multipaper.MultiPaperEntitiesHandler;
+import puregero.multipaper.MultiPaperInactiveTracker;
 import puregero.multipaper.externalserverprotocol.SpawnUpdatePacket;
 import puregero.multipaper.externalserverprotocol.TimeUpdatePacket;
 
@@ -214,6 +215,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents;
     private final List<BlockEventData> blockEventsToReschedule;
     private boolean handlingTick;
+    public final MultiPaperInactiveTracker inactiveTracker = new MultiPaperInactiveTracker(this); // MultiPaper
     private final List<CustomSpawner> customSpawners;
     @Nullable
     private final EndDragonFight dragonFight;
@@ -1302,6 +1304,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         //gameprofilerfiller.incrementCounter("tickNonPassenger"); // Purpur
         if (isActive) { // Paper - EAR 2
             TimingHistory.activatedEntityTicks++;
+        inactiveTracker.track(entity); // MultiPaper
         entity.tick();
         entity.postTick(); // CraftBukkit
         } else { entity.inactiveTick(); } // Paper - EAR 2
@@ -1342,6 +1345,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 //gameprofilerfiller.incrementCounter("tickPassenger"); // Purpur
                 // Paper start - EAR 2
                 if (isActive) {
+                inactiveTracker.track(passenger); // MultiPaper
                 passenger.rideTick();
                 passenger.postTick(); // CraftBukkit
                 } else {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f99828c7a8afd96c2c263b42f033b4cfdacb7e17..d44a5fa201f5de068f8930929618b98a4849ee9c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -328,6 +328,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public float maxUpStep; // Purpur - private -> public
     public boolean noPhysics;
     protected final RandomSource random;
+    public boolean sendChangesThisTick = false; // MultiPaper
     public int tickCount;
     public int remainingFireTicks;
     public boolean wasTouchingWater;
diff --git a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
index a4e789271c720f2969011610cbcf82b59b4c0b1c..422c06c39d0189acc9f4e0a66f97a90f2feb48b5 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
@@ -187,6 +187,10 @@ public class MultiPaperEntitiesHandler {
     }
 
     public static void handleEntityUpdate(ExternalServerConnection connection, Entity entity, Packet<?> packet) {
+        if (entity.level instanceof ServerLevel serverLevel) {
+            serverLevel.inactiveTracker.track(entity);
+        }
+
         if (packet instanceof ClientboundMoveEntityPacket moveEntityPacket) {
             VecDeltaCodec vecDeltaCodec = new VecDeltaCodec();
             vecDeltaCodec.setBase(entity.position());
diff --git a/src/main/java/puregero/multipaper/MultiPaperInactiveTracker.java b/src/main/java/puregero/multipaper/MultiPaperInactiveTracker.java
new file mode 100644
index 0000000000000000000000000000000000000000..bdc544bd0cc908349c66b31aa178297140b36d55
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperInactiveTracker.java
@@ -0,0 +1,57 @@
+package puregero.multipaper;
+
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import puregero.multipaper.config.MultiPaperConfiguration;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class MultiPaperInactiveTracker {
+    private final ServerLevel level;
+    private final Queue<Entity> entitiesToSendChangesThisTick = new LinkedList<>();
+
+    public MultiPaperInactiveTracker(ServerLevel level) {
+        this.level = level;
+    }
+
+    public void track(Entity entity) {
+        if (MultiPaperConfiguration.get().optimizations.ticksPerInactiveEntityTracking > 1 && !entity.sendChangesThisTick && entity.tracker != null) {
+            entity.sendChangesThisTick = true;
+            entitiesToSendChangesThisTick.add(entity);
+        }
+    }
+
+    public boolean tick() {
+        int tickCount = level.getServer().getTickCount();
+        if (MultiPaperConfiguration.get().optimizations.ticksPerInactiveEntityTracking > 1 && tickCount % MultiPaperConfiguration.get().optimizations.ticksPerInactiveEntityTracking != 0) {
+            level.timings.trackerInactive.startTiming();
+            try {
+                Iterator<Entity> iterator = entitiesToSendChangesThisTick.iterator();
+                while (iterator.hasNext()) {
+                    Entity entity = iterator.next();
+
+                    if (entity.tracker == null) {
+                        entity.sendChangesThisTick = false;
+                        iterator.remove();
+                        continue;
+                    }
+
+                    if (tickCount - entity.tracker.serverEntity.lastChangesSent >= entity.tracker.serverEntity.updateInterval) {
+                        entity.tracker.serverEntity.lastChangesSent = tickCount;
+                        entity.tracker.updatePlayers(entity.getPlayersInTrackRange());
+                        entity.tracker.serverEntity.sendChanges();
+                        entity.sendChangesThisTick = false;
+                        iterator.remove();
+                    }
+                }
+            } finally {
+                level.timings.trackerInactive.stopTiming();
+            }
+            return true;
+        }
+
+        return false;
+    }
+}
