From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jan 2022 16:43:52 +1000
Subject: [PATCH] Send server port to the master server


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 681ed57929602535ebae9852bddba9a0b3920da9..65e51146e9ad892b66523f3238fbb783ee09de50 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -270,6 +270,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             this.setPort(dedicatedserverproperties.serverPort);
         }
         bindAddress = new java.net.InetSocketAddress(inetaddress, this.getPort());
+        MultiPaper.setPort(this.getPort()); // MultiPaper
         }
         // Paper end
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 6e9d0c7080416000aaf9148ae3c26d4cd7863b65..12f6e543b7ac19b6f9ad66b424f9a1b6695dc560 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -99,6 +99,14 @@ public class MultiPaper {
         return syncJsonFiles;
     }
 
+    private static Boolean advertiseToBuiltInProxy = null;
+    public static boolean shouldAdvertiseToBuiltInProxy() {
+        if (advertiseToBuiltInProxy == null) {
+            loadConfig();
+        }
+        return advertiseToBuiltInProxy;
+    }
+
     private static void loadConfig() {
         YamlConfiguration config = new YamlConfiguration();
         File file = new File("multipaper.yml");
@@ -140,6 +148,13 @@ public class MultiPaper {
 
         syncJsonFiles = config.getBoolean("syncJsonFiles");
 
+        if (!config.contains("advertiseToBuiltInProxy")) {
+            save = true;
+            config.set("advertiseToBuiltInProxy", true);
+        }
+
+        advertiseToBuiltInProxy = config.getBoolean("advertiseToBuiltInProxy");
+
         try {
             if (save) {
                 config.save(file);
@@ -586,4 +601,8 @@ public class MultiPaper {
     public static void broadcastPlayerAction(ServerPlayer player, ServerboundSetCarriedItemPacket packet) {
         broadcastPacketToExternalServers(new PlayerActionPacket(player, packet));
     }
+
+    public static void setPort(int port) {
+        getConnection().sendPort(port);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 827993c7708b62bca51e01aef75c2e39ba66029c..19cdcfeb77f1552f544024427cf313d49fa98dae 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -41,6 +41,7 @@ public class MultiPaperConnection extends Thread {
     private final Map<Integer, Request> callbacks = new ConcurrentHashMap<>();
     private final Map<String, ExternalServer> serversMap = new ConcurrentHashMap<>();
     private String myName = null;
+    private int port = -1;
 
     public MultiPaperConnection() {
         super("MultiPaperConnection Thread");
@@ -118,6 +119,10 @@ public class MultiPaperConnection extends Thread {
                     }
                 }
 
+                if (port >= 0) {
+                    sendPort(port);
+                }
+
                 DataInputStream in = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
                 while (!socket.isClosed()) {
                     int id = in.readInt();
@@ -838,4 +843,20 @@ public class MultiPaperConnection extends Thread {
             e.printStackTrace();
         }
     }
+
+    public void sendPort(int port) {
+        if (!MultiPaper.shouldAdvertiseToBuiltInProxy()) {
+            return;
+        }
+
+        this.port = port;
+        try {
+            DataOutputSender out = new DataOutputSender(this);
+            out.writeUTF("setPort");
+            out.writeInt(port);
+            out.send(null);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
 }
