From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jan 2022 16:43:52 +1000
Subject: [PATCH] Send server port to the master server


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index e80411cf991d9fd8b7d280cb381c54eff123be73..a9b349476570be3ed7748b439f5151c58c6e714a 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -270,6 +270,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             this.setPort(dedicatedserverproperties.serverPort);
         }
         bindAddress = new java.net.InetSocketAddress(inetaddress, this.getPort());
+        MultiPaper.setPort(this.getPort()); // MultiPaper
         }
         // Paper end
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index bbf8fa55a9815ba687add0bab63389e66cfc7590..928491af98ed842108f39a7b3cc0e26086998911 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -92,6 +92,14 @@ public class MultiPaper {
         return syncJsonFiles;
     }
 
+    private static Boolean advertiseToBuiltInProxy = null;
+    public static boolean shouldAdvertiseToBuiltInProxy() {
+        if (advertiseToBuiltInProxy == null) {
+            loadConfig();
+        }
+        return advertiseToBuiltInProxy;
+    }
+
     private static void loadConfig() {
         YamlConfiguration config = new YamlConfiguration();
         File file = new File("multipaper.yml");
@@ -126,6 +134,13 @@ public class MultiPaper {
 
         syncJsonFiles = config.getBoolean("syncJsonFiles");
 
+        if (!config.contains("advertiseToBuiltInProxy")) {
+            save = true;
+            config.set("advertiseToBuiltInProxy", true);
+        }
+
+        advertiseToBuiltInProxy = config.getBoolean("advertiseToBuiltInProxy");
+
         try {
             if (save) {
                 config.save(file);
@@ -571,4 +586,8 @@ public class MultiPaper {
     public static void broadcastPlayerAction(ServerPlayer player, ServerboundSetCarriedItemPacket packet) {
         broadcastPacketToExternalServers(new PlayerActionPacket(player, packet));
     }
+
+    public static void setPort(int port) {
+        getConnection().sendPort(port);
+    }
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 425f999ea29a86375b3c63cc3dd5b4de424b82dc..b19c23dbe41819f0b7e7958b4c26e1786c81a7fb 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -48,6 +48,7 @@ public class MultiPaperConnection extends Thread {
     private final Map<String, ExternalServer> serversMap = new ConcurrentHashMap<>();
     private String myName = null;
     public String secret;
+    private int port = -1;
 
     public MultiPaperConnection() {
         super("MultiPaperConnection Thread");
@@ -125,6 +126,10 @@ public class MultiPaperConnection extends Thread {
                     }
                 }
 
+                if (port >= 0) {
+                    sendPort(port);
+                }
+
                 DataInputStream in = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
                 while (!socket.isClosed()) {
                     int id = in.readInt();
@@ -865,4 +870,20 @@ public class MultiPaperConnection extends Thread {
             e.printStackTrace();
         }
     }
+
+    public void sendPort(int port) {
+        if (!MultiPaper.shouldAdvertiseToBuiltInProxy()) {
+            return;
+        }
+
+        this.port = port;
+        try {
+            DataOutputSender out = new DataOutputSender(this);
+            out.writeUTF("setPort");
+            out.writeInt(port);
+            out.send(null);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
 }
