From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 9 Jan 2022 15:50:30 +1000
Subject: [PATCH] Add temporary debug log for dispensers


diff --git a/src/main/java/net/minecraft/world/level/block/DispenserBlock.java b/src/main/java/net/minecraft/world/level/block/DispenserBlock.java
index 83ac86b3c1e7b9233f2db8e5488f97c5b44f8843..6f62be95bd657883af1916dda84c6e2af7256d28 100644
--- a/src/main/java/net/minecraft/world/level/block/DispenserBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DispenserBlock.java
@@ -78,6 +78,7 @@ public class DispenserBlock extends BaseEntityBlock {
     }
 
     public void dispenseFrom(ServerLevel world, BlockPos pos) {
+        if (pos.getZ() == -43 && pos.getY() == 107 && pos.getX() == -132) new Exception("Dispensing " + pos).printStackTrace();
         BlockSourceImpl sourceblock = new BlockSourceImpl(world, pos);
         DispenserBlockEntity tileentitydispenser = (DispenserBlockEntity) sourceblock.getEntity();
         int i = tileentitydispenser.getRandomSlot();
diff --git a/src/main/java/net/minecraft/world/level/block/DropperBlock.java b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
index b3203049eade7d11602fa2a12a8104a732d67552..a6cbf19640caca359c6d13e783af6af518bf824e 100644
--- a/src/main/java/net/minecraft/world/level/block/DropperBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/DropperBlock.java
@@ -39,6 +39,7 @@ public class DropperBlock extends DispenserBlock {
 
     @Override
     public void dispenseFrom(ServerLevel world, BlockPos pos) {
+        if (pos.getZ() == -43 && pos.getY() == 107 && pos.getX() == -132) new Exception("Dispensing " + pos).printStackTrace();
         BlockSourceImpl sourceblock = new BlockSourceImpl(world, pos);
         DispenserBlockEntity tileentitydispenser = (DispenserBlockEntity) sourceblock.getEntity();
         int i = tileentitydispenser.getRandomSlot();
