From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 19:54:53 +1000
Subject: [PATCH] Add shutdown packet handler


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 22aa3dacc2c9ca34aa8e2dd4c60ae84183a18f3a..a3d230bba4e3168f5b05864d1c95634ba0aed450 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -394,6 +394,10 @@ public class MultiPaperConnection extends Thread {
         }
     }
 
+    public void shutdown(DataInputStream in, DataOutputSender out) throws IOException {
+        Bukkit.shutdown();
+    }
+
     public void writeTickTime(long time, double tps) throws IOException {
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeTickTime");
