From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Mar 2022 01:01:01 +1000
Subject: [PATCH] Temporarily disable piercing to prevent crashes


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 3f8e10c33e0f8739602f8327da3a8f66318c5c94..e4b5ade5e36a11109cd79376765babdaa08e6f1a 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -234,7 +234,7 @@ public abstract class AbstractArrow extends Projectile {
                     this.hasImpulse = true;
                 }
 
-                if (movingobjectpositionentity == null || this.getPierceLevel() <= 0) {
+                if (true || movingobjectpositionentity == null || this.getPierceLevel() <= 0) { // MultiPaper - temporarily disable piercing to prevent crashes
                     break;
                 }
 
