From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 2 Apr 2022 23:20:07 +1000
Subject: [PATCH] Reset horse inventory during load


diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index fdf41f4990923aa4309febb9c0ca8009de8e16f1..4a98ed11d735ef75639167796cfd298af1cc290d 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -942,6 +942,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
     @Override
     public void readAdditionalSaveData(CompoundTag nbt) {
         super.readAdditionalSaveData(nbt);
+        this.inventory.clearContent(); // MultiPaper - reset inventory before loading data
         this.setEating(nbt.getBoolean("EatingHaystack"));
         this.setBred(nbt.getBoolean("Bred"));
         this.setTemper(nbt.getInt("Temper"));
