From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jan 2022 19:49:32 +1000
Subject: [PATCH] Allow for only ip forwarding with no uuid


diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 828f10d34a74e42043fd16c81e4834e860071c0d..c18713bb166afd7c8e68593884c9416fd1e1ad01 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -133,12 +133,13 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                     if (!handledByEvent && proxyLogicEnabled) {
                         // Paper end
                     // if (org.spigotmc.SpigotConfig.bungee) { // Paper - comment out, we check above!
-                        if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.BYPASS_HOSTCHECK || ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) { // Paper
+                        if ( ( split.length == 2 || split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.BYPASS_HOSTCHECK || ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) { // Paper // MultiPaper - Allow for only ip-forwarding
                             // Paper start - Unix domain socket support
                             java.net.SocketAddress socketAddress = connection.getRemoteAddress();
                             packet.hostName = split[0];
                             connection.address = new java.net.InetSocketAddress(split[1], socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getPort() : 0);
                             // Paper end
+                            if (split.length > 2) // MultiPaper - Allow for only ip-forwarding
                             connection.spoofedUUID = com.mojang.util.UUIDTypeAdapter.fromString( split[2] );
                         } else
                         {
@@ -151,7 +152,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                         {
                             connection.spoofedProfile = ServerHandshakePacketListenerImpl.gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
                         }
-                    } else if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) {
+                    } else if ( ( split.length == 2 || split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) { // MultiPaper - Allow for only ip-forwarding
                         Component chatmessage = Component.literal("Unknown data in login hostname, did you forget to enable BungeeCord in spigot.yml?");
                         this.connection.send(new ClientboundLoginDisconnectPacket(chatmessage));
                         this.connection.disconnect(chatmessage);
