From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Apr 2022 13:08:48 +1000
Subject: [PATCH] Tick death for external players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index d5cebdd9b68bbdd0561636b6554e7b8bd08e1a97..f6342c8c8241132dd3365f131cee6be228cf6c38 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -639,6 +639,14 @@ public abstract class LivingEntity extends Entity implements Attackable {
         return true;
     }
 
+    // MultiPaper
+    public void tickDeathIfDead() {
+        if (this.isDeadOrDying() && this.level().shouldTickDeath(this)) {
+            this.tickDeath();
+        }
+    }
+    // MultiPaper end
+
     protected void tickDeath() {
         ++this.deathTime;
         if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved()) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 12e42d35111bef768c42e6a9d2a520a142c8bea0..f7fbe93557be9b62fc6491a100afc31bc08c2844 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -95,6 +95,8 @@ public class MultiPaper {
                 if (player.takeXpDelay > 0) {
                     --player.takeXpDelay;
                 }
+
+                externalPlayer.tickDeathIfDead();
             }
 
             player.syncExperience();
