From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Apr 2022 13:08:48 +1000
Subject: [PATCH] Tick death for external players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 59765ea9ea528e964b067cdfe1516e22362f7b7b..aafed8591c10c6db34f23713b3e11914be3d18b0 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -639,6 +639,14 @@ public abstract class LivingEntity extends Entity implements Attackable {
         return true;
     }
 
+    // MultiPaper
+    public void tickDeathIfDead() {
+        if (this.isDeadOrDying() && this.level.shouldTickDeath(this)) {
+            this.tickDeath();
+        }
+    }
+    // MultiPaper end
+
     protected void tickDeath() {
         ++this.deathTime;
         if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved()) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 43015a08a574857593bcef82eb3c7c60a3ac0856..8ee90cfc3b492794ff4c89962f6d87bf7eca0fb4 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -89,6 +89,8 @@ public class MultiPaper {
                 if (player.takeXpDelay > 0) {
                     --player.takeXpDelay;
                 }
+
+                externalPlayer.tickDeathIfDead();
             }
 
             player.syncExperience();
