From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 6 Jan 2022 16:17:48 +1000
Subject: [PATCH] Run all the queued tasks before the next tick


diff --git a/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java b/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java
index 7a4ade1a4190bf4fbb048919ae2be230f7b80fff..347cd0d57398d4f3a22d3eab3ac227323576e106 100644
--- a/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java
+++ b/src/main/java/net/minecraft/util/thread/BlockableEventLoop.java
@@ -132,11 +132,12 @@ public abstract class BlockableEventLoop<R extends Runnable> implements Profiler
         ++this.blockingCount;
 
         try {
-            while(!stopCondition.getAsBoolean()) {
-                if (!this.pollTask()) {
-                    this.waitForTasks();
-                }
-            }
+            // MultiPaper start - run all the queued tasks before the next tick
+            do {
+                while (this.pollTask());
+                this.waitForTasks();
+            } while (!stopCondition.getAsBoolean());
+            // MultiPaper end - run all the queued tasks before the next tick
         } finally {
             --this.blockingCount;
         }
