From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 9 Jan 2022 15:23:24 +1000
Subject: [PATCH] Sync player cameras


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index e858579d3d6226421f8aae5547bfa57e2f43b688..f8cc83dc23197ca39ecb91bcbe9d2634fb5ffa4a 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2058,6 +2058,8 @@ public class ServerPlayer extends Player {
         }
         this.connection.send(new ClientboundSetCameraPacket(entity));
         // Paper end
+
+        if (!PlayerSetCameraPacket.handlingSetCamera) MultiPaper.broadcastPacketToExternalServers(this, new PlayerSetCameraPacket(this, entity)); // MultiPaper
     }
 
     @Override
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index b32dfaad0142746f9bdd2aee1fef919503d563a6..1813990401927705ed8f9b869cb92d87bde39a0f 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -56,6 +56,7 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerStatsIncreasePacket.class, PlayerStatsIncreasePacket::new);
         addPacket(PlayerExperienceUpdatePacket.class, PlayerExperienceUpdatePacket::new);
         addPacket(PlayerListNameUpdatePacket.class, PlayerListNameUpdatePacket::new);
+        addPacket(PlayerSetCameraPacket.class, PlayerSetCameraPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerSetCameraPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerSetCameraPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..5328f19ffd573098fd71adb97c0b638272a22469
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerSetCameraPacket.java
@@ -0,0 +1,50 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerSetCameraPacket extends ExternalServerPacket {
+
+    public static boolean handlingSetCamera = false;
+
+    private final UUID uuid;
+    private final UUID uuidCamera;
+
+    public PlayerSetCameraPacket(ServerPlayer player, Entity camera) {
+        this.uuid = player.getUUID();
+        this.uuidCamera = camera == null ? null : camera.getUUID();
+    }
+
+    public PlayerSetCameraPacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+        uuidCamera = in.readBoolean() ? in.readUUID() : null;
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeBoolean(uuidCamera != null);
+        if (uuidCamera != null) {
+            out.writeUUID(uuidCamera);
+        }
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            handlingSetCamera = true;
+
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+            Entity entity = player.getLevel().getEntity(uuidCamera);
+            player.setCamera(entity);
+
+            handlingSetCamera = false;
+        });
+    }
+}
