From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 6 Apr 2022 20:30:56 +1000
Subject: [PATCH] A view distance of 1 also works


diff --git a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
index a8cfdade0a992baaf9b5649ae34410d0e6964336..8cb9441b30f168dc5e7bd24a61d4b1db300c7a6c 100644
--- a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
@@ -43,7 +43,7 @@ public class RegionizedPlayerChunkLoader {
 
     public static final TicketType<Long> REGION_PLAYER_TICKET = TicketType.create("region_player_ticket", Long::compareTo);
 
-    public static final int MIN_VIEW_DISTANCE = 2;
+    public static final int MIN_VIEW_DISTANCE = 1; // MultiPaper - minimum can actually be 1
     public static final int MAX_VIEW_DISTANCE = 32;
 
     public static final int TICK_TICKET_LEVEL = 31;
