From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:37:10 +1000
Subject: [PATCH] Only run PlayerArmorChangeEvent for local players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index a49851954f270aef858fe052b6b70dc4dff4b5f8..e149df6acb49a94b181393135e4020c827f315d0 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3220,7 +3220,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             if (this.equipmentHasChanged(itemstack, itemstack1)) {
                 // Paper start - PlayerArmorChangeEvent
-                if (this instanceof ServerPlayer && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) {
+                if (MultiPaper.isRealPlayer(this) && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) { // MultiPaper - only run PlayerArmorChangeEvent for local players
                     final org.bukkit.inventory.ItemStack oldItem = CraftItemStack.asBukkitCopy(itemstack);
                     final org.bukkit.inventory.ItemStack newItem = CraftItemStack.asBukkitCopy(itemstack1);
                     new PlayerArmorChangeEvent((Player) this.getBukkitEntity(), PlayerArmorChangeEvent.SlotType.valueOf(enumitemslot.name()), oldItem, newItem).callEvent();
