From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 19:54:53 +1000
Subject: [PATCH] Add shutdown packet handler


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index f588ef3d2b1061ade23b581c5245bf9cad5e9cf6..b5b1c3530ba1f786705366f954a80bc72fc54297 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -453,6 +453,10 @@ public class MultiPaperConnection extends Thread {
         }
     }
 
+    public void shutdown(DataInputStream in, DataOutputSender out) throws IOException {
+        Bukkit.shutdown();
+    }
+
     public void writeTickTime(long time, double tps) throws IOException {
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeTickTime");
