From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:37:10 +1000
Subject: [PATCH] Only run PlayerArmorChangeEvent for local players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 0002d7dce8419252909d79afab8963a94e0b2f33..1e34e8108472f4a2833bcb9c1c5fee468c4b68db 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3218,7 +3218,7 @@ public abstract class LivingEntity extends Entity {
 
             if (this.equipmentHasChanged(itemstack, itemstack1)) {
                 // Paper start - PlayerArmorChangeEvent
-                if (this instanceof ServerPlayer && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) {
+                if (MultiPaper.isRealPlayer(this) && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) { // MultiPaper - only run PlayerArmorChangeEvent for local players
                     final org.bukkit.inventory.ItemStack oldItem = CraftItemStack.asBukkitCopy(itemstack);
                     final org.bukkit.inventory.ItemStack newItem = CraftItemStack.asBukkitCopy(itemstack1);
                     new PlayerArmorChangeEvent((Player) this.getBukkitEntity(), PlayerArmorChangeEvent.SlotType.valueOf(enumitemslot.name()), oldItem, newItem).callEvent();
