From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 19:54:53 +1000
Subject: [PATCH] Add shutdown packet handler


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index b19c23dbe41819f0b7e7958b4c26e1786c81a7fb..6168c143920f989b24c54bc620597397407c22c6 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -441,6 +441,10 @@ public class MultiPaperConnection extends Thread {
         }
     }
 
+    public void shutdown(DataInputStream in, DataOutputSender out) throws IOException {
+        Bukkit.shutdown();
+    }
+
     public void writeTickTime(long time, double tps) throws IOException {
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeTickTime");
