From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 19:54:53 +1000
Subject: [PATCH] Add shutdown packet handler


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index ae66082eaa4c79fcd14503408ec9ad5a926a582d..e0d804a680740b493cef2dc599f2bc13f0be488e 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -441,6 +441,10 @@ public class MultiPaperConnection extends Thread {
         }
     }
 
+    public void shutdown(DataInputStream in, DataOutputSender out) throws IOException {
+        Bukkit.shutdown();
+    }
+
     public void writeTickTime(long time, double tps) throws IOException {
         DataOutputSender out = new DataOutputSender(this);
         out.writeUTF("writeTickTime");
