From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:54:41 +1000
Subject: [PATCH] Don't compare items if they're the same item


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 8db2cc74c19ac5d61a7c6688bad511c2fe15c012..b7a47c07a51a042a7cbb60f618a1bb410579281a 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -347,12 +347,12 @@ public abstract class Player extends LivingEntity {
 
         ItemStack itemstack = this.getMainHandItem();
 
-        if (!ItemStack.matches(this.lastItemInMainHand, itemstack)) {
+        if (this.lastItemInMainHand != itemstack && !ItemStack.matches(this.lastItemInMainHand, itemstack)) { // MultiPaper - don't compare items if they're the same item
             if (!ItemStack.isSame(this.lastItemInMainHand, itemstack)) {
                 this.resetAttackStrengthTicker();
             }
 
-            this.lastItemInMainHand = itemstack.copy();
+            this.lastItemInMainHand = itemstack; // MultiPaper - don't compare items if they're the same item
         }
 
         this.turtleHelmetTick();
