From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 19:54:53 +1000
Subject: [PATCH] Add shutdown packet handler


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 5aa135c683c7b3e511ec56c73eae5bea5484c1eb..aa2104fb1e526f4ebd66a76ca5d4f182d4d6b4ff 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -405,4 +405,9 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
             }
         });
     }
+
+    @Override
+    public void handle(ShutdownMessage message) {
+        Bukkit.shutdown();
+    }
 }
