From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 16 Jan 2022 11:51:16 +1000
Subject: [PATCH] Lock the chunk when writting it


diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 6701e45d6877e338afde0fd6355b941149788d6d..3c43c87a4866e6148506d85af332882db3cc771b 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1474,6 +1474,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             asyncSaveData = ChunkSerializer.getAsyncSaveData(this.level, chunk);
         }
 
+        MultiPaper.willSaveChunk(level, chunk);
         this.level.asyncChunkTaskManager.scheduleChunkSave(chunkPos.x, chunkPos.z, com.destroystokyo.paper.io.PrioritizedTaskQueue.NORMAL_PRIORITY,
             asyncSaveData, chunk);
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 3368a02d8dcad7da4bee964a62ac31cf259c0ed6..993fa10141c3814ea4cd0fc5e59f0dacd5b95ce3 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -535,6 +535,10 @@ public class MultiPaper {
         }
     }
 
+    public static void willSaveChunk(ServerLevel level, ChunkAccess chunk) {
+        getConnection().willSaveChunk(level.getWorld().getName(), chunk.locX, chunk.locZ);
+    }
+
     public static byte[] nbtToBytes(CompoundTag compoundTag) throws IOException {
         if (compoundTag == null) {
             return new byte[0];
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 13c4640ed8081abcbe5be4780aec7e6a2a94a85d..6b081485be5417e865ac90d97e0d00f785cb13ba 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -915,4 +915,17 @@ public class MultiPaperConnection extends Thread {
             e.printStackTrace();
         }
     }
+
+    public void willSaveChunk(String world, int cx, int cz) {
+        try {
+            DataOutputSender out = new DataOutputSender(this);
+            out.writeUTF("willSaveChunk");
+            out.writeUTF(world);
+            out.writeInt(cx);
+            out.writeInt(cz);
+            out.send(null);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
 }
