From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 19:54:53 +1000
Subject: [PATCH] Add shutdown packet handler


diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index 103956dbb88e6fc9b3dab0ec32b9fc5e0d992f6c..fd0ffcca8203d2af0efa5eba089c76f383af1769 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -410,4 +410,9 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
             }
         });
     }
+
+    @Override
+    public void handle(ShutdownMessage message) {
+        Bukkit.shutdown();
+    }
 }
