From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 21 Jan 2022 09:58:45 +1000
Subject: [PATCH] Reset mustNotSave on chunk load


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 878d28ace3bdac00fbe4296de99f56e30ff8d601..caaa58b3b7466b70cc8f3d0b72e5326dc905e08d 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -815,6 +815,7 @@ public class LevelChunk extends ChunkAccess {
     // CraftBukkit start
     public void loadCallback() {
         if (this.loadedTicketLevel) { LOGGER.error("Double calling chunk load!", new Throwable()); } // Paper
+        this.mustNotSave = false; // MultiPaper - Reset mustNotSave on chunk load
         // Paper start - neighbour cache
         int chunkX = this.chunkPos.x;
         int chunkZ = this.chunkPos.z;
