From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 16 Jan 2022 11:51:16 +1000
Subject: [PATCH] Lock the chunk when writting it


diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 9cb6a3ccd46d17620faa82c93e6bbca0b448fb48..eef93254d470523d250bcfe3df9afc82d9f30c09 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1540,6 +1540,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             asyncSaveData = ChunkSerializer.getAsyncSaveData(this.level, chunk);
         }
 
+        MultiPaper.willSaveChunk(level, chunk);
         this.level.asyncChunkTaskManager.scheduleChunkSave(chunkPos.x, chunkPos.z, com.destroystokyo.paper.io.PrioritizedTaskQueue.NORMAL_PRIORITY,
             asyncSaveData, chunk);
 
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 1184c9743be35f2b77d0663e2e5daecbc36ce50b..625eb330b91fb426ef8b894c6e17724f4eee300a 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -510,6 +510,10 @@ public class MultiPaper {
         chunk.hasExternalLockRequest = false;
     }
 
+    public static void willSaveChunk(ServerLevel level, ChunkAccess chunk) {
+        getConnection().send(new WillSaveChunkLaterMessage(level.getWorld().getName(), chunk.locX, chunk.locZ));
+    }
+
     public static byte[] nbtToBytes(CompoundTag compoundTag) throws IOException {
         if (compoundTag == null) {
             return new byte[0];
