From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 27 Jan 2022 22:43:56 +1000
Subject: [PATCH] Run tasks while we wait for the entity storage worker to
 close


diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
index 8d074ebb7c9c7e568f48e3475816ac2cde67dc4b..18504b91ff4f10e5f06dd2f0cca7fb4d41ff172f 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -125,7 +125,12 @@ public class EntityStorage implements EntityPersistentStorage<Entity> {
 
     @Override
     public void flush(boolean sync) {
-        this.worker.synchronize(sync).join();
+        // MultiPaper start - Run tasks while we wait for the entity storage worker to close
+        CompletableFuture<Void> completableFuture = this.worker.synchronize(sync);
+        if (!completableFuture.isDone()) {
+            level.chunkSource.mainThreadProcessor.managedBlock(completableFuture::isDone);
+        }
+        // MultiPaper end
         this.entityDeserializerQueue.runAll();
     }
 
