From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 21 Jan 2022 09:58:45 +1000
Subject: [PATCH] Reset mustNotSave on chunk load


diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 455ee65c8eaa77b315b6d81417c25553398f1faf..1039eb818689d88cc0a6122eaf01037c603a2787 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -867,6 +867,7 @@ public class LevelChunk extends ChunkAccess {
     // CraftBukkit start
     public void loadCallback() {
         if (this.loadedTicketLevel) { LOGGER.error("Double calling chunk load!", new Throwable()); } // Paper
+        this.mustNotSave = false; // MultiPaper - Reset mustNotSave on chunk load
         // Paper start - neighbour cache
         int chunkX = this.chunkPos.x;
         int chunkZ = this.chunkPos.z;
