From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 26 Jun 2022 16:01:21 +1000
Subject: [PATCH] Add option to not save lighting updates


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index 213d9a06c785d2c62d6f54ca06746daf781e3df2..a6c57e64c482ee47a4bffebe3087aaceee0e9d77 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -45,6 +45,7 @@ import net.minecraft.server.MinecraftServer;
 
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
+import puregero.multipaper.config.MultiPaperConfiguration;
 
 public class ChunkHolder {
 
@@ -261,6 +262,7 @@ public class ChunkHolder {
         LevelChunk chunk = this.chunkMap.level.getChunkSource().getChunkAtIfLoadedImmediately(this.pos.x, this.pos.z); // this.getSendingChunk(); // Paper - no-tick view distance // MultiPaper - if we modify this chunk, even if it's not sent to players, we still need to send it to other servers
 
         if (chunk != null) {
+            if (!MultiPaperConfiguration.get().optimizations.dontSaveJustForLightingUpdates) chunk.setUnsaved(true); // MultiPaper
             int i = this.levelHeightAccessor.getSectionIndex(pos.getY());
 
             if (i < 0 || i >= this.changedBlocksPerSection.length) return; // CraftBukkit - SPIGOT-6086, SPIGOT-6296
