From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 31 Jan 2022 13:35:42 +1000
Subject: [PATCH] Send crit animations as the entity being critted


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 76d87f6129c86783947f13e48e3df736ec321c65..04eb502a56bcb64d990086896c6c44be2979774a 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1829,12 +1829,12 @@ public class ServerPlayer extends Player {
 
     @Override
     public void crit(Entity target) {
-        this.getLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(target, 4));
+        this.getLevel().getChunkSource().broadcastAndSend(target, new ClientboundAnimatePacket(target, 4));
     }
 
     @Override
     public void magicCrit(Entity target) {
-        this.getLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(target, 5));
+        this.getLevel().getChunkSource().broadcastAndSend(target, new ClientboundAnimatePacket(target, 5));
     }
 
     @Override
