From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 31 Jan 2022 13:35:42 +1000
Subject: [PATCH] Send crit animations as the entity being critted


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d58d10f3b4c4cc3efc590d442f87042d05c1daed..461525e9cc6084963feef26210d790caa89bd0b7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1834,12 +1834,12 @@ public class ServerPlayer extends Player {
 
     @Override
     public void crit(Entity target) {
-        this.getLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(target, 4));
+        this.getLevel().getChunkSource().broadcastAndSend(target, new ClientboundAnimatePacket(target, 4));
     }
 
     @Override
     public void magicCrit(Entity target) {
-        this.getLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(target, 5));
+        this.getLevel().getChunkSource().broadcastAndSend(target, new ClientboundAnimatePacket(target, 5));
     }
 
     @Override
