From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 1 Feb 2022 00:41:27 +1000
Subject: [PATCH] Sync beacons


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index ee8443bdcdc5add432d4b92bf1a35394e3ac1401..34f21f17cde3184d30d3b36baf9333906273fa1d 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -134,6 +134,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider {
                         BeaconBlockEntity.this.secondaryPower = BeaconBlockEntity.getValidEffectById(value);
                 }
 
+                BeaconBlockEntity.this.setChanged(); // MultiPaper - send this block entity to other server
             }
 
             @Override
