From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:44:09 +1000
Subject: [PATCH] Tick sleepCounter for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index d4d54eefa1090302701bd1744b53785db1b56102..a93a68cbfae991356844cd9000071b6b2308d2e9 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -168,7 +168,7 @@ public abstract class Player extends LivingEntity {
     public double xCloak;
     public double yCloak;
     public double zCloak;
-    public int sleepCounter;
+    public int sleepCounter; public int getSleepCounter() { return this.sleepCounter; } public void setSleepCounter(int sleepCounter) { this.sleepCounter = sleepCounter; } // MultiPaper - add getter and setter
     protected boolean wasUnderwater;
     private final Abilities abilities = new Abilities();
     public int experienceLevel;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 8ee90cfc3b492794ff4c89962f6d87bf7eca0fb4..89d12724be748c9af28905669019103c786ed31d 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -91,6 +91,12 @@ public class MultiPaper {
                 }
 
                 externalPlayer.tickDeathIfDead();
+
+                if (externalPlayer.isSleeping()) {
+                    externalPlayer.setSleepCounter(Math.min(100, externalPlayer.getSleepCounter() + 1));
+                } else {
+                    externalPlayer.setSleepCounter(0);
+                }
             }
 
             player.syncExperience();
