From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 1 Feb 2022 00:41:27 +1000
Subject: [PATCH] Sync beacons


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index 1e7d14a021e870ae0126cd68a09ecfdcad2f203f..0269169a42563442d0a9799e295ef730d23fbe63 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -141,6 +141,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
                         BeaconBlockEntity.this.secondaryPower = BeaconBlockEntity.getValidEffectById(value);
                 }
 
+                BeaconBlockEntity.this.setChanged(); // MultiPaper - send this block entity to other server
             }
 
             @Override
