From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:48:07 +1000
Subject: [PATCH] Only wake up local players


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 0fc69756af9f35aec63b1c2f70bc0775cb54e710..970d29a9e7b188bd9aec4028497149b3270b6d11 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -997,7 +997,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     private void wakeUpAllPlayers() {
         this.sleepStatus.removeAllSleepers();
-        (this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList())).forEach((entityplayer) -> { // CraftBukkit - decompile error
+        (this.players.stream().filter(MultiPaper::isRealPlayer).filter(LivingEntity::isSleeping).collect(Collectors.toList())).forEach((entityplayer) -> { // CraftBukkit - decompile error // MultiPaper - only wake up local players
             entityplayer.stopSleepInBed(false, false);
         });
     }
