From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 23:35:02 +1000
Subject: [PATCH] Save lightning bolt data


diff --git a/src/main/java/net/minecraft/world/entity/LightningBolt.java b/src/main/java/net/minecraft/world/entity/LightningBolt.java
index a01a8ba27534d2b0d544423bc2c48fd29d0a6d64..6b96958b481c459bc21037fe34fd6501389d3b40 100644
--- a/src/main/java/net/minecraft/world/entity/LightningBolt.java
+++ b/src/main/java/net/minecraft/world/entity/LightningBolt.java
@@ -274,10 +274,34 @@ public class LightningBolt extends Entity {
     protected void defineSynchedData() {}
 
     @Override
-    protected void readAdditionalSaveData(CompoundTag nbt) {}
+    protected void readAdditionalSaveData(CompoundTag nbt) {
+        // MultiPaper start - save lightning bolt data
+        this.visualOnly = nbt.getBoolean("VisualOnly");
+        this.seed = nbt.getLong("Seed");
+        this.flashes = nbt.getInt("Flashes");
+        this.life = nbt.getInt("Life");
+        this.isSilent = nbt.getBoolean("IsSilent");
+
+        if (nbt.hasUUID("Cause")) {
+            this.cause = level.getServer().getPlayerList().getPlayer(nbt.getUUID("Cause"));
+        }
+        // MultiPaper end
+    }
 
     @Override
-    protected void addAdditionalSaveData(CompoundTag nbt) {}
+    protected void addAdditionalSaveData(CompoundTag nbt) {
+        // MultiPaper start - save lightning bolt data
+        nbt.putBoolean("VisualOnly", this.visualOnly);
+        nbt.putLong("Seed", this.seed);
+        nbt.putInt("Flashes", this.flashes);
+        nbt.putInt("Life", this.life);
+        nbt.putBoolean("IsSilent", this.isSilent);
+
+        if (this.cause != null) {
+            nbt.putUUID("Cause", this.cause.getUUID());
+        }
+        // MultiPaper end
+    }
 
     public int getBlocksSetOnFire() {
         return this.blocksSetOnFire;
