From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 5 Feb 2022 22:11:44 +1000
Subject: [PATCH] Take control of external chunks when extending pistons across
 server borders


diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 95d65d3ac2b8def54e0e309839021a8270854541..50b6b532331617ca6e06a16ed4274413bbb17006 100644
--- a/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -17,6 +17,7 @@ import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
@@ -41,6 +42,10 @@ import com.google.common.collect.ImmutableList;
 import org.bukkit.craftbukkit.block.CraftBlock;
 import org.bukkit.event.block.BlockPistonRetractEvent;
 import org.bukkit.event.block.BlockPistonExtendEvent;
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.mastermessagingprotocol.ChunkKey;
+import puregero.multipaper.mastermessagingprotocol.messages.masterbound.RequestChunkOwnershipMessage;
+import puregero.multipaper.mastermessagingprotocol.messages.serverbound.BooleanMessageReply;
 // CraftBukkit end
 
 public class PistonBaseBlock extends DirectionalBlock {
@@ -422,6 +427,9 @@ public class PistonBaseBlock extends DirectionalBlock {
                 aiblockdata[j++] = iblockdata1;
             }
 
+            HashSet<ChunkPos> takeControlOfChunks = new HashSet<>(); // MultiPaper
+            takeControlOfChunks.add(new ChunkPos(pos)); // MultiPaper
+
             for (k = list.size() - 1; k >= 0; --k) {
                 // Paper start - fix a variety of piston desync dupes
                 boolean allowDesync = io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPistonDuplication;
@@ -432,6 +440,8 @@ public class PistonBaseBlock extends DirectionalBlock {
                 map.remove(blockposition3);
                 BlockState iblockdata2 = (BlockState) Blocks.MOVING_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, dir);
 
+                takeControlOfChunks.add(new ChunkPos(blockposition3)); // MultiPaper
+
                 world.setBlock(blockposition3, iblockdata2, 68);
                 // Paper start - fix a variety of piston desync dupes
                 if (!allowDesync) {
@@ -448,6 +458,7 @@ public class PistonBaseBlock extends DirectionalBlock {
             }
 
             if (retract) {
+                takeControlOfChunks.add(new ChunkPos(blockposition1)); // MultiPaper
                 PistonType blockpropertypistontype = this.isSticky ? PistonType.STICKY : PistonType.DEFAULT;
                 BlockState iblockdata3 = (BlockState) ((BlockState) Blocks.PISTON_HEAD.defaultBlockState().setValue(PistonHeadBlock.FACING, dir)).setValue(PistonHeadBlock.TYPE, blockpropertypistontype);
 
@@ -458,6 +469,8 @@ public class PistonBaseBlock extends DirectionalBlock {
                 ((ServerLevel) world).getChunkSource().blockChanged(blockposition1); // MultiPaper
             }
 
+            takeControlOfChunks((ServerLevel) world, takeControlOfChunks); // MultiPaper
+
             BlockState iblockdata4 = Blocks.AIR.defaultBlockState();
             Iterator iterator = map.keySet().iterator();
 
@@ -505,6 +518,21 @@ public class PistonBaseBlock extends DirectionalBlock {
         }
     }
 
+    // MultiPaper start
+    private void takeControlOfChunks(ServerLevel level, HashSet<ChunkPos> takeControlOfChunks) {
+        if (takeControlOfChunks.stream().anyMatch(pos -> MultiPaper.isChunkExternal(MultiPaper.getChunkHolder(level, pos.x, pos.z)))
+                && takeControlOfChunks.stream().allMatch(pos -> MultiPaper.getChunkHolder(level, pos.x, pos.z).hasExternalLockRequest)) {
+            MultiPaper.getConnection().sendAndAwaitReply(
+                    new RequestChunkOwnershipMessage(
+                            level.getWorld().getName(),
+                            takeControlOfChunks.stream().map(pos -> new ChunkKey(level.getWorld().getName(), pos.x, pos.z)).toArray(ChunkKey[]::new)
+                    ),
+                    BooleanMessageReply.class
+            );
+        }
+    }
+    // MultiPaper end
+
     @Override
     public BlockState rotate(BlockState state, Rotation rotation) {
         return (BlockState) state.setValue(PistonBaseBlock.FACING, rotation.rotate((Direction) state.getValue(PistonBaseBlock.FACING)));
