From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 6 Jul 2022 16:15:42 +1000
Subject: [PATCH] Sync FireworkRocket's attachedToEntity


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index fca27f98989bf106060ba08196255fe32f850df5..ac435b09c59157fbcc6b282518a55bbed79f76aa 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -323,6 +323,11 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
             nbt.putUUID("SpawningEntity", this.spawningEntity);
         }
         // Paper end
+        // MultiPaper start
+        if (this.attachedToEntity != null) {
+            nbt.putUUID("AttachedToEntity", this.attachedToEntity.getUUID());
+        }
+        // MultiPaper end
     }
 
     @Override
@@ -344,6 +349,11 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
             this.spawningEntity = nbt.getUUID("SpawningEntity");
         }
         // Paper end
+        // MultiPaper start
+        if (nbt.hasUUID("AttachedToEntity") && level.getEntities().get(nbt.getUUID("AttachedToEntity")) instanceof LivingEntity livingEntity) {
+            this.attachedToEntity = livingEntity;
+        }
+        // MultiPaper end
     }
 
     @Override
