From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 23 Feb 2022 14:47:18 +1000
Subject: [PATCH] Fix infinite recursion in FishingHook


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index d99cdd7fdc9190f21b1558521487217e433511f7..2882af00c5dea7f1d0f2f157bd0a2ecb1805988c 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -604,6 +604,7 @@ public class FishingHook extends Projectile {
     @Override
     public void setOwner(@Nullable Entity entity) {
         super.setOwner(entity);
+        if (entity != null) // MultiPaper
         this.updateOwnerInfo(this);
     }
 
