From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 18 Feb 2022 21:58:13 +1000
Subject: [PATCH] Save Bukkit MapCanvas'


diff --git a/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java b/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java
index 01e4897e623ace1b18a4dc8559bb178ab0643cfc..7323a4018da98124de41dc396111e659242ce3a7 100644
--- a/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java
+++ b/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java
@@ -57,6 +57,7 @@ public class CraftMapCanvas implements MapCanvas {
             return;
         if (this.buffer[y * 128 + x] != color) {
             this.buffer[y * 128 + x] = color;
+            mapView.worldMap.setColor(x, y, color); // MultiPaper - save this map
             mapView.worldMap.setColorsDirty(x, y);
         }
     }
