From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 11 Feb 2022 00:13:18 +1000
Subject: [PATCH] Add --add-plugin-dir command line option


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a92f7b73270537fc1ec130cbc6d86d2aa0db06e8..e9e7a37ae9d1d50270daff2d2441e3b045d35946 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -464,6 +464,20 @@ public final class CraftServer implements Server {
             }
             list.add(file);
         }
+        // MultiPaper start
+        @SuppressWarnings("unchecked")
+        List<File> dirs = (List<File>) this.console.options.valuesOf("add-plugin-dir");
+        for (File dir : dirs) {
+            File[] files = dir.listFiles();
+            if (files != null) {
+                for (File file : files) {
+                    if (file.getName().endsWith(".jar")) {
+                        list.add(file);
+                    }
+                }
+            }
+        }
+        // MultiPaper end
         return list;
     }
     // Paper end
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index e1e51af894f4d6bbe9b1114c6d69b7abd92ea42d..7963cf74b51eda8c9e1c1c20a2517f3caa46af2b 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -164,6 +164,12 @@ public class Main {
                         .ofType(File.class)
                         .defaultsTo(new File("multipaper.yml"))
                         .describedAs("Yml file");
+
+                acceptsAll(asList("add-plugin-dir", "add-plugin-directory", "add-extra-plugin-jars"), "Specify paths of a directory containing extra plugin jars to be loaded in addition to those in the plugins folder. This argument can be specified multiple times, once for each extra plugin directory path.")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File[] {})
+                        .describedAs("Plugin directory");
                 // MultiPaper end
 
                 // Paper start
