From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 26 Feb 2022 18:39:19 +1000
Subject: [PATCH] Sync whitelist toggle


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 1a3ba902f46221484952f660c97c9d52cbfab175..bc01b3038666a3dd61c40d8f7051940f6169111a 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1373,6 +1373,7 @@ public abstract class PlayerList {
     public void setUsingWhiteList(boolean whitelistEnabled) {
         new com.destroystokyo.paper.event.server.WhitelistToggleEvent(whitelistEnabled).callEvent();
         this.doWhiteList = whitelistEnabled;
+        if (!WhitelistTogglePacket.updatingWhitelistToggle) MultiPaper.broadcastPacketToExternalServers(new WhitelistTogglePacket(whitelistEnabled)); // MultiPaper
     }
 
     public List<ServerPlayer> getPlayersWithAddress(String ip) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 4c8729bf65750df558df88b8c6540c02aa2b0da7..53950d191b5620811a8e7ff4d70b6fd3ed3eaf1d 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -62,6 +62,7 @@ public class ExternalServerPacketSerializer {
         addPacket(DifficultyUpdatePacket.class, DifficultyUpdatePacket::new);
         addPacket(SpawnUpdatePacket.class, SpawnUpdatePacket::new);
         addPacket(ScoreboardUpdatePacket.class, ScoreboardUpdatePacket::new);
+        addPacket(WhitelistTogglePacket.class, WhitelistTogglePacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/WhitelistTogglePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/WhitelistTogglePacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..ca0c69ffec95ee01c30eacfaebd7cd428bcd7782
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/WhitelistTogglePacket.java
@@ -0,0 +1,40 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.Difficulty;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+public class WhitelistTogglePacket extends ExternalServerPacket {
+
+    public static boolean updatingWhitelistToggle = false;
+
+    private final boolean whitelistEnabled;
+
+    public WhitelistTogglePacket(boolean whitelistEnabled) {
+        this.whitelistEnabled = whitelistEnabled;
+    }
+
+    public WhitelistTogglePacket(FriendlyByteBuf in) {
+        whitelistEnabled = in.readBoolean();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeBoolean(whitelistEnabled);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            updatingWhitelistToggle = true;
+            MinecraftServer.getServer().getPlayerList().setUsingWhiteList(whitelistEnabled);
+            updatingWhitelistToggle = false;
+        });
+    }
+}
