From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 27 Feb 2022 11:24:14 +1000
Subject: [PATCH] Update external player spam counters


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 7594aa45a02d8bcf7e2a2f8215f67fe529d62cf9..8ab1f3d4b5802bff0e1c3dd4c74421ebf379f906 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -371,20 +371,23 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         // Paper end
 
         this.server.getProfiler().pop();
-        // CraftBukkit start
-        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
-        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
-        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
-        /* Use thread-safe field access instead
-        if (this.chatSpamTickCount > 0) {
-            --this.chatSpamTickCount;
-        }
-        */
-        // CraftBukkit end
-
-        if (this.dropSpamTickCount > 0) {
-            --this.dropSpamTickCount;
-        }
+// MultiPaper start - move to reduceSpamCounters()
+//        // CraftBukkit start
+//        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
+//        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
+//        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
+//        /* Use thread-safe field access instead
+//        if (this.chatSpamTickCount > 0) {
+//            --this.chatSpamTickCount;
+//        }
+//        */
+//        // CraftBukkit end
+//
+//        if (this.dropSpamTickCount > 0) {
+//            --this.dropSpamTickCount;
+//        }
+        reduceSpamCounters();
+// MultiPaper end
 
         if (this.player.getLastActionTime() > 0L && this.server.getPlayerIdleTimeout() > 0 && Util.getMillis() - this.player.getLastActionTime() > (long) (this.server.getPlayerIdleTimeout() * 1000 * 60) && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits.
             this.player.resetLastActionTime(); // CraftBukkit - SPIGOT-854
@@ -402,6 +405,25 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         this.lastGoodZ = this.player.getZ();
     }
 
+    // MultiPaper start
+    public void reduceSpamCounters() {
+        // CraftBukkit start
+        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
+        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
+        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
+        /* Use thread-safe field access instead
+        if (this.chatSpamTickCount > 0) {
+            --this.chatSpamTickCount;
+        }
+        */
+        // CraftBukkit end
+
+        if (this.dropSpamTickCount > 0) {
+            --this.dropSpamTickCount;
+        }
+    }
+    // MultiPaper end
+
     @Override
     public Connection getConnection() {
         return this.connection;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 982533696d74323f44ac22d1b46e0c73c2e44339..0104baa1aad8d12102db7dad913f325b71dc43d0 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -101,6 +101,8 @@ public class MultiPaper {
             }
 
             player.syncExperience();
+
+            player.connection.reduceSpamCounters();
         }
 
         for (ExternalServer server : getConnection().getServersMap().values()) {
