From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 7 Dec 2022 19:55:45 +1000
Subject: [PATCH] Process transient entity chunks on the master


diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
index 6dca0c6ddf452d7b699fc630a1f07ba984d9104e..233917d2ea80d74f8dd2f1d2ae63565b09507e25 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
@@ -1962,7 +1962,7 @@ public final class NewChunkHolder {
                     return false;
                 }
                 try {
-                    mergeFrom = RegionFileIOThread.loadData(this.world, this.chunkX, this.chunkZ, RegionFileIOThread.RegionFileType.ENTITY_DATA, PrioritisedExecutor.Priority.BLOCKING);
+                    mergeFrom = new CompoundTag(); // mergeFrom = RegionFileIOThread.loadData(this.world, this.chunkX, this.chunkZ, RegionFileIOThread.RegionFileType.ENTITY_DATA, PrioritisedExecutor.Priority.BLOCKING); // MultiPaper - Process transient entity chunks on the master
                 } catch (final Exception ex) {
                     LOGGER.error("Cannot merge transient entities for chunk (" + this.chunkX + "," + this.chunkZ + ") in world '" + this.world.getWorld().getName() + "', data on disk will be replaced", ex);
                 }
@@ -1975,6 +1975,7 @@ public final class NewChunkHolder {
                     return false;
                 } else {
                     EntityStorage.copyEntities(mergeFrom, save);
+                    save.putBoolean("multipaper.transient", true); // MultiPaper - Process transient entity chunks on the master
                 }
             }
             if (save == null && this.lastEntitySaveNull) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 1c1b8721308421e326b78b1c0089a7a4ab223bc1..97e8ce0ed0c51b600e732e3d4e0c61231e8a7b08 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -399,7 +399,7 @@ public class MultiPaper {
     }
 
     public static void writeRegionFileNBT(String world, String path, int cx, int cz, CompoundTag compoundTag) throws IOException {
-        writeRegionFile(world, path, cx, cz, nbtToBytes(compoundTag));
+        writeRegionFile(world, path, cx, cz, nbtToBytes(compoundTag), compoundTag != null && compoundTag.contains("multipaper.transient"));
     }
 
     public static CompoundTag readLevel(String world) throws IOException {
