From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 8 Mar 2022 00:37:35 +1000
Subject: [PATCH] Only drop leads on the local server


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index f7bbed77796aad9cf95f38f223ee0dfe120aff51..ebc5086b6d0a131e39232c506dd52f2a522382c1 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -81,6 +81,7 @@ import org.bukkit.event.entity.EntityTargetEvent;
 import org.bukkit.event.entity.EntityTransformEvent;
 import org.bukkit.event.entity.EntityUnleashEvent;
 import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public abstract class Mob extends LivingEntity {
@@ -1422,7 +1423,7 @@ public abstract class Mob extends LivingEntity {
         if (this.leashHolder != null) {
             this.leashHolder = null;
             this.leashInfoTag = null;
-            if (!this.level.isClientSide && dropItem) {
+            if (!this.level.isClientSide && dropItem && MultiPaper.isChunkLocal(level.getChunkIfLoaded(blockPosition))) { // MultiPaper - only drop leads on the local server
                 this.forceDrops = true; // CraftBukkit
                 this.spawnAtLocation((ItemLike) Items.LEAD);
                 this.forceDrops = false; // CraftBukkit
