From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 5 Mar 2022 21:59:06 +1000
Subject: [PATCH] Add null check in LegacyStructureDataHandler


diff --git a/src/main/java/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java b/src/main/java/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
index e8c334866c6e4962fefffd19928c6c39eb9f5c90..4f872b2d942f26cb579ac08bfce576a8fd8e2d76 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -178,7 +178,11 @@ public class LegacyStructureDataHandler {
                 CompoundTag nbttagcompound = new CompoundTag();
 
                 try {
-                    nbttagcompound = persistentStateManager.readTagFromDisk(s, 1493).getCompound("data").getCompound("Features");
+                    // MultiPaper start
+                    CompoundTag tag = persistentStateManager.readTagFromDisk(s, 1493);
+                    if (tag == null) continue;
+                    nbttagcompound = tag.getCompound("data").getCompound("Features");
+                    // MultiPaper end
                     if (nbttagcompound.isEmpty()) {
                         continue;
                     }
