From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 29 Jan 2023 16:33:09 +0100
Subject: [PATCH] Fix Ikea V1 dupe


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c6a6fa3e9727d4eb9ced84e9fb83324036a2bff3..97164492cd6550a1e34a40e737be2dc0ca5fbb79 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3191,6 +3191,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     return entity1 != entity;
                 }).collect(ImmutableList.toImmutableList());
             }
+            // MultiPaper start
+            if (entity instanceof ServerPlayer player) {
+                player.closeContainer();
+            }
+            // MultiPaper end
 
             entity.boardingCooldown = 60;
             this.gameEvent(GameEvent.ENTITY_DISMOUNT, entity);
