From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 8 Mar 2022 00:37:35 +1000
Subject: [PATCH] Only drop leads on the local server


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index dfd30400bcecd4b2e917c417374c2478e6aea01f..8de8bb2fa7e23a13a12d7cc1f4b519a099c2c9bb 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -84,6 +84,7 @@ import org.bukkit.event.entity.EntityTargetEvent;
 import org.bukkit.event.entity.EntityTransformEvent;
 import org.bukkit.event.entity.EntityUnleashEvent;
 import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public abstract class Mob extends LivingEntity {
@@ -1443,7 +1444,7 @@ public abstract class Mob extends LivingEntity {
         if (this.leashHolder != null) {
             this.leashHolder = null;
             this.leashInfoTag = null;
-            if (!this.level.isClientSide && dropItem) {
+            if (!this.level.isClientSide && dropItem && MultiPaper.isChunkLocal(level.getChunkIfLoaded(blockPosition))) { // MultiPaper - only drop leads on the local server
                 this.forceDrops = true; // CraftBukkit
                 this.spawnAtLocation((ItemLike) Items.LEAD);
                 this.forceDrops = false; // CraftBukkit
