From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 27 Feb 2022 19:21:10 +1000
Subject: [PATCH] Don't double save the json lists


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
index 009d0db589b47de973774d2eed40e863513300d5..0def15cd99b77370cdd8576585b8e11b5feb7e8a 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -46,13 +46,13 @@ public class DedicatedPlayerList extends PlayerList {
     @Override
     public void op(GameProfile profile) {
         super.op(profile);
-        this.saveOps();
+        // this.saveOps(); // MultiPaper - don't double save the op list
     }
 
     @Override
     public void deop(GameProfile profile) {
         super.deop(profile);
-        this.saveOps();
+        // this.saveOps(); // MultiPaper - don't double save the op list
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
index 61cb647b2aa590303402e6652bd37b5bca0e0b1d..1f3f373d59b1c72aa05136d3279aec17d1c5ed51 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
@@ -41,11 +41,13 @@ public class CraftIpBanList implements org.bukkit.BanList {
 
         this.list.add(entry);
 
-        try {
-            this.list.save();
-        } catch (IOException ex) {
-            Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-ips.json, {0}", ex.getMessage());
-        }
+        // MultiPaper start - Don't double save the ban list
+        // try {
+        //     this.list.save();
+        // } catch (IOException ex) {
+        //     Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-ips.json, {0}", ex.getMessage());
+        // }
+        // MultiPaper end
 
         return new CraftIpBanEntry(target, entry, this.list);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanEntry.java b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanEntry.java
index 126968d557263b3efddc9d53d0f8f436a401cf78..b7425e88609efe31112426c8c86dc85d0496d6e1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanEntry.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanEntry.java
@@ -78,10 +78,12 @@ public final class CraftProfileBanEntry implements org.bukkit.BanEntry {
     public void save() {
         UserBanListEntry entry = new UserBanListEntry(this.profile, this.created, this.source, this.expiration, this.reason);
         this.list.add(entry);
-        try {
-            this.list.save();
-        } catch (IOException ex) {
-            Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-players.json, {0}", ex.getMessage());
-        }
+        // MultiPaper start - Don't double save the ban list
+        // try {
+        //     this.list.save();
+        // } catch (IOException ex) {
+        //     Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-players.json, {0}", ex.getMessage());
+        // }
+        // MultiPaper end
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
index 6e2047f1b45799c7a1ddfdcd24d76e6e12f91e4b..66e693404478b33819531a95b55978bb3a1c6860 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
@@ -54,11 +54,13 @@ public class CraftProfileBanList implements org.bukkit.BanList {
 
         this.list.add(entry);
 
-        try {
-            this.list.save();
-        } catch (IOException ex) {
-            Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-players.json, {0}", ex.getMessage());
-        }
+        // MultiPaper start - Don't double save the ban list
+        // try {
+        //     this.list.save();
+        // } catch (IOException ex) {
+        //     Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-players.json, {0}", ex.getMessage());
+        // }
+        // MultiPaper end
 
         return new CraftProfileBanEntry(profile, entry, this.list);
     }
