From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 29 Jan 2023 16:33:09 +0100
Subject: [PATCH] Fix Ikea V1 dupe


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 8bf2d99247302f55dbdc721356dd4f688c145b45..69c8d12db796e71382cbb9bab8ab7020a01f7694 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3021,6 +3021,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     return entity1 != entity;
                 }).collect(ImmutableList.toImmutableList());
             }
+            // MultiPaper start
+            if (entity instanceof ServerPlayer player) {
+                player.closeContainer();
+            }
+            // MultiPaper end
 
             entity.boardingCooldown = 60;
         }
