From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 29 Jan 2023 16:33:09 +0100
Subject: [PATCH] Fix Ikea V1 dupe


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index b6318b4fe594938755b4a6cfa7da5a3b771c031d..73da6159165acc5ec74cc4b2f12b3295b7ded889 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3034,6 +3034,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     return entity1 != entity;
                 }).collect(ImmutableList.toImmutableList());
             }
+            // MultiPaper start
+            if (entity instanceof ServerPlayer player) {
+                player.closeContainer();
+            }
+            // MultiPaper end
 
             entity.boardingCooldown = 60;
             this.gameEvent(GameEvent.ENTITY_DISMOUNT, entity);
