From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Mon, 6 Feb 2023 00:02:10 +0400
Subject: [PATCH] load chunk at owner for pets who owner is external player


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
index 4fe177c950c769f5c6e76b522019b5d5b78259a5..03513a82a33a8d6598b981149a040f6fbc2fc85b 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -16,7 +16,9 @@ import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 // CraftBukkit start
 import org.bukkit.Location;
 import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.entity.CraftPlayer; // MultiPaper
 import org.bukkit.event.entity.EntityTeleportEvent;
+import puregero.multipaper.ExternalPlayer; // MultiPaper
 // CraftBukkit end
 
 public class FollowOwnerGoal extends Goal {
@@ -145,6 +147,13 @@ public class FollowOwnerGoal extends Goal {
     }
 
     private boolean canTeleportTo(BlockPos pos) {
+        // MultiPaper start - load chunks at owner for pets who owner is external player.
+        if (owner instanceof ExternalPlayer externalPlayer) {
+            // use bukkit api to load that chunk aysnc
+            CraftPlayer player = externalPlayer.getBukkitEntity();
+            player.getWorld().getChunkAtAsync(player.getLocation());
+        }
+        // MultiPaper end
         BlockPathTypes pathtype = WalkNodeEvaluator.getBlockPathTypeStatic(this.level, pos.mutable());
 
         if (pathtype != BlockPathTypes.WALKABLE) {
