From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 21:41:32 +1000
Subject: [PATCH] Cache ChunkHolder's getAvailableChunkNow


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index 7cf6ac74bfb35cf7f7624d10d367451329c5fb52..d0ddbd0117e781cd6ae15632d747dc50831e3cd3 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -73,6 +73,7 @@ public class ChunkHolder {
     private boolean wasAccessibleSinceLastSave;
     private boolean resendLight;
     private CompletableFuture<Void> pendingFullStateConfirmation;
+    @Nullable private ChunkAccess cachedAvailableChunkNow; // MultiPaper
 
     public ServerLevel getWorld() { return chunkMap.level; } // Paper
     boolean isUpdateQueued = false; // Paper
@@ -176,16 +177,17 @@ public class ChunkHolder {
 
     // Paper start
     public ChunkAccess getAvailableChunkNow() {
-        // TODO can we just getStatusFuture(EMPTY)?
-        for (ChunkStatus curr = ChunkStatus.FULL, next = curr.getParent(); curr != next; curr = next, next = next.getParent()) {
-            CompletableFuture<Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure>> future = this.getFutureIfPresentUnchecked(curr);
-            Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure> either = future.getNow(null);
-            if (either == null || !either.left().isPresent()) {
-                continue;
+        // MultiPaper start
+        return cachedAvailableChunkNow;
+    }
+
+    public void setAvailableChunkNow(Either<? extends ChunkAccess, ChunkLoadingFailure> either) {
+        either.ifLeft(chunkAccess -> {
+            if (cachedAvailableChunkNow == null || chunkAccess.getStatus().isOrAfter(cachedAvailableChunkNow.getStatus())) {
+                cachedAvailableChunkNow = chunkAccess;
             }
-            return either.left().get();
-        }
-        return null;
+        });
+        // MultiPaper end
     }
     // Paper end
 
@@ -440,6 +442,7 @@ public class ChunkHolder {
 
             this.updateChunkToSave(completablefuture1, "schedule " + targetStatus);
             this.futures.set(i, completablefuture1);
+            completablefuture1.thenAccept(this::setAvailableChunkNow);
             return completablefuture1;
         } else {
             return completablefuture == null ? ChunkHolder.UNLOADED_CHUNK_FUTURE : completablefuture;
@@ -632,6 +635,7 @@ public class ChunkHolder {
                     this.chunkMap.distanceManager.clearPriorityTickets(pos);
                 }
             });
+            this.fullChunkFuture.thenAccept(this::setAvailableChunkNow); // MultiPaper
             this.updateChunkToSave(this.fullChunkFuture, "full");
         }
 
@@ -660,6 +664,7 @@ public class ChunkHolder {
                 });
             });
             // Paper end
+            this.tickingChunkFuture.thenAccept(this::setAvailableChunkNow); // MultiPaper
             this.updateChunkToSave(this.tickingChunkFuture, "ticking");
         }
 
@@ -695,6 +700,7 @@ public class ChunkHolder {
                 });
             });
             // Paper end
+            this.entityTickingChunkFuture.thenAccept(this::setAvailableChunkNow); // MultiPaper
             this.updateChunkToSave(this.entityTickingChunkFuture, "entity ticking");
         }
 
@@ -821,6 +827,8 @@ public class ChunkHolder {
         }
 
         this.updateChunkToSave(CompletableFuture.completedFuture(Either.left(chunk.getWrapped())), "replaceProto");
+
+        cachedAvailableChunkNow = chunk; // MultiPaper
     }
 
     // MultiPaper start
@@ -840,6 +848,8 @@ public class ChunkHolder {
         }
 
         this.updateChunkToSave(CompletableFuture.completedFuture(Either.left(protoChunk)), "replaceProtoWithProto");
+
+        cachedAvailableChunkNow = protoChunk;
     }
 
     public CompletableFuture<Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure>> getLastAvailableChunkFuture() {
