From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 12 Jan 2022 21:41:32 +1000
Subject: [PATCH] Cache ChunkHolder's getAvailableChunkNow


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index ae957d932fb9f9df54ac53475778b38ce5394420..0f01607f78dfbc72e22c4d913dbe130c5bc13df6 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -76,6 +76,7 @@ public class ChunkHolder {
     private boolean wasAccessibleSinceLastSave;
     private boolean resendLight;
     private CompletableFuture<Void> pendingFullStateConfirmation;
+    @Nullable private ChunkAccess cachedAvailableChunkNow; // MultiPaper
 
     public ServerLevel getWorld() { return chunkMap.level; } // Paper
     boolean isUpdateQueued = false; // Paper
@@ -166,16 +167,17 @@ public class ChunkHolder {
 
     // Paper start
     public @Nullable ChunkAccess getAvailableChunkNow() {
-        // TODO can we just getStatusFuture(EMPTY)?
-        for (ChunkStatus curr = ChunkStatus.FULL, next = curr.getParent(); curr != next; curr = next, next = next.getParent()) {
-            CompletableFuture<Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure>> future = this.getFutureIfPresentUnchecked(curr);
-            Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure> either = future.getNow(null);
-            if (either == null || either.left().isEmpty()) {
-                continue;
+        // MultiPaper start
+        return cachedAvailableChunkNow;
+    }
+
+    public void setAvailableChunkNow(Either<? extends ChunkAccess, ChunkLoadingFailure> either) {
+        either.ifLeft(chunkAccess -> {
+            if (cachedAvailableChunkNow == null || chunkAccess.getStatus().isOrAfter(cachedAvailableChunkNow.getStatus())) {
+                cachedAvailableChunkNow = chunkAccess;
             }
-            return either.left().get();
-        }
-        return null;
+        });
+        // MultiPaper end
     }
     // Paper end
     // CraftBukkit start
@@ -465,6 +467,7 @@ public class ChunkHolder {
 
             this.updateChunkToSave(completablefuture1, "schedule " + targetStatus);
             this.futures.set(i, completablefuture1);
+            completablefuture1.thenAccept(this::setAvailableChunkNow);
             return completablefuture1;
         } else {
             return completablefuture == null ? ChunkHolder.UNLOADED_CHUNK_FUTURE : completablefuture;
@@ -657,6 +660,7 @@ public class ChunkHolder {
                     this.chunkMap.distanceManager.clearPriorityTickets(pos);
                 }
             });
+            this.fullChunkFuture.thenAccept(this::setAvailableChunkNow); // MultiPaper
             this.updateChunkToSave(this.fullChunkFuture, "full");
         }
 
@@ -685,6 +689,7 @@ public class ChunkHolder {
                 });
             });
             // Paper end
+            this.tickingChunkFuture.thenAccept(this::setAvailableChunkNow); // MultiPaper
             this.updateChunkToSave(this.tickingChunkFuture, "ticking");
         }
 
@@ -720,6 +725,7 @@ public class ChunkHolder {
                 });
             });
             // Paper end
+            this.entityTickingChunkFuture.thenAccept(this::setAvailableChunkNow); // MultiPaper
             this.updateChunkToSave(this.entityTickingChunkFuture, "entity ticking");
         }
 
@@ -846,6 +852,8 @@ public class ChunkHolder {
         }
 
         this.updateChunkToSave(CompletableFuture.completedFuture(Either.left(chunk.getWrapped())), "replaceProto");
+
+        cachedAvailableChunkNow = chunk; // MultiPaper
     }
 
     public List<Pair<ChunkStatus, CompletableFuture<Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure>>>> getAllFutures() {
@@ -875,6 +883,8 @@ public class ChunkHolder {
         }
 
         this.updateChunkToSave(CompletableFuture.completedFuture(Either.left(protoChunk)), "replaceProtoWithProto");
+
+        cachedAvailableChunkNow = protoChunk;
     }
 
     public CompletableFuture<Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure>> getLastAvailableChunkFuture() {
