From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Wed, 8 Feb 2023 23:38:47 +0400
Subject: [PATCH] fixup! Hurt external entities


diff --git a/src/main/java/puregero/multipaper/util/DamageSourceSerializer.java b/src/main/java/puregero/multipaper/util/DamageSourceSerializer.java
index b4d194d70cceacd9ef795ece2a40bdc00a403258..e47a8737bb897d91cbd1850545f126448ae8bb44 100644
--- a/src/main/java/puregero/multipaper/util/DamageSourceSerializer.java
+++ b/src/main/java/puregero/multipaper/util/DamageSourceSerializer.java
@@ -172,7 +172,7 @@ public class DamageSourceSerializer {
             case "indirectMagic" -> DamageSource.indirectMagic(entities[0], entities[1]);
 //            case "mob" -> DamageSource.indirectMobAttack(entities[0], (LivingEntity) entities[1]); // Handled above
             case "player" -> DamageSource.playerAttack((Player) entities[0]);
-            case "badRespawnPoint" -> DamageSource.badRespawnPointExplosion(sourcePosition);
+            case "badRespawnPoint" -> DamageSource.badRespawnPointExplosion(sourcePosition, null);
             case "fireball" -> DamageSource.fireball((Fireball) entities[0], entities[1]);
             case "sonic_boom" -> DamageSource.sonicBoom(entities[0]);
             default -> throw new IOException("Unknown damage cause msgId of " + msgId);
