From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Sun, 19 Feb 2023 16:10:33 +0400
Subject: [PATCH] add fake entity api


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index cd4dea302b567e5ac317bd70dae49f108ad48a95..2e81c1f1d9e1466636fbe0f230d145c246635f6f 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -59,6 +59,7 @@ import net.minecraft.network.protocol.game.VecDeltaCodec;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.obfuscate.DontObfuscate;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import io.papermc.paper.util.MCUtil;
@@ -432,7 +433,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean collidingWithWorldBorder; // Paper
     public boolean fixedPose = false; // Paper
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
-
+    public volatile boolean fake = false; // MultiPaper - Fake entity nms
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
         this.originWorld = location.getWorld().getUID();
@@ -5163,4 +5164,19 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return false;
     }
     // Purpur end
+    // MultiPaper start - Fake entity nms
+    @DontObfuscate
+    public boolean isFake() {
+        return this.fake;
+    }
+
+    @DontObfuscate
+    public void setFake() {
+        // we need to make sure that actual players are not set fake
+        if (this instanceof ServerPlayer serverPlayer && getServer().getPlayerList().players.contains(serverPlayer)) {
+            throw new IllegalStateException("Online player cannot be set as fake entity");
+        }
+        this.fake = true;
+    }
+    // MultiPaper end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index ffda74205a95d638a399dd11eae0857703428f3a..bee79e4d4fa3bb6a2501934e6bc47db445690a39 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1572,4 +1572,15 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return !getHandle().dismountsUnderwater();
     }
     // Purpur end
+    // MultiPaper start - add fake entity api
+    @Override
+    public void setFake() {
+        this.getHandle().setFake();
+    }
+
+    @Override
+    public boolean isFake() {
+        return this.getHandle().isFake();
+    }
+    // MultiPaper stop
 }
diff --git a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
index f7085a42ec7320cfba7758c3eee72252e5c42590..ca2ce74fee9400679d991bbe4f410318369afca2 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
@@ -351,7 +351,7 @@ public class MultiPaperEntitiesHandler {
     }
 
     public static boolean shouldSyncEntity(Entity entity) {
-        return entity.shouldBeSaved() || entity instanceof LightningBolt || entity instanceof LeashFenceKnotEntity;
+        return !entity.isFake() && (entity.shouldBeSaved() || entity instanceof LightningBolt || entity instanceof LeashFenceKnotEntity);
     }
 
     public static void onEntityRemove(Entity entity, Entity.RemovalReason reason) {
diff --git a/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java
index 6b212628a7c4e4bc9d3458fc7a79d88bbf866c47..54126110ba25f6ac16bfe55b19c0e39d3217da3b 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java
@@ -17,6 +17,7 @@ public class MultiPaperEntityInteractHandler {
     private static final Logger LOGGER = LogManager.getLogger(MultiPaperEntityInteractHandler.class.getSimpleName());
 
     public static ExternalServerConnection getOwner(Entity entity) {
+        if (entity.isFake()) return null;
         NewChunkHolder newChunkHolder = MultiPaper.getChunkHolder(entity);
 
         if (MultiPaperEntitiesHandler.getControllingPassenger(entity) instanceof ExternalPlayer externalPlayer) {
@@ -29,6 +30,7 @@ public class MultiPaperEntityInteractHandler {
     }
 
     public static boolean handleEntityInteract(ServerPlayer player, Entity entity, ServerboundInteractPacket packet) {
+        if (entity.isFake()) return false;
         ExternalServerConnection owner = getOwner(entity);
         if (owner != null) {
             owner.send(new PlayerActionOnEntityPacket(player, entity, packet));
@@ -40,7 +42,7 @@ public class MultiPaperEntityInteractHandler {
 
     public static boolean touchEntity(Player player, Entity entity) {
         NewChunkHolder newChunkHolder = MultiPaper.getChunkHolder(entity);
-
+        if (entity.isFake()) return false;
         if (MultiPaper.isRealPlayer(entity)) {
             return false;
         } else if (MultiPaper.isChunkExternal(newChunkHolder) && !(entity instanceof FishingHook)) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/RequestEntityPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/RequestEntityPacket.java
index 26defc673cbb481f3d60eefd8fdabc346ea2753e..01a07a3ef3d414f8bfd557e6a924bfc98660d06f 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/RequestEntityPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/RequestEntityPacket.java
@@ -79,6 +79,7 @@ public class RequestEntityPacket extends ExternalServerPacket {
             ServerLevel level = craftWorld.getHandle();
             Entity entity = level.getEntityLookup().getEntityIgnoringAccessible(uuid);
             if (entity != null) {
+                if (entity.isFake()) return;
                 if (entity instanceof ServerPlayer serverPlayer) {
                     triedToRequestPlayer(connection, serverPlayer);
                     return;
