From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Wed, 8 Feb 2023 23:39:18 +0400
Subject: [PATCH] Disable Mojang's profile call in MultiPaperChunkHandler


diff --git a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
index 09e4266bbdc06450ca72108dd2db357bc1e192b2..301b6c9daa3da6091b9f27ae36af815788dbd61b 100644
--- a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
@@ -218,9 +218,9 @@ public class MultiPaperChunkHandler {
         holder.vanillaChunkHolder.blockChanged(pos);
 
         if (oldState != null && blockState != oldState && (blockState.getLightBlock(chunk, pos) != oldState.getLightBlock(chunk, pos) || blockState.getLightEmission() != oldState.getLightEmission() || blockState.useShapeForLightOcclusion() || oldState.useShapeForLightOcclusion())) {
-            chunk.level.getProfiler().push("queueCheckLightExternalUpdate");
+            //chunk.level.getProfiler().push("queueCheckLightExternalUpdate");
             chunk.level.getChunkSource().getLightEngine().checkBlock(pos);
-            chunk.level.getProfiler().pop();
+            //chunk.level.getProfiler().pop();
         }
     }
 
