From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Fri, 17 Feb 2023 20:58:00 +0400
Subject: [PATCH] fix enderpearls teleport on external entities


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index f35547c6b5951bc6eb4df74b2a94496fd20d69b5..0e91114aa88a0bff1e5df80f137618c0287428e1 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -20,6 +20,7 @@ import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.event.player.PlayerTeleportEvent;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public class ThrownEnderpearl extends ThrowableItemProjectile {
@@ -51,7 +52,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
             this.level.addParticle(ParticleTypes.PORTAL, this.getX(), this.getY() + this.random.nextDouble() * 2.0D, this.getZ(), this.random.nextGaussian(), 0.0D, this.random.nextGaussian());
         }
 
-        if (!this.level.isClientSide && !this.isRemoved()) {
+        if (!this.level.isClientSide && !this.isRemoved() && MultiPaper.isChunkLocal((ServerLevel) level, blockPosition))  { // MultiPaper - Fix ender pearls on external entities
             Entity entity = this.getOwner();
 
             if (entity instanceof ServerPlayer) {
