From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Mar 2022 16:48:05 +1000
Subject: [PATCH] Only touch tridents that we are ticking


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
index a367f50b0e3fe9e7a1b87892a8c98e88bd678f6f..4c55a3905987609f504072c7ce063a072e54e810 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -22,6 +22,7 @@ import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
+import puregero.multipaper.MultiPaper;
 
 public class ThrownTrident extends AbstractArrow {
 
@@ -181,6 +182,7 @@ public class ThrownTrident extends AbstractArrow {
 
     @Override
     public void playerTouch(Player player) {
+        if (MultiPaper.isChunkLocal(level.getChunkIfLoaded(blockPosition))) // MultiPaper - ensure we're still ticking this entity
         if (this.ownedBy(player) || this.getOwner() == null) {
             super.playerTouch(player);
         }
