From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 18 Mar 2022 01:43:33 +1000
Subject: [PATCH] Add ability to pardon floodgate players


diff --git a/src/main/java/net/minecraft/server/commands/PardonCommand.java b/src/main/java/net/minecraft/server/commands/PardonCommand.java
index f1605ec163f6d178df1b8203ff16fbe97cd53475..900d8b6b1b55b0e0f4ea833a0b3f6f4380133598 100644
--- a/src/main/java/net/minecraft/server/commands/PardonCommand.java
+++ b/src/main/java/net/minecraft/server/commands/PardonCommand.java
@@ -2,16 +2,21 @@ package net.minecraft.server.commands;
 
 import com.mojang.authlib.GameProfile;
 import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import java.util.Collection;
+import java.util.Collections;
+
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.SharedSuggestionProvider;
 import net.minecraft.commands.arguments.GameProfileArgument;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.players.UserBanList;
+import net.minecraft.server.players.UserBanListEntry;
 
 public class PardonCommand {
     private static final SimpleCommandExceptionType ERROR_NOT_BANNED = new SimpleCommandExceptionType(Component.translatable("commands.pardon.failed"));
@@ -19,13 +24,38 @@ public class PardonCommand {
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(Commands.literal("pardon").requires((source) -> {
             return source.hasPermission(3);
-        }).then(Commands.argument("targets", GameProfileArgument.gameProfile()).suggests((context, builder) -> {
+        }).then(Commands.argument("targets", new BanListGameProfileArgument()).suggests((context, builder) -> { // MultiPaper - also get gameprofiles from the ban list
             return SharedSuggestionProvider.suggest(context.getSource().getServer().getPlayerList().getBans().getUserList(), builder);
         }).executes((context) -> {
             return pardonPlayers(context.getSource(), GameProfileArgument.getGameProfiles(context, "targets"));
         })));
     }
 
+    // MultiPaper start - also get gameprofiles from the ban list
+    public static class BanListGameProfileArgument extends GameProfileArgument {
+        @Override
+        public GameProfileArgument.Result parse(StringReader stringReader) throws CommandSyntaxException {
+            int cursor = stringReader.getCursor();
+
+            while(stringReader.canRead() && stringReader.peek() != ' ') {
+                stringReader.skip();
+            }
+
+            String username = stringReader.getString().substring(cursor, stringReader.getCursor());
+
+            for (UserBanListEntry entry : MinecraftServer.getServer().getPlayerList().getBans().getEntries()) {
+                if (entry.getUser() != null && entry.getUser().getName() != null && entry.getUser().getName().equalsIgnoreCase(username)) {
+                    return (source) -> Collections.singleton(entry.getUser());
+                }
+            }
+
+            stringReader.setCursor(cursor);
+
+            return super.parse(stringReader);
+        }
+    }
+    // MultiPaper end
+
     private static int pardonPlayers(CommandSourceStack source, Collection<GameProfile> targets) throws CommandSyntaxException {
         UserBanList userBanList = source.getServer().getPlayerList().getBans();
         int i = 0;
