From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Mar 2022 01:01:01 +1000
Subject: [PATCH] Temporarily disable piercing to prevent crashes


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index e7ef36dac559d03d127cf45373a7e0dc935b80a8..b6e57569511303a33feb9ec3f5141dac9f9194d4 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -244,7 +244,7 @@ public abstract class AbstractArrow extends Projectile {
                     this.hasImpulse = true;
                 }
 
-                if (movingobjectpositionentity == null || this.getPierceLevel() <= 0) {
+                if (true || movingobjectpositionentity == null || this.getPierceLevel() <= 0) { // MultiPaper - temporarily disable piercing to prevent crashes
                     break;
                 }
 
