From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Mar 2022 01:01:01 +1000
Subject: [PATCH] Temporarily disable piercing to prevent crashes


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 9788e477ff1446ad2ea3669922cc7dfc09900ce8..543124bac5722b27bee25bd4e87b8d4b258dbad1 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -233,7 +233,7 @@ public abstract class AbstractArrow extends Projectile {
                     this.hasImpulse = true;
                 }
 
-                if (movingobjectpositionentity == null || this.getPierceLevel() <= 0) {
+                if (true || movingobjectpositionentity == null || this.getPierceLevel() <= 0) { // MultiPaper - temporarily disable piercing to prevent crashes
                     break;
                 }
 
