From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 26 Apr 2024 14:39:23 +0900
Subject: [PATCH] Only save local players


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c3462753df53ee197b4a002cbff21d538932420f..e1d1ce61e588908635b963672d41f12ac6349359 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1353,8 +1353,8 @@ public abstract class PlayerList {
         //MinecraftTimings.savePlayers.startTiming(); // Paper // Purpur
         int numSaved = 0;
         long now = MinecraftServer.currentTick;
-        for (int i = 0; i < this.players.size(); ++i) {
-            ServerPlayer entityplayer = this.players.get(i);
+        for (int i = 0; i < this.localPlayers.size(); ++i) {
+            ServerPlayer entityplayer = this.localPlayers.get(i);
             if (interval == -1 || now - entityplayer.lastSave >= interval) {
                 this.save(entityplayer);
                 if (interval != -1 && ++numSaved >= io.papermc.paper.configuration.GlobalConfiguration.get().playerAutoSave.maxPerTick()) { break; }
