From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 2 Apr 2022 23:20:07 +1000
Subject: [PATCH] Reset horse inventory during load


diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 821b88d0c32f946af9af1c2e211b38446fab4ecb..23eef06c0badc9b677bce0f657f852f20ac7cba8 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -865,6 +865,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
     @Override
     public void readAdditionalSaveData(CompoundTag nbt) {
         super.readAdditionalSaveData(nbt);
+        this.inventory.clearContent(); // MultiPaper - reset inventory before loading data
         this.setEating(nbt.getBoolean("EatingHaystack"));
         this.setBred(nbt.getBoolean("Bred"));
         this.setTemper(nbt.getInt("Temper"));
