From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 19 Mar 2022 01:38:18 +1000
Subject: [PATCH] Open certain block menus on the player's local server


diff --git a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
index 05c46f3b3bce5225b819d86e6e06729a5093e092..c79e2d5c167a03152a53584fa50fadd8eb043a85 100644
--- a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -37,13 +37,7 @@ import net.minecraft.world.level.EmptyBlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
-import net.minecraft.world.level.block.Block;
-import net.minecraft.world.level.block.EntityBlock;
-import net.minecraft.world.level.block.Mirror;
-import net.minecraft.world.level.block.RenderShape;
-import net.minecraft.world.level.block.Rotation;
-import net.minecraft.world.level.block.SoundType;
-import net.minecraft.world.level.block.SupportType;
+import net.minecraft.world.level.block.*;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntityTicker;
 import net.minecraft.world.level.block.entity.BlockEntityType;
@@ -65,6 +59,8 @@ import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import puregero.multipaper.ExternalPlayer;
+import puregero.multipaper.externalserverprotocol.PlayerUseBlockPacket;
 
 public abstract class BlockBehaviour {
 
@@ -975,6 +971,21 @@ public abstract class BlockBehaviour {
         }
 
         public InteractionResult use(Level world, Player player, InteractionHand hand, BlockHitResult hit) {
+            // MultiPaper start - Open these menus on the player's server
+            if ((player instanceof ExternalPlayer externalPlayer) && (
+                    this.getBlock() instanceof CraftingTableBlock ||
+                    this.getBlock() instanceof EnderChestBlock ||
+                    this.getBlock() instanceof AnvilBlock ||
+                    this.getBlock() instanceof LoomBlock ||
+                    this.getBlock() instanceof FletchingTableBlock ||
+                    this.getBlock() instanceof CartographyTableBlock ||
+                    this.getBlock() instanceof GrindstoneBlock ||
+                    this.getBlock() instanceof SmithingTableBlock ||
+                    this.getBlock() instanceof StonecutterBlock
+            )) {
+                externalPlayer.externalServerConnection.send(new PlayerUseBlockPacket(player, hit.getBlockPos()));
+            }
+            // MultiPaper end
             return this.getBlock().use(this.asState(), world, hit.getBlockPos(), player, hand, hit);
         }
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 9127dab7b5431f637a4dcb5a0657da60665eab0d..cc9ef4a5291b9858d954e06bace7b0166d7cb99c 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -67,6 +67,7 @@ public class ExternalServerPacketSerializer {
         addPacket(DestroyAndAckPacket.class, DestroyAndAckPacket::new);
         addPacket(EntityTeleportPacket.class, EntityTeleportPacket::new);
         addPacket(ProjectileHitEntityPacket.class, ProjectileHitEntityPacket::new);
+        addPacket(PlayerUseBlockPacket.class, PlayerUseBlockPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerUseBlockPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerUseBlockPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..eceb57019cb450c9bb230fe1cdb986d7c5555f21
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerUseBlockPacket.java
@@ -0,0 +1,55 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.Vec3;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerUseBlockPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(PlayerUseBlockPacket.class.getSimpleName());
+
+    private final UUID uuid;
+    private final BlockPos pos;
+
+    public PlayerUseBlockPacket(Player player, BlockPos blockPos) {
+        this.uuid = player.getUUID();
+        this.pos = blockPos;
+    }
+
+    public PlayerUseBlockPacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+        pos = BlockPos.of(in.readLong());
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeLong(pos.asLong());
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            if (player == null) {
+                LOGGER.warn("Tried to use a block " + pos + " with a non-existent player of uuid " + uuid);
+                return;
+            }
+
+            player.level.getBlockState(pos).use(player.level, player, InteractionHand.MAIN_HAND, new BlockHitResult(new Vec3(pos.getX(), pos.getY(), pos.getZ()), Direction.NORTH, pos, false));
+        });
+    }
+}
