From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Apr 2022 13:08:48 +1000
Subject: [PATCH] Tick death for external players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 4247199e81dd16dfb80b4505da507d165c7c86d3..18252c3a121d58fe7b004a255a3a1251c38fd6f9 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -625,6 +625,14 @@ public abstract class LivingEntity extends Entity {
         return false;
     }
 
+    // MultiPaper
+    public void tickDeathIfDead() {
+        if (this.isDeadOrDying() && this.level.shouldTickDeath(this)) {
+            this.tickDeath();
+        }
+    }
+    // MultiPaper end
+
     protected void tickDeath() {
         ++this.deathTime;
         if (this.deathTime >= 20 && !this.level.isClientSide() && !this.isRemoved()) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 0251479932cf25af6f34ddf168fdf7b5a80f8c00..668555367671c566c2b6e8d7d2abbf05be6397be 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -89,6 +89,8 @@ public class MultiPaper {
                 if (player.takeXpDelay > 0) {
                     --player.takeXpDelay;
                 }
+
+                externalPlayer.tickDeathIfDead();
             }
 
             player.syncExperience();
