From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Mar 2022 01:01:01 +1000
Subject: [PATCH] Temporarily disable piercing to prevent crashes


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 53d0024daf6963ac4dab575666b0d6a74a39a958..f2f178b49cf435d4a46aa9c4de37281b3f030467 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -242,7 +242,7 @@ public abstract class AbstractArrow extends Projectile {
                     this.hasImpulse = true;
                 }
 
-                if (movingobjectpositionentity == null || this.getPierceLevel() <= 0) {
+                if (true || movingobjectpositionentity == null || this.getPierceLevel() <= 0) { // MultiPaper - temporarily disable piercing to prevent crashes
                     break;
                 }
 
