From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 6 Apr 2022 20:30:56 +1000
Subject: [PATCH] A view distance of 1 also works


diff --git a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
index 3eff1a562ad1bfa866a09d1e0464bb82e9cb36fe..fbf9596bfd9aa5a0206923da20f8dbb8b69d8f2d 100644
--- a/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/PlayerChunkLoader.java
@@ -32,7 +32,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 
 public final class PlayerChunkLoader {
 
-    public static final int MIN_VIEW_DISTANCE = 2;
+    public static final int MIN_VIEW_DISTANCE = 1; // MultiPaper - minimum can actually be 1
     public static final int MAX_VIEW_DISTANCE = 32;
 
     public static final int TICK_TICKET_LEVEL = 31;
@@ -227,7 +227,7 @@ public final class PlayerChunkLoader {
     }
 
     public void setTickDistance(final int distance) {
-        if (distance < MIN_VIEW_DISTANCE || distance > MAX_VIEW_DISTANCE) {
+        if (distance < MIN_VIEW_DISTANCE || distance > MAX_VIEW_DISTANCE) { // MultiPaper - minimum is actually 1. This line is unchanged by MultiPaper, but is marked to track changes to the minimum value
             throw new IllegalArgumentException("View distance must be a number between " + MIN_VIEW_DISTANCE + " and " + MAX_VIEW_DISTANCE + ", got: " + distance);
         }
         this.rawTickDistance = distance;
