From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:54:41 +1000
Subject: [PATCH] Don't compare items if they're the same item


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2ff6c8a5c8782e1bee8571181268b3e777f0d922..9c2b7dbeec9ba26935e305e7f0be720a59171c1c 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -317,12 +317,12 @@ public abstract class Player extends LivingEntity {
 
         ItemStack itemstack = this.getMainHandItem();
 
-        if (!ItemStack.matches(this.lastItemInMainHand, itemstack)) {
+        if (this.lastItemInMainHand != itemstack && !ItemStack.matches(this.lastItemInMainHand, itemstack)) { // MultiPaper - don't compare items if they're the same item
             if (!ItemStack.isSame(this.lastItemInMainHand, itemstack)) {
                 this.resetAttackStrengthTicker();
             }
 
-            this.lastItemInMainHand = itemstack.copy();
+            this.lastItemInMainHand = itemstack; // MultiPaper - don't compare items if they're the same item
         }
 
         this.turtleHelmetTick();
