From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Apr 2022 13:34:05 +1000
Subject: [PATCH] Fix lectern quick move


diff --git a/src/main/java/net/minecraft/world/inventory/LecternMenu.java b/src/main/java/net/minecraft/world/inventory/LecternMenu.java
index 52471ef87994ac91b4b0017159b61ff50774848a..bd19e47c04ba8988a070b890b611e371badbcd4f 100644
--- a/src/main/java/net/minecraft/world/inventory/LecternMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/LecternMenu.java
@@ -120,6 +120,13 @@ public class LecternMenu extends AbstractContainerMenu {
         }
     }
 
+    // Paper start
+    @Override
+    public ItemStack quickMoveStack(net.minecraft.world.entity.player.Player player, int index) {
+        return ItemStack.EMPTY;
+    }
+    // Paper end
+
     @Override
     public ItemStack quickMoveStack(net.minecraft.world.entity.player.Player player, int index) {
         return ItemStack.EMPTY;
