From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 21 Mar 2022 12:51:41 +1000
Subject: [PATCH] Add getAllOnlinePlayers and getLocalOnlinePlayers


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 49d695001e5137abd11363e6ff8c884483313598..4a2cc403fe31b4541843c2c87ec3c2f9f20c48e2 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -114,6 +114,7 @@ public abstract class PlayerList {
     private static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z");
     private final MinecraftServer server;
     public final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
+    public final List<ServerPlayer> localPlayers = new java.util.concurrent.CopyOnWriteArrayList(); // MultiPaper
     private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap();
     private final UserBanList bans;
     private final IpBanList ipBans;
@@ -349,6 +350,7 @@ public abstract class PlayerList {
 
         playerconnection.teleport(player.getX(), player.getY(), player.getZ(), player.getYRot(), player.getXRot());
         this.players.add(player);
+        if (!(player instanceof ExternalPlayer)) this.localPlayers.add(player); // MultiPaper
         this.playersByName.put(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT), player); // Spigot
         this.playersByUUID.put(player.getUUID(), player);
         // this.broadcastAll(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER, new EntityPlayer[]{entityplayer})); // CraftBukkit - replaced with loop below
@@ -673,6 +675,7 @@ public abstract class PlayerList {
         worldserver.removePlayerImmediately(entityplayer, Entity.RemovalReason.UNLOADED_WITH_PLAYER);
         entityplayer.getAdvancements().stopListening();
         this.players.remove(entityplayer);
+        if (!(entityplayer instanceof ExternalPlayer)) this.localPlayers.remove(entityplayer); // MultiPaper
         this.playersByName.remove(entityplayer.getScoreboardName().toLowerCase(java.util.Locale.ROOT)); // Spigot
         this.server.getCustomBossEvents().onPlayerDisconnect(entityplayer);
         UUID uuid = entityplayer.getUUID();
@@ -854,6 +857,7 @@ public abstract class PlayerList {
         // MultiPaper end
         entityplayer.stopRiding(); // CraftBukkit
         this.players.remove(entityplayer);
+        if (!(entityplayer instanceof ExternalPlayer)) this.localPlayers.remove(entityplayer); // MultiPaper
         this.playersByName.remove(entityplayer.getScoreboardName().toLowerCase(java.util.Locale.ROOT)); // Spigot
         entityplayer.getLevel().removePlayerImmediately(entityplayer, Entity.RemovalReason.DISCARDED);
         BlockPos blockposition = entityplayer.getRespawnPosition();
@@ -989,6 +993,7 @@ public abstract class PlayerList {
         if (!entityplayer.connection.isDisconnected()) {
             worldserver1.addRespawnedPlayer(entityplayer1);
             this.players.add(entityplayer1);
+            if (!(entityplayer1 instanceof ExternalPlayer)) this.localPlayers.add(entityplayer1); // MultiPaper
             this.playersByName.put(entityplayer1.getScoreboardName().toLowerCase(java.util.Locale.ROOT), entityplayer1); // Spigot
             this.playersByUUID.put(entityplayer1.getUUID(), entityplayer1);
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 68f6664fb4b5edb68a873c7a0de0674d868b9856..f62d1294351313b0cb7e6f47c0dca107ac511165 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -279,6 +279,7 @@ public final class CraftServer implements Server {
     private final Object2IntOpenHashMap<SpawnCategory> spawnCategoryLimit = new Object2IntOpenHashMap<>();
     private final CraftMultiPaperNotificationManager multipaperNotificationManager = new CraftMultiPaperNotificationManager(); // MultiPaper
     private File container;
+    private Thread inTabCompletion; // MultiPaper
     private WarningState warningState = WarningState.DEFAULT;
     public String minimumAPI;
     public CraftScoreboardManager scoreboardManager;
@@ -288,6 +289,7 @@ public final class CraftServer implements Server {
     private boolean overrideAllCommandBlockCommands = false;
     public boolean ignoreVanillaPermissions = false;
     private final List<CraftPlayer> playerView;
+    private final List<CraftPlayer> localPlayerView; // Multipaper
     public int reloadCount;
     private final io.papermc.paper.datapack.PaperDatapackManager datapackManager; // Paper
     public static Exception excessiveVelEx; // Paper - Velocity warnings
@@ -310,6 +312,14 @@ public final class CraftServer implements Server {
                 return player.getBukkitEntity();
             }
         }));
+        // MultiPaper start
+        this.localPlayerView = Collections.unmodifiableList(Lists.transform(playerList.localPlayers, new Function<ServerPlayer, CraftPlayer>() {
+            @Override
+            public CraftPlayer apply(ServerPlayer player) {
+                return player.getBukkitEntity();
+            }
+        }));
+        // MultiPaper end
         this.serverVersion = CraftServer.class.getPackage().getImplementationVersion();
         this.structureManager = new CraftStructureManager(console.getStructureManager());
 
@@ -612,10 +622,28 @@ public final class CraftServer implements Server {
     }
     // Paper end
 
+    // MultiPaper start
     @Override
     public List<CraftPlayer> getOnlinePlayers() {
+        if (inTabCompletion != null && inTabCompletion == Thread.currentThread()) {
+            // If we're tab-completing, we probably want all the players
+            // instead of just the ones on our server.
+            return getAllOnlinePlayers();
+        } else {
+            return getLocalOnlinePlayers();
+        }
+    }
+
+    @Override
+    public List<CraftPlayer> getLocalOnlinePlayers() {
+        return this.localPlayerView;
+    }
+
+    @Override
+    public List<CraftPlayer> getAllOnlinePlayers() {
         return this.playerView;
     }
+    // MultiPaper end
 
     @Override
     @Deprecated
@@ -630,7 +658,7 @@ public final class CraftServer implements Server {
 
         String lowerName = name.toLowerCase(java.util.Locale.ENGLISH);
         int delta = Integer.MAX_VALUE;
-        for (Player player : this.getOnlinePlayers()) {
+        for (Player player : this.getAllOnlinePlayers()) { // MultiPaper
             if (player.getName().toLowerCase(java.util.Locale.ENGLISH).startsWith(lowerName)) {
                 int curDelta = Math.abs(player.getName().length() - lowerName.length());
                 if (curDelta < delta) {
@@ -677,7 +705,7 @@ public final class CraftServer implements Server {
 
         List<Player> matchedPlayers = new ArrayList<Player>();
 
-        for (Player iterPlayer : this.getOnlinePlayers()) {
+        for (Player iterPlayer : this.getAllOnlinePlayers()) { // MultiPaper
             String iterPlayerName = iterPlayer.getName();
 
             if (partialName.equalsIgnoreCase(iterPlayerName)) {
@@ -2073,7 +2101,7 @@ public final class CraftServer implements Server {
             }
         }
 
-        players.addAll(this.getOnlinePlayers());
+        players.addAll(this.getAllOnlinePlayers()); // MultiPaper
 
         return players.toArray(new OfflinePlayer[players.size()]);
     }
@@ -2244,6 +2272,8 @@ public final class CraftServer implements Server {
             return ImmutableList.of();
         }
 
+        inTabCompletion = Thread.currentThread(); // MultiPaper
+
         List<String> offers;
         Player player = (Player) sender;
         if (message.startsWith("/") || forceCommand) {
@@ -2255,6 +2285,8 @@ public final class CraftServer implements Server {
         TabCompleteEvent tabEvent = new TabCompleteEvent(player, message, offers, message.startsWith("/") || forceCommand, pos != null ? net.minecraft.server.MCUtil.toLocation(((CraftWorld) player.getWorld()).getHandle(), new BlockPos(pos)) : null); // Paper
         this.getPluginManager().callEvent(tabEvent);
 
+        inTabCompletion = null; // MultiPaper
+
         return tabEvent.isCancelled() ? Collections.EMPTY_LIST : tabEvent.getCompletions();
     }
 
@@ -2289,7 +2321,7 @@ public final class CraftServer implements Server {
         List<String> completions = new ArrayList<String>();
         PlayerChatTabCompleteEvent event = new PlayerChatTabCompleteEvent(player, message, completions);
         String token = event.getLastToken();
-        for (Player p : this.getOnlinePlayers()) {
+        for (Player p : this.getAllOnlinePlayers()) { // MultiPaper
             if (player.canSee(p) && StringUtil.startsWithIgnoreCase(p.getName(), token)) {
                 completions.add(p.getName());
             }
@@ -2778,14 +2810,14 @@ public final class CraftServer implements Server {
 
         @Override
         public void broadcast(BaseComponent component) {
-            for (Player player : CraftServer.this.getOnlinePlayers()) {
+            for (Player player : CraftServer.this.getAllOnlinePlayers()) { // MultiPaper
                 player.spigot().sendMessage(component);
             }
         }
 
         @Override
         public void broadcast(BaseComponent... components) {
-            for (Player player : CraftServer.this.getOnlinePlayers()) {
+            for (Player player : CraftServer.this.getAllOnlinePlayers()) { // MultiPaper
                 player.spigot().sendMessage(components);
             }
         }
diff --git a/src/main/java/puregero/multipaper/MultiPaperPermissionSyncer.java b/src/main/java/puregero/multipaper/MultiPaperPermissionSyncer.java
index 5c52e8351d82e799b383f44906c1c74bd0c5e8d9..27c4d5fce8aff4303b371464b8a13b488d62e261 100644
--- a/src/main/java/puregero/multipaper/MultiPaperPermissionSyncer.java
+++ b/src/main/java/puregero/multipaper/MultiPaperPermissionSyncer.java
@@ -25,7 +25,7 @@ public class MultiPaperPermissionSyncer {
             return;
         }
 
-        for (Player player : Bukkit.getOnlinePlayers()) {
+        for (Player player : Bukkit.getAllOnlinePlayers()) {
             syncers.computeIfAbsent(player, MultiPaperPermissionSyncer::new);
         }
 
diff --git a/src/main/java/puregero/multipaper/commands/SListCommand.java b/src/main/java/puregero/multipaper/commands/SListCommand.java
index 2b9d8875353e9353929d85ec528eee31bfdd3e03..474bc0197f440b96344ec4796dec24a8477d35c6 100644
--- a/src/main/java/puregero/multipaper/commands/SListCommand.java
+++ b/src/main/java/puregero/multipaper/commands/SListCommand.java
@@ -20,7 +20,7 @@ public class SListCommand extends Command {
     public boolean execute(CommandSender sender, String commandLabel, String[] args) {
         if (!testPermission(sender)) return false;
 
-        sender.sendMessage(ChatColor.WHITE + "There are " + Bukkit.getOnlinePlayers().size() + " out of " + Bukkit.getMaxPlayers() + " players online");
+        sender.sendMessage(ChatColor.WHITE + "There are " + Bukkit.getAllOnlinePlayers().size() + " out of " + Bukkit.getMaxPlayers() + " players online");
 
         for (ExternalServer server : MultiPaper.getConnection().getServersMap().values()) {
             String name = ChatColor.GREEN + "[" + server.getName() + "] ";
