From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:37:10 +1000
Subject: [PATCH] Only run PlayerArmorChangeEvent for local players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index f4c0b8e504924b55029c2b2d27798e7d7362790e..0a7ec5d5d660d318bfae530fb86e5ea6872248ab 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3137,7 +3137,7 @@ public abstract class LivingEntity extends Entity {
 
             if (!ItemStack.matches(itemstack1, itemstack)) {
                 // Paper start - PlayerArmorChangeEvent
-                if (this instanceof ServerPlayer && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) {
+                if (MultiPaper.isRealPlayer(this) && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) { // MultiPaper - only run PlayerArmorChangeEvent for local players
                     final org.bukkit.inventory.ItemStack oldItem = CraftItemStack.asBukkitCopy(itemstack);
                     final org.bukkit.inventory.ItemStack newItem = CraftItemStack.asBukkitCopy(itemstack1);
                     new PlayerArmorChangeEvent((Player) this.getBukkitEntity(), PlayerArmorChangeEvent.SlotType.valueOf(enumitemslot.name()), oldItem, newItem).callEvent();
