From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:54:41 +1000
Subject: [PATCH] Don't compare items if they're the same item


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 7f5c2dd786258a7b47ef785afe6c715085132445..488d14fca7c973dcf90480cc86baa8d307df264d 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -322,12 +322,12 @@ public abstract class Player extends LivingEntity {
 
         ItemStack itemstack = this.getMainHandItem();
 
-        if (!ItemStack.matches(this.lastItemInMainHand, itemstack)) {
+        if (this.lastItemInMainHand != itemstack && !ItemStack.matches(this.lastItemInMainHand, itemstack)) { // MultiPaper - don't compare items if they're the same item
             if (!ItemStack.isSameIgnoreDurability(this.lastItemInMainHand, itemstack)) {
                 this.resetAttackStrengthTicker();
             }
 
-            this.lastItemInMainHand = itemstack.copy();
+            this.lastItemInMainHand = itemstack; // MultiPaper - don't compare items if they're the same item
         }
 
         this.turtleHelmetTick();
