From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 14 Apr 2022 23:24:51 +1000
Subject: [PATCH] Check if player is still connected after
 PlayerSpawnLocationEvent


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 53aa26213db2ef9d6c6601bc60fb98dc848fcd0b..52748bdec17e30396c8492e2a6e329fa3af02c84 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -259,11 +259,19 @@ public abstract class PlayerList {
         }
         // Paper end
 
-        // Spigot start - spawn location event
+       // Spigot start - spawn location event
         Player spawnPlayer = player.getBukkitEntity();
         org.spigotmc.event.player.PlayerSpawnLocationEvent ev = new com.destroystokyo.paper.event.player.PlayerInitialSpawnEvent(spawnPlayer, spawnPlayer.getLocation()); // Paper use our duplicate event
         this.cserver.getPluginManager().callEvent(ev);
 
+        // MultiPaper start
+        if (!connection.isConnected() || player.quitReason != null) {
+            MultiPaper.sendPlayerDisconnect(player);
+            pendingPlayers.remove(player.getUUID(), player);
+            return;
+        }
+        // MultiPaper end
+
         Location loc = ev.getSpawnLocation();
         worldserver1 = ((CraftWorld) loc.getWorld()).getHandle();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 32c0bedaf3a264dd8341d327297d3bd8043515d6..5bfc23b170cf4ed3d2f1d117af934a3a56e3d6ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -28,6 +28,8 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.adventure.PaperAdventure;
 import net.minecraft.Util;
 import net.minecraft.advancements.AdvancementProgress;
 import net.minecraft.core.BlockPos;
@@ -519,6 +521,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         final ServerGamePacketListenerImpl connection = this.getHandle().connection;
         if (connection != null) {
             connection.disconnect(message == null ? net.kyori.adventure.text.Component.empty() : message, cause);
+        // MultiPaper start - kick even without a connection
+        } else {
+            final Component ichatbasecomponent = PaperAdventure.asVanilla(message);
+            this.getHandle().quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.KICKED;
+            this.getHandle().networkManager.send(new ClientboundDisconnectPacket(ichatbasecomponent), (future) -> {
+                this.getHandle().networkManager.disconnect(ichatbasecomponent);
+            });
+        // MultiPaper end
         }
     }
 
