From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 20 Apr 2022 07:26:42 +0200
Subject: [PATCH] Fix an error when player's boat is removed in a plugin.


diff --git a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
index 4e6d5a872afc0c7a8e6035d51a89c0118d6d4b05..d2efb9259e81479e147174715037974164260745 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
@@ -41,6 +41,8 @@ public class MultiPaperEntitiesHandler {
     public static boolean takingItem = false;
 
     public static Entity getControllingPassenger(Entity entity) {
+        if (entity == null) return null;
+
         if (entity instanceof AbstractMinecart) {
             // Because Minecarts are special and the passenger doesn't 'technically' control them
             return entity.getFirstPassenger();
