From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 20 Apr 2022 07:26:42 +0200
Subject: [PATCH] Fix an error when player's boat is removed in a plugin.


diff --git a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
index 71c7e0719b1412153162b396cd5d68234625140b..51fa74637f70c1428774b654294aa5bd84e74f4f 100644
--- a/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperEntitiesHandler.java
@@ -40,6 +40,8 @@ public class MultiPaperEntitiesHandler {
     public static boolean takingItem = false;
 
     public static Entity getControllingPassenger(Entity entity) {
+        if (entity == null) return null;
+
         if (entity instanceof AbstractMinecart) {
             // Because Minecarts are special and the passenger doesn't 'technically' control them
             return entity.getFirstPassenger();
