From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 18 Apr 2022 21:46:09 +1000
Subject: [PATCH] Check if chunk is loaded when broadcasting changes


diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index d13f258946fb9156eea4dc0388540bb932e221ad..2d5ce36e0ca77d436dabc9ebdcc7fab09e5e586e 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -910,7 +910,7 @@ public class ServerChunkCache extends ChunkSource {
                 ReferenceOpenHashSet<ChunkHolder> copy = this.chunkMap.needsChangeBroadcasting.clone();
                 this.chunkMap.needsChangeBroadcasting.clear();
                 for (ChunkHolder holder : copy) {
-                    holder.broadcastChanges(holder.getFullChunkNowUnchecked()); // LevelChunks are NEVER unloaded
+                    Optional.ofNullable(holder.getFullChunkNowUnchecked()).ifPresent(holder::broadcastChanges); // LevelChunks are NEVER unloaded // MultiPaper - Add null check. MultiPaper might not remove an unloaded chunk from needsChangeBroadcasting, and getFullChunkNowUnchecked may be null
                     if (holder.needsBroadcastChanges()) {
                         // I DON'T want to KNOW what DUMB plugins might be doing.
                         this.chunkMap.needsChangeBroadcasting.add(holder);
