From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 14 Apr 2022 23:24:51 +1000
Subject: [PATCH] Check if player is still connected after
 PlayerSpawnLocationEvent


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 4a2cc403fe31b4541843c2c87ec3c2f9f20c48e2..844e9831f19b887ded86b0075cb2d94ddcf1020e 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -246,11 +246,19 @@ public abstract class PlayerList {
         }
         // Paper end
 
-        // Spigot start - spawn location event
+       // Spigot start - spawn location event
         Player spawnPlayer = player.getBukkitEntity();
         org.spigotmc.event.player.PlayerSpawnLocationEvent ev = new com.destroystokyo.paper.event.player.PlayerInitialSpawnEvent(spawnPlayer, spawnPlayer.getLocation()); // Paper use our duplicate event
         this.cserver.getPluginManager().callEvent(ev);
 
+        // MultiPaper start
+        if (!connection.isConnected() || player.quitReason != null) {
+            MultiPaper.sendPlayerDisconnect(player);
+            pendingPlayers.remove(player.getUUID(), player);
+            return;
+        }
+        // MultiPaper end
+
         Location loc = ev.getSpawnLocation();
         worldserver1 = ((CraftWorld) loc.getWorld()).getHandle();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index aabb67f2706a4eaceca0ff0beac02dab9f2d8811..bd4651123153237c2d77b7fb2f61cd0c9c006ff0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -30,11 +30,14 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import javax.annotation.Nullable;
+import io.papermc.paper.adventure.PaperAdventure;
+import net.minecraft.Util;
 import net.minecraft.advancements.AdvancementProgress;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.PacketSendListener;
 import net.minecraft.network.chat.ChatMessageContent;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.PlayerChatMessage;
@@ -563,6 +566,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         final ServerGamePacketListenerImpl connection = this.getHandle().connection;
         if (connection != null) {
             connection.disconnect(message == null ? net.kyori.adventure.text.Component.empty() : message, cause);
+        // MultiPaper start - kick even without a connection
+        } else {
+            final Component ichatbasecomponent = PaperAdventure.asVanilla(message);
+            this.getHandle().quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.KICKED;
+            this.getHandle().networkManager.send(new ClientboundDisconnectPacket(ichatbasecomponent), PacketSendListener.thenRun(() -> {
+                this.getHandle().networkManager.disconnect(ichatbasecomponent);
+            }));
+        // MultiPaper end
         }
     }
 
