From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 28 Apr 2022 21:49:48 +1000
Subject: [PATCH] Add useLocalPlayerCountForServerIsFullKick


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 52748bdec17e30396c8492e2a6e329fa3af02c84..ba62a386ea88ce3e9301a8922930a0b7183bc265 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -115,6 +115,7 @@ import org.bukkit.event.player.PlayerRespawnEvent;
 import puregero.multipaper.ExternalPlayer;
 import puregero.multipaper.ExternalServerConnection;
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperConfig;
 import puregero.multipaper.externalserverprotocol.*;
 // CraftBukkit end
 
@@ -807,7 +808,8 @@ public abstract class PlayerList {
             event.disallow(PlayerLoginEvent.Result.KICK_BANNED, PaperAdventure.asAdventure(chatmessage)); // Paper - Adventure
         } else {
             // return this.players.size() >= this.maxPlayers && !this.canBypassPlayerLimit(gameprofile) ? new ChatMessage("multiplayer.disconnect.server_full") : null;
-            if (this.players.size() >= this.maxPlayers && !this.canBypassPlayerLimit(gameprofile)) {
+            boolean isFull = MultiPaperConfig.useLocalPlayerCountForServerIsFullKick ? this.localPlayers.size() >= this.maxPlayers : this.players.size() >= this.maxPlayers; // MultiPaper - add isFull
+            if (isFull && !this.canBypassPlayerLimit(gameprofile)) { // MultiPaper - add isFull
                 event.disallow(PlayerLoginEvent.Result.KICK_FULL, PaperAdventure.LEGACY_SECTION_UXRC.deserialize(org.spigotmc.SpigotConfig.serverFullMessage)); // Spigot // Paper - Adventure
             }
         }
