From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 1 Jun 2022 14:01:39 +1000
Subject: [PATCH] Add support for proxy protocol


diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index 7fc7f967ff3cad2615ee2f45c5d29ea992296f07..f70a5cb955ef9682a60cd7816ed93e381ca6a59e 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -1,10 +1,12 @@
 package puregero.multipaper;
 
 import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.socket.SocketChannel;
 import io.netty.handler.codec.compression.Zstd;
+import io.netty.handler.codec.haproxy.*;
 import io.netty.util.internal.SystemPropertyUtil;
 import net.minecraft.network.*;
 import net.minecraft.network.protocol.Packet;
@@ -26,6 +28,7 @@ import java.io.IOException;
 import java.util.*;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 
 public class ExternalServerConnection extends ChannelInitializer<SocketChannel> implements Closeable {
@@ -69,6 +72,23 @@ public class ExternalServerConnection extends ChannelInitializer<SocketChannel>
     }
 
     public void sendMinecraftHandshake(String address, String secret, int port) {
+        // MultiPaper start - Add support for Proxy Protocol
+        if (io.papermc.paper.configuration.GlobalConfiguration.get() == null) {
+            // Paper config hasn't been loaded yet, try again 100ms later
+            CompletableFuture.delayedExecutor(100, TimeUnit.MILLISECONDS).execute(() -> sendMinecraftHandshake(address, secret, port));
+            return;
+        }
+
+        ChannelFuture future2 = channel.newSucceededFuture();
+
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.proxyProtocol) {
+            channel.pipeline().addLast("haproxy-encoder", HAProxyMessageEncoder.INSTANCE);
+            future2 = channel.writeAndFlush(new HAProxyMessage(HAProxyProtocolVersion.V2, HAProxyCommand.LOCAL, HAProxyProxiedProtocol.UNKNOWN, null, null, 0, 0));
+        }
+
+        future2.addListener(future1 -> {
+        // MultiPaper end
+
         channel.pipeline()
                 .addLast("prepender", new MessageLengthEncoder())
                 .addLast("encoder", new PacketEncoder(PacketFlow.SERVERBOUND));
@@ -81,6 +101,8 @@ public class ExternalServerConnection extends ChannelInitializer<SocketChannel>
                         future.cause().printStackTrace();
                     }
                 });
+
+        }); // MultiPaper - Add support for Proxy Protocol
     }
 
     public void setupPipeline() {
