From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:44:09 +1000
Subject: [PATCH] Tick sleepCounter for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index b26d9e27bd8900d4e5a7c34b8d5e47c3220dbe8e..3ec9fe6fcc0e6b245f1e9c87154019b5153c875d 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -168,7 +168,7 @@ public abstract class Player extends LivingEntity {
     public double xCloak;
     public double yCloak;
     public double zCloak;
-    public int sleepCounter;
+    public int sleepCounter; public int getSleepCounter() { return this.sleepCounter; } public void setSleepCounter(int sleepCounter) { this.sleepCounter = sleepCounter; } // MultiPaper - add getter and setter
     protected boolean wasUnderwater;
     private final Abilities abilities = new Abilities();
     public int experienceLevel;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 81ceb952d198feea780afdc74b8a6789141d0f6a..108854c636c824874e3c7f2dbcfcf2c9636fa1c1 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -91,6 +91,12 @@ public class MultiPaper {
                 }
 
                 externalPlayer.tickDeathIfDead();
+
+                if (externalPlayer.isSleeping()) {
+                    externalPlayer.setSleepCounter(Math.min(100, externalPlayer.getSleepCounter() + 1));
+                } else {
+                    externalPlayer.setSleepCounter(0);
+                }
             }
 
             player.syncExperience();
