From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 26 Jun 2022 16:01:21 +1000
Subject: [PATCH] Add option to not save lighting updates


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index e8347b625be246a03e7e4ca34b393ccbc15ed604..05b0ebe8fead031dd742a1344580f4140e72b576 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -43,6 +43,7 @@ import net.minecraft.server.MinecraftServer;
 
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
+import puregero.multipaper.MultiPaperConfig;
 
 public class ChunkHolder {
 
@@ -288,6 +289,7 @@ public class ChunkHolder {
         LevelChunk chunk = this.getSendingChunk(); // Paper - no-tick view distance
 
         if (chunk != null) {
+            if (!MultiPaperConfig.dontSaveLightingUpdates) chunk.setUnsaved(true); // MultiPaper
             int i = this.levelHeightAccessor.getSectionIndex(pos.getY());
 
             if (i < 0 || i >= this.changedBlocksPerSection.length) return; // CraftBukkit - SPIGOT-6086, SPIGOT-6296
