From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 6 Jul 2022 16:15:42 +1000
Subject: [PATCH] Sync FireworkRocket's attachedToEntity


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 61252ef8a79bac0ea2d3d231ec6b12166f39d072..55b34372debe944d27c119e6d8ad027f3220cc3c 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -324,6 +324,11 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
             nbt.putUUID("SpawningEntity", this.spawningEntity);
         }
         // Paper end
+        // MultiPaper start
+        if (this.attachedToEntity != null) {
+            nbt.putUUID("AttachedToEntity", this.attachedToEntity.getUUID());
+        }
+        // MultiPaper end
     }
 
     @Override
@@ -345,6 +350,11 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
             this.spawningEntity = nbt.getUUID("SpawningEntity");
         }
         // Paper end
+        // MultiPaper start
+        if (nbt.hasUUID("AttachedToEntity") && level.getEntities().get(nbt.getUUID("AttachedToEntity")) instanceof LivingEntity livingEntity) {
+            this.attachedToEntity = livingEntity;
+        }
+        // MultiPaper end
     }
 
     @Override
