From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 26 Jun 2022 16:01:21 +1000
Subject: [PATCH] Add option to not save lighting updates


diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index 335f7e121b2b69a6b9b40bc87b1d675f41d8f6f6..7e46333c997e251f757661d50aa1b697084be771 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -45,6 +45,7 @@ import net.minecraft.server.MinecraftServer;
 
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
+import puregero.multipaper.config.MultiPaperConfiguration;
 
 public class ChunkHolder {
 
@@ -300,6 +301,7 @@ public class ChunkHolder {
         LevelChunk chunk = this.getSendingChunk(); // Paper - no-tick view distance
 
         if (chunk != null) {
+            if (!MultiPaperConfiguration.get().optimizations.dontSaveJustForLightingUpdates) chunk.setUnsaved(true); // MultiPaper
             int i = this.levelHeightAccessor.getSectionIndex(pos.getY());
 
             if (i < 0 || i >= this.changedBlocksPerSection.length) return; // CraftBukkit - SPIGOT-6086, SPIGOT-6296
