From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:44:09 +1000
Subject: [PATCH] Tick sleepCounter for external players


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index f8c736d24cbdb65c34d6167ec6fa731eea7658b4..e4fae438c89f664b71c84e0b50643e09aa69eb9e 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -97,6 +97,12 @@ public class MultiPaper {
                 }
 
                 externalPlayer.tickDeathIfDead();
+
+                if (externalPlayer.isSleeping()) {
+                    externalPlayer.sleepCounter = Math.min(100, externalPlayer.sleepCounter + 1);
+                } else {
+                    externalPlayer.sleepCounter = 0;
+                }
             }
 
             player.syncExperience();
