From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:44:09 +1000
Subject: [PATCH] Tick sleepCounter for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 488d14fca7c973dcf90480cc86baa8d307df264d..b5b34f5e068da99cf8d37cd56ae16a6576dea94c 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -167,7 +167,7 @@ public abstract class Player extends LivingEntity {
     public double xCloak;
     public double yCloak;
     public double zCloak;
-    public int sleepCounter;
+    public int sleepCounter; public int getSleepCounter() { return this.sleepCounter; } public void setSleepCounter(int sleepCounter) { this.sleepCounter = sleepCounter; } // MultiPaper - add getter and setter
     protected boolean wasUnderwater;
     private final Abilities abilities = new Abilities();
     public int experienceLevel;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index d1cf3c705dc7cd75752d19a87dcd854e230991d4..6278616b55cbba9da42e61fd8f90826e07a1c145 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -97,6 +97,12 @@ public class MultiPaper {
                 }
 
                 externalPlayer.tickDeathIfDead();
+
+                if (externalPlayer.isSleeping()) {
+                    externalPlayer.setSleepCounter(Math.min(100, externalPlayer.getSleepCounter() + 1));
+                } else {
+                    externalPlayer.setSleepCounter(0);
+                }
             }
 
             player.syncExperience();
