From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:44:09 +1000
Subject: [PATCH] Tick sleepCounter for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index bbb46853c6dc97383cb091dabf801164e90f9f12..7b36c7aaf093ad6144abcc2472c6c2e6a77d20c2 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -168,7 +168,7 @@ public abstract class Player extends LivingEntity {
     public double xCloak;
     public double yCloak;
     public double zCloak;
-    public int sleepCounter;
+    public int sleepCounter; public int getSleepCounter() { return this.sleepCounter; } public void setSleepCounter(int sleepCounter) { this.sleepCounter = sleepCounter; } // MultiPaper - add getter and setter
     protected boolean wasUnderwater;
     private final Abilities abilities = new Abilities();
     public int experienceLevel;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index 9255d7a9d3881666d8eee0263e17b85aa4f2a86c..63eb8c9884ff300fa70a7cc6fedbbae8a9b5ff7f 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -96,6 +96,12 @@ public class MultiPaper {
                 }
 
                 externalPlayer.tickDeathIfDead();
+
+                if (externalPlayer.isSleeping()) {
+                    externalPlayer.setSleepCounter(Math.min(100, externalPlayer.getSleepCounter() + 1));
+                } else {
+                    externalPlayer.setSleepCounter(0);
+                }
             }
 
             player.syncExperience();
