From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 7 Dec 2022 19:55:45 +1000
Subject: [PATCH] Process transient entity chunks on the master


diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
index e01bce0a30c997ca316f133346f49b87ea81fa1b..d4f86e4975f913b255dc828869c0bfe582c91ae9 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
@@ -1930,7 +1930,7 @@ public final class NewChunkHolder {
                     return false;
                 }
                 try {
-                    mergeFrom = RegionFileIOThread.loadData(this.world, this.chunkX, this.chunkZ, RegionFileIOThread.RegionFileType.ENTITY_DATA, PrioritisedExecutor.Priority.BLOCKING);
+                    mergeFrom = new CompoundTag(); // mergeFrom = RegionFileIOThread.loadData(this.world, this.chunkX, this.chunkZ, RegionFileIOThread.RegionFileType.ENTITY_DATA, PrioritisedExecutor.Priority.BLOCKING); // MultiPaper - Process transient entity chunks on the master
                 } catch (final Exception ex) {
                     LOGGER.error("Cannot merge transient entities for chunk (" + this.chunkX + "," + this.chunkZ + ") in world '" + this.world.getWorld().getName() + "', data on disk will be replaced", ex);
                 }
@@ -1943,6 +1943,7 @@ public final class NewChunkHolder {
                     return false;
                 } else {
                     EntityStorage.copyEntities(mergeFrom, save);
+                    save.putBoolean("multipaper.transient", true); // MultiPaper - Process transient entity chunks on the master
                 }
             }
             if (save == null && this.lastEntitySaveNull) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index b43d8c5b38798b5d9042992ca9ba167077bbc247..9a278e849af8bc9b2f75686d58a5cd06a8e77b5c 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -395,7 +395,7 @@ public class MultiPaper {
     }
 
     public static void writeRegionFileNBT(String world, String path, int cx, int cz, CompoundTag compoundTag) throws IOException {
-        writeRegionFile(world, path, cx, cz, nbtToBytes(compoundTag));
+        writeRegionFile(world, path, cx, cz, nbtToBytes(compoundTag), compoundTag != null && compoundTag.contains("multipaper.transient"));
     }
 
     public static CompoundTag readLevel(String world) throws IOException {
