From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 9 Jul 2022 12:34:14 +1000
Subject: [PATCH] Sync entity velocity changes


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index cd65dc29095926fd60a27b89030037cac46ee12a..d890347d516b1e700a489b13dd2eb24d08906eae 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -154,6 +154,7 @@ import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperEntitiesHandler;
 import puregero.multipaper.MultiPaperEntityIdManager;
 import puregero.multipaper.config.MultiPaperConfiguration;
+import puregero.multipaper.externalserverprotocol.AddDeltaMovementPacket;
 import puregero.multipaper.externalserverprotocol.EntityRemovePacket;
 import puregero.multipaper.externalserverprotocol.EntityUpdateNBTPacket;
 import puregero.multipaper.externalserverprotocol.EntityUpdateWithDependenciesPacket;
@@ -4346,6 +4347,17 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.setDeltaMovement(new Vec3(x, y, z));
     }
 
+    // MultiPaper start
+    public void addDeltaMovement(double x, double y, double z) {
+        addDeltaMovement(new Vec3(x, y, z));
+    }
+
+    public void addDeltaMovement(Vec3 velocity) {
+        setDeltaMovement(getDeltaMovement().add(velocity));
+        AddDeltaMovementPacket.broadcast(this, velocity);
+    }
+    // MultiPaper end
+
     public final int getBlockX() {
         return this.blockPosition.getX();
     }
diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 01477e7240f9e33d08d416a7d40ee10f3e5d4abf..da499acf294a5760ec22698f82287a8202855720 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -273,7 +273,7 @@ public class Explosion {
                             d14 = entity instanceof Player && level.paperConfig().environment.disableExplosionKnockback ? 0 : ProtectionEnchantment.getExplosionKnockbackAfterDampener((LivingEntity) entity, d13); // Paper - Disable explosion knockback
                         }
 
-                        entity.setDeltaMovement(entity.getDeltaMovement().add(d8 * d14, d9 * d14, d10 * d14));
+                        entity.addDeltaMovement(d8 * d14, d9 * d14, d10 * d14); // MultiPaper
                         if (entity instanceof Player) {
                             Player entityhuman = (Player) entity;
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/AddDeltaMovementPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/AddDeltaMovementPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..e34d8e9feec2b58f8b9ea6c9f0df4f4c54392a4e
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/AddDeltaMovementPacket.java
@@ -0,0 +1,75 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.phys.Vec3;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalPlayer;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperEntitiesHandler;
+
+import java.util.UUID;
+
+public class AddDeltaMovementPacket extends ExternalServerPacket {
+    private static boolean handlingPacket = false;
+    private final UUID world;
+    private final UUID entityUUID;
+    private final Vec3 velocity;
+
+    public AddDeltaMovementPacket(ServerLevel level, Entity entity, Vec3 velocity) {
+        this.world = level.getWorld().getUID();
+        this.entityUUID = entity.getUUID();
+        this.velocity = velocity;
+    }
+
+    public static void broadcast(Entity entity, Vec3 velocity) {
+        if (!handlingPacket) {
+            Entity controller = MultiPaperEntitiesHandler.getControllingPassenger(entity);
+            if (controller instanceof ExternalPlayer externalPlayer) {
+                externalPlayer.externalServerConnection.send(new AddDeltaMovementPacket((ServerLevel) entity.level, entity, velocity));
+            } else {
+                LevelChunk chunk = entity.level.getChunkIfLoaded(entity.blockPosition());
+                if (chunk != null && chunk.externalOwner != null && chunk.externalOwner.getConnection() != null) {
+                    chunk.externalOwner.getConnection().send(new AddDeltaMovementPacket((ServerLevel) entity.level, entity, velocity));
+                }
+            }
+        }
+    }
+
+    public AddDeltaMovementPacket(FriendlyByteBuf in) {
+        this.world = in.readUUID();
+        this.entityUUID = in.readUUID();
+        this.velocity = new Vec3(in.readDouble(), in.readDouble(), in.readDouble());
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(this.world);
+        out.writeUUID(this.entityUUID);
+        out.writeDouble(this.velocity.x);
+        out.writeDouble(this.velocity.y);
+        out.writeDouble(this.velocity.z);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            handlingPacket = true;
+            World world = Bukkit.getWorld(this.world);
+            if (world instanceof CraftWorld craftWorld) {
+                Entity entity = craftWorld.getHandle().getEntity(entityUUID);
+                if (entity != null) {
+                    entity.addDeltaMovement(velocity);
+                }
+            }
+            handlingPacket = false;
+        });
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 2fa9dba8f23008a9b6946a0b6ebe8a28733f51c2..3fe08e04f8669089f5cb64730401315808e78085 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -73,6 +73,7 @@ public class ExternalServerPacketSerializer {
         addPacket(RaidUpdatePacket.class, RaidUpdatePacket::new);
         addPacket(RaidJoinPacket.class, RaidJoinPacket::new);
         addPacket(SetPoiPacket.class, SetPoiPacket::new);
+        addPacket(AddDeltaMovementPacket.class, AddDeltaMovementPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
