From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Jul 2022 12:14:11 +1000
Subject: [PATCH] Sync game events


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 96784c08689dcb5f02b9db1c26e2ed2c7fcefe88..5da04d66c76773759cb4260739fc16235acfe69b 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -177,6 +177,7 @@ import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet; // Paper
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperChunkHandler;
 import puregero.multipaper.MultiPaperEntitiesHandler;
+import puregero.multipaper.MultiPaperGameEventHandler;
 import puregero.multipaper.MultiPaperInactiveTracker;
 import puregero.multipaper.externalserverprotocol.SpawnUpdatePacket;
 import puregero.multipaper.externalserverprotocol.TimeUpdatePacket;
@@ -1661,6 +1662,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             GameEvent.Message gameevent_b = (GameEvent.Message) iterator.next();
             GameEventListener gameeventlistener = gameevent_b.recipient();
 
+            if (MultiPaperGameEventHandler.onGameEvent(gameeventlistener, this, gameevent_b)) continue; // MultiPaper
             gameeventlistener.handleGameEvent(this, gameevent_b);
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
index 2d87b740e84435ef9bd3eee3d3e2af9f707a0826..89df7ae39be44ab299961795cca92c9fd6afbfc6 100644
--- a/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/src/main/java/net/minecraft/world/entity/animal/allay/Allay.java
@@ -70,6 +70,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
+import puregero.multipaper.MultiPaperGameEventHandler;
 
 public class Allay extends PathfinderMob implements InventoryCarrier {
 
@@ -575,7 +576,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier {
         this.level.addParticle(ParticleTypes.HEART, this.getRandomX(1.0D), this.getRandomY() + 0.5D, this.getRandomZ(1.0D), d0, d1, d2);
     }
 
-    private class AllayVibrationListenerConfig implements VibrationListener.VibrationListenerConfig {
+    class AllayVibrationListenerConfig implements VibrationListener.VibrationListenerConfig, MultiPaperGameEventHandler.SourceContainer { // MultiPaper - get allay
 
         AllayVibrationListenerConfig() {}
 
@@ -608,9 +609,13 @@ public class Allay extends PathfinderMob implements InventoryCarrier {
         public TagKey<GameEvent> getListenableEvents() {
             return GameEventTags.ALLAY_CAN_LISTEN;
         }
+
+        @Override public VibrationListener getVibrationListener() { return dynamicVibrationListener.getListener(); } // MultiPaper - add getter
+
+        @Override public Object getSource() { return Allay.this; } // MultiPaper - get allay
     }
 
-    private class JukeboxListener implements GameEventListener {
+    class JukeboxListener implements GameEventListener, MultiPaperGameEventHandler.SourceContainer { // MultiPaper - get allay
 
         private final PositionSource listenerSource;
         private final int listenerRadius;
@@ -642,5 +647,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier {
                 return false;
             }
         }
+
+        @Override public Object getSource() { return Allay.this; } // MultiPaper - get allay
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java b/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
index 27bd70dc30c8472e5a80f3273f9233a0392f831d..2ab710af5d9727b4c8a062e31b5def5cafd47ee6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
@@ -107,7 +107,7 @@ public class Warden extends Monster implements VibrationListener.VibrationListen
     public AnimationState diggingAnimationState = new AnimationState();
     public AnimationState attackAnimationState = new AnimationState();
     public AnimationState sonicBoomAnimationState = new AnimationState();
-    private final DynamicGameEventListener<VibrationListener> dynamicGameEventListener = new DynamicGameEventListener<>(new VibrationListener(new EntityPositionSource(this, this.getEyeHeight()), 16, this, (VibrationListener.ReceivingEvent) null, 0.0F, 0));
+    private final DynamicGameEventListener<VibrationListener> dynamicGameEventListener = new DynamicGameEventListener<>(new VibrationListener(new EntityPositionSource(this, this.getEyeHeight()), 16, this, (VibrationListener.ReceivingEvent) null, 0.0F, 0)); public VibrationListener getVibrationListener() { return dynamicGameEventListener.getListener(); } // MultiPaper - add getter
     private AngerManagement angerManagement = new AngerManagement(this::canTargetEntity, Collections.emptyList());
 
     public Warden(EntityType<? extends Monster> type, Level world) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index 22c309343299e60ed8028229b7f134109001ff35..4c1515fe7d5c05f4f15007e7f06cc5a7d7377d90 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -18,6 +18,7 @@ import net.minecraft.world.level.gameevent.BlockPositionSource;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.gameevent.GameEventListener;
 import net.minecraft.world.level.gameevent.PositionSource;
+import puregero.multipaper.MultiPaperGameEventHandler;
 
 public class SculkCatalystBlockEntity extends BlockEntity implements GameEventListener {
 
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
index 579a96d2d6c99d6587ea182f52b6df918f595f17..977334a13d0feab3a498da69c41c9f630dcc59c6 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
@@ -19,7 +19,7 @@ import org.slf4j.Logger;
 
 public class SculkSensorBlockEntity extends BlockEntity implements VibrationListener.VibrationListenerConfig {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private VibrationListener listener;
+    private VibrationListener listener; public VibrationListener getVibrationListener() { return listener; } // MultiPaper - add getter
     public int lastVibrationFrequency;
 
     public SculkSensorBlockEntity(BlockPos pos, BlockState state) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
index ec0eba674c940ee1d743cd32f9c2b690868757e0..58f462359ec2b29870cefad8745295139038e6df 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -54,7 +54,7 @@ public class SculkShriekerBlockEntity extends BlockEntity implements VibrationLi
     });
     private static final int SHRIEKING_TICKS = 90;
     public int warningLevel;
-    private VibrationListener listener = new VibrationListener(new BlockPositionSource(this.worldPosition), 8, this, (VibrationListener.ReceivingEvent)null, 0.0F, 0);
+    private VibrationListener listener = new VibrationListener(new BlockPositionSource(this.worldPosition), 8, this, (VibrationListener.ReceivingEvent)null, 0.0F, 0); public VibrationListener getVibrationListener() { return listener; } // MultiPaper - add getter
 
     public SculkShriekerBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.SCULK_SHRIEKER, pos, state);
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 4976caa63d06da4278174b7f646af77ed6f077f8..c47f23dca3c96d84992d4094278f2939aa8842c0 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -828,7 +828,7 @@ public class LevelChunk extends ChunkAccess {
         this.removeBlockEntityTicker(pos);
     }
 
-    private <T extends BlockEntity> void removeGameEventListener(T blockEntity, ServerLevel worldserver) {
+    public <T extends BlockEntity> void removeGameEventListener(T blockEntity, ServerLevel worldserver) { // MultiPaper - make public
         Block block = blockEntity.getBlockState().getBlock();
 
         if (block instanceof EntityBlock) {
@@ -1147,7 +1147,7 @@ public class LevelChunk extends ChunkAccess {
         });
     }
 
-    private <T extends BlockEntity> void addGameEventListener(T blockEntity, ServerLevel worldserver) {
+    public <T extends BlockEntity> void addGameEventListener(T blockEntity, ServerLevel worldserver) { // MultiPaper - make public
         Block block = blockEntity.getBlockState().getBlock();
 
         if (block instanceof EntityBlock) {
diff --git a/src/main/java/net/minecraft/world/level/gameevent/vibrations/VibrationListener.java b/src/main/java/net/minecraft/world/level/gameevent/vibrations/VibrationListener.java
index e45f54534bbf054eaf0008546ff459d4c11ddd50..f2af1a52a4a36846f9db623ba8ad47c918a8e929 100644
--- a/src/main/java/net/minecraft/world/level/gameevent/vibrations/VibrationListener.java
+++ b/src/main/java/net/minecraft/world/level/gameevent/vibrations/VibrationListener.java
@@ -30,13 +30,14 @@ import net.minecraft.world.phys.Vec3;
 import org.bukkit.craftbukkit.block.CraftBlock;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
 import org.bukkit.event.block.BlockReceiveGameEvent;
+import puregero.multipaper.MultiPaperGameEventHandler;
 // CraftBukkit end
 
 public class VibrationListener implements GameEventListener {
 
     protected final PositionSource listenerSource;
     public int listenerRange;
-    protected final VibrationListener.VibrationListenerConfig config;
+    public final VibrationListener.VibrationListenerConfig config; // MultiPaper - make public
     @Nullable
     protected VibrationListener.ReceivingEvent receivingEvent;
     protected float receivingDistance;
@@ -207,6 +208,8 @@ public class VibrationListener implements GameEventListener {
         void onSignalReceive(ServerLevel world, GameEventListener listener, BlockPos pos, GameEvent event, @Nullable Entity entity, @Nullable Entity sourceEntity, float distance);
 
         default void onSignalSchedule() {}
+
+        VibrationListener getVibrationListener(); // MultiPaper
     }
 
     public static record ReceivingEvent(GameEvent gameEvent, float distance, Vec3 pos, @Nullable UUID uuid, @Nullable UUID projectileOwnerUuid, @Nullable Entity entity) {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index a517ca77b3f8dc915c2ad1e1074037a9a67a2738..0428f7d8702c8f1e837a9a95d31dc501374880bf 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -102,6 +102,8 @@ public class MultiPaper {
                 } else {
                     externalPlayer.setSleepCounter(0);
                 }
+
+                externalPlayer.getWardenSpawnTracker().tick();
             }
 
             player.syncExperience();
diff --git a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
index 382257ce8b3f2b5361ce8a8e05e514ad5561c0b1..5b4e31e9576d0549e9dacdeef7a53d012b00c86c 100644
--- a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
@@ -167,7 +167,9 @@ public class MultiPaperChunkHandler {
             } else if (packet instanceof ClientboundBlockEntityDataPacket update) {
                 BlockEntity existingBlockEntity = blockUpdateChunk.getBlockEntity(update.getPos());
                 if (existingBlockEntity != null && existingBlockEntity.minecraftKey.toString().equals(update.getTag().getString("id"))) {
+                    ((LevelChunk) blockUpdateChunk).removeGameEventListener(existingBlockEntity, level);
                     existingBlockEntity.load(update.getTag());
+                    ((LevelChunk) blockUpdateChunk).addGameEventListener(existingBlockEntity, level);
                     holder.blockChanged(update.getPos());
                 } else if (!blockUpdateChunk.getBlockState(update.getPos()).hasBlockEntity() && depth < 1) {
                     MCUtil.scheduleTask(1, () -> handleBlockUpdate(world, packet, depth + 1));
diff --git a/src/main/java/puregero/multipaper/MultiPaperGameEventHandler.java b/src/main/java/puregero/multipaper/MultiPaperGameEventHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..5de0f032fc3398d7e5eaf81b87bad843d693945f
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperGameEventHandler.java
@@ -0,0 +1,77 @@
+package puregero.multipaper;
+
+import com.mojang.datafixers.util.Either;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.level.gameevent.GameEventListener;
+import net.minecraft.world.level.gameevent.vibrations.VibrationListener;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.externalserverprotocol.GameEventPacket;
+
+import java.util.Optional;
+
+public class MultiPaperGameEventHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(MultiPaperGameEventHandler.class.getSimpleName());
+
+    public interface SourceContainer {
+        Object getSource();
+    }
+
+    public static Either<Entity, BlockEntity> getEntity(Object object) {
+        if (object instanceof Entity entity) {
+            return Either.left(entity);
+        } else if (object instanceof BlockEntity blockEntity) {
+            return Either.right(blockEntity);
+        } else if (object instanceof SourceContainer sourceContainer) {
+            return getEntity(sourceContainer.getSource());
+        } else if (object instanceof VibrationListener vibrationListener) {
+            return getEntity(vibrationListener.config);
+        } else {
+            throw new IllegalArgumentException(object + " could not be converted to either an Entity or a BlockEntity!");
+        }
+    }
+
+    public static Optional<ExternalServer> getControllingServer(Either<Entity, BlockEntity> either) {
+        return either.map(
+                entity -> {
+                    Entity controller = MultiPaperEntitiesHandler.getControllingPassenger(entity);
+                    if (controller instanceof ExternalPlayer externalPlayer) {
+                        return Optional.of(externalPlayer.externalServerConnection.externalServer);
+                    } else if (controller instanceof ServerPlayer) {
+                        return Optional.empty();
+                    } else {
+                        LevelChunk chunk = entity.level.getChunkIfLoaded(entity.blockPosition());
+                        if (chunk != null && chunk.externalOwner != null && !chunk.externalOwner.isMe()) {
+                            return Optional.of(chunk.externalOwner);
+                        } else {
+                            return Optional.empty();
+                        }
+                    }
+                },
+                blockEntity -> {
+                    LevelChunk chunk = blockEntity.getLevel().getChunkIfLoaded(blockEntity.getBlockPos());
+                    if (chunk != null && chunk.externalOwner != null && !chunk.externalOwner.isMe()) {
+                        return Optional.of(chunk.externalOwner);
+                    } else {
+                        return Optional.empty();
+                    }
+                }
+        );
+    }
+
+    public static boolean onGameEvent(GameEventListener gameEventListener, ServerLevel world, GameEvent.Message event) {
+        Either<Entity, BlockEntity> either = getEntity(gameEventListener);
+        Optional<ExternalServer> controllingServer = getControllingServer(either);
+
+        controllingServer.ifPresent(
+                externalServer -> externalServer.getConnection().send(new GameEventPacket(either, world, event))
+        );
+        return controllingServer.isPresent();
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 7436f3c9565099769c8abcb66e8ba719bec9bde8..0aa7f26833c1e572e3339c5734b7d4c53b17781a 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -76,6 +76,7 @@ public class ExternalServerPacketSerializer {
         addPacket(RaidJoinPacket.class, RaidJoinPacket::new);
         addPacket(SetPoiPacket.class, SetPoiPacket::new);
         addPacket(AddDeltaMovementPacket.class, AddDeltaMovementPacket::new);
+        addPacket(GameEventPacket.class, GameEventPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/GameEventPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/GameEventPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..427f149dc4d3db9311c481ab318e6db51967ab4c
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/GameEventPacket.java
@@ -0,0 +1,132 @@
+package puregero.multipaper.externalserverprotocol;
+
+import com.mojang.datafixers.util.Either;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Registry;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.animal.allay.Allay;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.gameevent.GameEvent;
+import net.minecraft.world.level.gameevent.GameEventListener;
+import net.minecraft.world.level.gameevent.vibrations.VibrationListener;
+import net.minecraft.world.phys.Vec3;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.Optional;
+import java.util.UUID;
+
+public class GameEventPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(GameEventPacket.class.getSimpleName());
+
+    private final UUID world;
+    private final Either<UUID, BlockPos> id;
+    private final GameEvent gameEvent;
+    private final Vec3 sourcePos;
+    private final Optional<UUID> sourceUuid;
+    private final Optional<BlockState> affectedState;
+
+    public GameEventPacket(Either<Entity, BlockEntity> either, ServerLevel level, GameEvent.Message gameEvent) {
+        this.world = level.getWorld().getUID();
+        this.id = either.mapBoth(
+                Entity::getUUID,
+                BlockEntity::getBlockPos
+        );
+        this.gameEvent = gameEvent.gameEvent();
+        this.sourcePos = gameEvent.source();
+        this.sourceUuid = Optional.ofNullable(gameEvent.context().sourceEntity()).map(Entity::getUUID);
+        this.affectedState = Optional.ofNullable(gameEvent.context().affectedState());
+    }
+
+    public GameEventPacket(FriendlyByteBuf in) {
+        this.world = in.readUUID();
+        if (in.readBoolean()) {
+            this.id = Either.left(in.readUUID());
+        } else {
+            this.id = Either.right(in.readBlockPos());
+        }
+        this.gameEvent = in.readById(Registry.GAME_EVENT);
+        this.sourcePos = new Vec3(in.readDouble(), in.readDouble(), in.readDouble());
+        this.sourceUuid = in.readOptional(FriendlyByteBuf::readUUID);
+        this.affectedState = in.readOptional(buf -> buf.readById(Block.BLOCK_STATE_REGISTRY));
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(this.world);
+        out.writeBoolean(this.id.left().isPresent());
+        this.id.ifLeft(out::writeUUID);
+        this.id.ifRight(out::writeBlockPos);
+        out.writeId(Registry.GAME_EVENT, this.gameEvent);
+        out.writeDouble(this.sourcePos.x());
+        out.writeDouble(this.sourcePos.y());
+        out.writeDouble(this.sourcePos.z());
+        out.writeOptional(this.sourceUuid, FriendlyByteBuf::writeUUID);
+        out.writeOptional(this.affectedState, (buf, value) -> buf.writeId(Block.BLOCK_STATE_REGISTRY, value));
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerLevel level = ((CraftWorld) Bukkit.getWorld(this.world)).getHandle();
+            Object entity = resolveId(level);
+            if (entity == null) {
+                LOGGER.warn("Could not find entity for " + this.id);
+                return;
+            }
+
+            if (entity instanceof Allay allay) {
+                allay.updateDynamicGameEventListener(
+                        (dynamicGameEventListener, serverLevel) -> handleGameEvent(dynamicGameEventListener.getListener(), level, entity)
+                );
+            } else {
+                GameEventListener gameEventListener = getGameEventListener(entity);
+                handleGameEvent(gameEventListener, level, entity);
+            }
+        });
+    }
+
+    private void handleGameEvent(GameEventListener gameEventListener, ServerLevel level, Object entity) {
+        gameEventListener.handleGameEvent(level, new GameEvent.Message(gameEvent, sourcePos, GameEvent.Context.of(sourceUuid.map(level::getEntity).orElse(null), affectedState.orElse(null)), gameEventListener, getPos(entity)));
+    }
+
+    private GameEventListener getGameEventListener(Object entity) {
+        if (entity instanceof GameEventListener gameEventListener) {
+            return gameEventListener;
+        } else if (entity instanceof VibrationListener.VibrationListenerConfig vibrationListenerConfig) {
+            return vibrationListenerConfig.getVibrationListener();
+        } else {
+            throw new IllegalArgumentException("Could not figure out how to convert " + entity + " to a GameEventListener");
+        }
+    }
+
+    private Vec3 getPos(Object object) {
+        if (object instanceof Entity entity) {
+            return entity.position();
+        } else if (object instanceof BlockEntity blockEntity) {
+            return new Vec3(blockEntity.getBlockPos().getX() + 0.5, blockEntity.getBlockPos().getY(), blockEntity.getBlockPos().getZ() + 0.5);
+        } else {
+            throw new IllegalArgumentException(object + " could not be converted to either an Entity or a BlockEntity to get it's position!");
+        }
+    }
+
+    private Object resolveId(ServerLevel level) {
+        return this.id.map(
+                level::getEntity,
+                blockPos -> {
+                    LevelChunk levelChunk = level.getChunkIfLoaded(blockPos);
+                    return levelChunk == null ? null : levelChunk.getBlockEntity(blockPos);
+                }
+        );
+    }
+}
