From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Fri, 2 Sep 2022 22:33:38 +0400
Subject: [PATCH] add mpdebugbossbar command


diff --git a/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java b/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
index 75c47d4c51e41cce641d79b5cd3d0bb4eccd5cff..9ebfe24803c926b8c1b23bd2626e4280a55c0c10 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
@@ -4,10 +4,7 @@ import java.util.Map;
 import org.bukkit.Server;
 import org.bukkit.command.Command;
 import org.bukkit.command.SimpleCommandMap;
-import puregero.multipaper.commands.MPDebugCommand;
-import puregero.multipaper.commands.MPMapCommand;
-import puregero.multipaper.commands.SListCommand;
-import puregero.multipaper.commands.ServersCommand;
+import puregero.multipaper.commands.*;
 
 public class CraftCommandMap extends SimpleCommandMap {
 
@@ -29,6 +26,7 @@ public class CraftCommandMap extends SimpleCommandMap {
         register("multipaper", new ServersCommand("servers"));
         register("multipaper", new MPMapCommand("mpmap"));
         register("multipaper", new SListCommand("slist"));
+        register("multipaper", new MPDebugBossBarCommand("mpdebugbossbar"));
     }
     // MultiPaper end
 
