From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mohammed jasem alaajel <xrambad@gmail.com>
Date: Sat, 19 Nov 2022 17:18:31 +0400
Subject: [PATCH] Make sure if Owner is ExternalPlayer to load chunk that
 currently in


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
index 7dc8cc93e8da37e1096c3263c1c8fc55bbf60401..e05153a699f7022e004b7169f290a9b8be09b138 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -16,7 +16,11 @@ import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
 // CraftBukkit start
 import org.bukkit.Location;
 import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.entity.Player;
 import org.bukkit.event.entity.EntityTeleportEvent;
+// MultiPaper start
+import puregero.multipaper.ExternalPlayer;
+// MultiPaper end
 // CraftBukkit end
 
 public class FollowOwnerGoal extends Goal {
@@ -143,6 +147,12 @@ public class FollowOwnerGoal extends Goal {
     }
 
     private boolean canTeleportTo(BlockPos pos) {
+        // MultiPaper start
+        if (owner instanceof ExternalPlayer externalPlayer) {
+            Player player = externalPlayer.getBukkitEntity();
+            player.getWorld().getChunkAtAsync(player.getLocation());
+        }
+        // MultiPaper end
         BlockPathTypes pathtype = WalkNodeEvaluator.getBlockPathTypeStatic(this.level, pos.mutable());
 
         if (pathtype != BlockPathTypes.WALKABLE) {
